---
title: "FigureYa257armCNV"
author: "Xiaofan Lu"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

比较Im-H和Im-L两个组臂级扩增和缺失频率和焦点级扩增和缺失水平，画出图4B的最后两个图。示例数据随便就行。

# Requirement Description

Compare the arm-level amplification and deletion frequencies and focus-level amplification and deletion levels between the two groups of Im-H and Im-L, and plot the last two plots of Figure 4B. Sample data is just fine.

![](example.png)

出自<https://linkinghub.elsevier.com/retrieve/pii/S2001037021003366>
from<https://linkinghub.elsevier.com/retrieve/pii/S2001037021003366>

Fig. 4. Molecular features of the immune subtypes of melanoma.
B. Comparisons of tumor mutation burden (TMB), homologous recombination deficiency (HRD) scores, and global methylation levels among the three immune subtypes in TCGA-melanoma;
prediction of **high-immune-score (upper third) versus low-immune-score (bottom third)** melanomas using **TMB and somatic copy number alteration (SCNA) score** by the logistic regression model;
comparisons of **arm-level amplification and deletion frequencies** and **focal-level amplification and deletion levels** between Im-H and Im-L melanomas.

# 应用场景

循环计算不同组别的扩增和缺失比例，对臂水平扩增和缺失做检验，并绘图。

更多基因组相关FigureYa看这里<https://k.youshop10.com/gZBI0YzR>

另外，Figure 5E、7D、9这种分布图的画法可参考FigureYa227boxdensity。

# Application Scenarios

The amplification and deletion ratios of different groups were calculated in cycles, and the horizontal amplification and deletion of the arm were tested and plotted.

For more genome-related FigureYa see here <https://k.youshop10.com/gZBI0YzR>

In addition, for the distribution chart of Figure 5E, 7D, and 9, please refer to FigureYa227boxdensity.

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")

```

加载包
load packages

```{r}
library(ggplot2)
library(patchwork)
library(data.table)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入文件

用GenePattern上的GISTIC分析可得到broad_data_by_genes.txt以及all_data_by_genes.txt文件，或可参照FigureYa79CNV。

easy_input_annotation.txt，样本分组信息。

# Input files

The broad_data_by_genes.txt and all_data_by_genes.txt files can be obtained by GISTIC analysis on GenePattern, or refer to FigureYa79CNV.

easy_input_annotation.txt. Sample grouping information.
