---
title: "FigureYa145target"
author: "Long Zhao"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述
## Requirement description

从形式上复现原图。
Reproduce the original image in form.

有时没有现成的包能画出一模一样的图，我们可以通过计算图中每个元素的位置，然后画出想要的效果。
Sometimes there is no ready-made package that can draw the same image. We can calculate the position of each element in the image and then draw the desired effect.

![](example.png)

出自<https://www.nature.com/articles/s41467-019-09222-w>
From <https://www.nature.com/articles/s41467-019-09222-w>

Fig. 2 Characteristics of glycosites identified with AI-ETD. f A glycoprotein-glycan network maps **which glycans (outer circle, 117 total) modify which proteins (inner bar, 771 total)**. Glycoproteins are **sorted by number of glycosites (scale to the right)**. Glycans are organized by classification, and **edges are colored by the glycan node from which they originate**, except for mannose-6-phosphate which has yellow edges. See Supplementary Fig. 11 and Supplementary Table 1 for glycan identifiers.

A glycoprotein-glycan network diagram in Fig. 2f maps which glycans (outer nodes) were observed on identified glycoproteins (inner column, organized by number of glycosites). Several discernable patterns appear, perhaps most notably the prevalence of high mannose glycosylation. The network diagram also indicates that the majority of **fucosylated, paucimannose, and sialylated glycans occur on proteins with multiple glycosylation sites**, and it indicates which glycans contribute more to heterogeneity. Supplementary Figure 11 provides a larger version of this network diagram with glycan identities in Supplementary Table 1.

**图的解析**
**Analysis of the graph**

这个图展示两层信息：糖蛋白上glycan的种类，糖蛋白上glycosites的数量（数量可替换成其他分类信息，例如上调/下调，或所在的通路等等）。
This graph shows two layers of information: the types of glycans on glycoproteins and the number of glycosites on glycoproteins (the number can be replaced by other classification information, such as up-regulation/down-regulation, or the pathway in which it is located, etc.).

- 外围弧线：classification，把glycans分为6种
- 点：117个glycans，按照classification定义颜色和位置，并与外围6个classification的弧线对应
- 中央矩形：看起来是矩形，其实每个矩形的纵坐标都是有意义的，其本质是771个糖蛋白，按照glycosites的数量排序，数量与右侧图例对应。
- 连线：连接外围点跟中央矩形。线的两端分别对应糖蛋白的glycan和glycosites数量。颜色跟classification一致，其中一个glycan：mannose-6-phosphate的连线用黄色突出显示。
- Outer arcs: classification, which divides glycans into 6 types
- Points: 117 glycans, with colors and positions defined by classification, and corresponding to the arcs of the 6 outer classifications
- Central rectangle: It looks like a rectangle, but in fact the vertical coordinates of each rectangle are meaningful. In essence, it is 771 glycoproteins, sorted by the number of glycosites, and the number corresponds to the legend on the right.
- Lines: Connect the outer points with the central rectangle. The two ends of the line correspond to the number of glycans and glycosites of the glycoprotein, respectively. The color is consistent with the classification, and the line connecting one of the glycans: mannose-6-phosphate is highlighted in yellow.

原文方法描述是用igraph画的：The **protein-glycan network**, glycan co-occurrence networks, and glycosylation profiles for subcellular groups were created in R 3.2.2 using the **igraph** library70, and the arcplot were created with arcdiagram library。
The original method description is drawn with igraph: The **protein-glycan network**, glycan co-occurrence networks, and glycosylation profiles for subcellular groups were created in R 3.2.2 using the **igraph** library70, and the arcplot were created with arcdiagram library.

这里用ggplot2画图，先计算位置再画图。
Here we use ggplot2 to draw the graph, first calculate the position and then draw the graph.

## 应用场景
## Application scenario

展示分类、分组跟元素之间的关系。例如：glycan-数量-glycoprotein，GO-TF-targe gene、表观修饰-上下调-gene、enhancer-motif-promoter等。
Show the relationship between classification, grouping and elements. For example: glycan-quantity-glycoprotein, GO-TF-target gene, epigenetic modification-up and down regulation-gene, enhancer-motif-promoter, etc.

以TF-GO-target gene互作为例，转录组找出几个GO 途径的基因可以放在外围的点，颜色代表不同GO term，点的大小代表基因表达水平或fold change。里面的矩形是我们共公共数据库找到的能够结合这些外围基因的转录因子。结果哪些转录因子激活基因，哪些抑制，就一目了然了。如果有HiC或ChIA-pet数据，hub 位点和靶基因的互作也是同理。
Taking the TF-GO-target gene interaction as an example, the transcriptome finds several genes in the GO pathway and can be placed on the outer points. The color represents different GO terms, and the size of the point represents the gene expression level or fold change. The rectangle inside is the transcription factor that can bind to these peripheral genes found in our public database. As a result, it is clear at a glance which transcription factors activate genes and which inhibit them. If there is HiC or ChIA-pet data, the interaction between hub sites and target genes is the same.

## 环境设置
## Environment settings


```{r}
source("install_dependencies.R")

```

加载包
Load package

```{r}
library(ggplot2) #用于画图 #For drawing
library(dplyr) #用于数据处理 #For data processing
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 #Display English error message
options(stringsAsFactors = FALSE) #禁止chr转成factor #Prohibit chr from being converted to factor
```

## 输入文件
## Input file

easy_input_node.txt和easy_input_group.txt。从例文41467_2019_9222_MOESM5_ESM.xlsx文件整理而来。
Organized from the example file 41467_2019_9222_MOESM5_ESM.xlsx.

- easy_input_node.txt，glycan及其分类，用来画外围的点和弧线；
- easy_input_group.txt，protein-glycan-数量的对应关系，用来画中央矩形和连线。每行一个protein，用`;`分隔多个glycon。
- easy_input_node.txt, glycan and its classification, used to draw the outer points and arcs;
- easy_input_group.txt, the correspondence between protein-glycan-quantity, used to draw the central rectangle and connecting lines. One protein per line, separate multiple glycans with `;`.

```{r}
Glycan <- read.table("easy_input_node.txt", sep = "\t", head=T)
head(Glycan)
dim(Glycan)

protein <- read.table("easy_input_group.txt",head=T)
head(protein)
dim(protein)
```

## 计算图中各元素位置
## Calculate the position of each element in the graph

### 计算点（Glycan）的坐标
### Calculate the coordinates of the point (Glycan)

一共有117个点，右侧（第一、四象限）59个，左侧（二、三象限）58个。
There are 117 points in total, 59 on the right (first and fourth quadrants) and 58 on the left (second and third quadrants).

先计算每个点与x轴的角度，再通过与x轴的角度和三角函数计算出每个点的横纵坐标。
First calculate the angle between each point and the x-axis, and then calculate the horizontal and vertical coordinates of each point through the angle with the x-axis and trigonometric functions.

计算角度时，因为不是整个圆，上下各缺了一块，缺的角度为30度。所以第一个蛋白角度是75度，第59个蛋白是-75度，第60个蛋白是-107度，最后一个是-255度。
When calculating the angle, because it is not a complete circle, there is a missing piece on the top and bottom, and the missing angle is 30 degrees. So the angle of the first protein is 75 degrees, the 59th protein is -75 degrees, the 60th protein is -107 degrees, and the last one is -255 degrees.

计算坐标时，考虑到后面还要画中央矩形，因此以蛋白数量/2作为半径。
When calculating the coordinates, we took into account that we would need to draw a central rectangle later, so we used the number of proteins/2 as the radius.

```{r, fig.width=5, fig.height=5}
Glycan$angle <- ifelse(Glycan$Node < 60,
                       75-(150/58*(Glycan$Node-1)),
                       75-(150/58*(Glycan$Node-1))-30)

Glycan$G.x <- cos(Glycan$angle*pi/180.0)*(nrow(protein)/2)
Glycan$G.y <- sin(Glycan$angle*pi/180.0)*(nrow(protein)/2)
head(Glycan)
```

### 计算最外围的弧线
### Calculate the outermost arc

外围的弧线内侧是圆半径的1.1倍，外侧是1.15倍
The inner side of the outer arc is 1.1 times the radius of the circle, and the outer side is 1.15 times
