---
title: "FigureYa43ManhattanV2"
author: "Zhaodong Hao; Ying Ge, Yijing Chen"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述
## Requirement description

画出paper里的曼哈顿图，文字标签好看些。

Draw a picture of Manhattan from the paper, the text labels look better.

![](example.png)

出自<https://www.nature.com/articles/s41588-018-0045-8>

from<https://www.nature.com/articles/s41588-018-0045-8>

## 应用场景
## Application scenario

展示每条染色体上的SNP分布，通常作为外显子组测序文章里的第一个图。

Show the distribution of SNP on each chromosome, usually presented as the first figure in an exon sequencing article.

## 输入文件
## Input file

easy_input.csv，包括染色体名字、SNP编号、在染色体上的位置（第几个碱基）、pvalue、类型type、基因名gene。

其中type和gene，非必需，如果能提供基因名，就写在gene列；如果要用不同形状来展示不同类的SNP，会把类型写到type。

easy_input.csv, including chromosome name, SNP number, position on the chromosome (which base), pvalue, type, gene name.

Type and gene are not required, if the gene name can be provided, it should be written in the gene column; if different shapes are used to display different classes of SNPs, the type will be written in type.

```{r,message=FALSE}
source("install_dependencies.R")
library(dplyr)

gwas <- read.csv("easy_input.csv", as.is = T)
head(gwas)
str(gwas)

gwas$chr <- factor(gwas$CHR,levels=c(1:18,"X","Y"))#给染色体排序 order chromosomes
```

如果自己不提供染色体长度，就用SNP位置最大值作为染色体长度，运行下面这段

If you don't provide the chromosome length yourself, use the SNP position maximum as the chromosome length and run the following paragraph

```r
chrx <- gwas %>%
  group_by(chr) %>%
  summarise(chr_len=max(POS)) %>%

  #计算每条染色体在x轴上的起始位置
  #calculate the starting position of each chromosome on the x-axis
  mutate(chrx=cumsum(chr_len)-chr_len) %>%
  select(-chr_len) %>% #删除染色体长度 delete chromosome length
  data.frame()
```

Karyotype.txt，如果你自己提供染色体长度文件，就运行下面这段

Karyotype.txt, if you provide your own chromosome length file, run the following paragraph

```{r}
chrom <- read.table("Karyotype.txt", header = T, as.is = T)
head(chrom)
#计算每条染色体在x轴上的起始位置
#calculate the starting position of each chromosome on the x-axis
chrom$chrx <- c(0,cumsum(as.numeric(chrom$chr_len))[-nrow(chrom)])
chrx <- data.frame(chr = chrom$chr, chrx = chrom$chrx)
chrx$chr <- factor(chrx$chr,levels=c(1:18,"X","Y"))#给染色体排序 order chromosomes
```

## 计算每个SNP的坐标
## Calculate the coordinates of each SNP
