---
title: "FigureYa19 Lollipop"
author: "Dekang Lv; Ying Ge, Yijing Chen"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## éœ€æ±‚æè¿°
## Requirement description

ç”¨Rä»£ç ç”»å‡ºpaperé‡Œçš„Lollipopï¼Œæ£’æ£’ç³–ğŸ­ã€‚

Draw the Lollipop in the paper using R code, lollipop ğŸ­.

![](example.png)

å‡ºè‡ª<https://www.nature.com/articles/nature22973>

from<https://www.nature.com/articles/nature22973>

## åº”ç”¨åœºæ™¯
## Application scenario

æŸä¸ªåŸºå› çš„çªå˜å‘ç”Ÿåœ¨å“ªä¸ªé‡è¦çš„ç»“æ„åŸŸï¼Ÿæ¨æµ‹è¯¥ä½ç‚¹çš„çªå˜å½±å“äº†è›‹ç™½è´¨åŠŸèƒ½ã€‚

åœºæ™¯ä¸€ï¼šå±•ç¤ºTCGAæŸä¸ªç™Œç—‡ç±»å‹é‡ŒæŸä¸ªåŸºå› çš„çªå˜ä½ç‚¹ã€‚

åœºæ™¯äºŒï¼šå¯¹æ¯”ä¸åŒç™Œç—‡/äºšå‹çš„çªå˜ä½ç‚¹ã€‚

In which important structural domain does the mutation of a gene occur? It is speculated that the mutation of this site affects protein function.

Scenario 1: Show the mutation site of a gene in a cancer type in TCGA.

Scenario 2: Compare the mutation sites of different cancers/subtypes.

## ç¯å¢ƒè®¾ç½®
## Environment setting

```r

install.packages("BiocManager")
library(BiocManager)
install("trackViewer")
install("TCGAbiolinks")

Sys.setenv(LANGUAGE = "en") #æ˜¾ç¤ºè‹±æ–‡æŠ¥é”™ä¿¡æ¯ display English error messages
options(stringsAsFactors = FALSE) #ç¦æ­¢chrè½¬æˆfactor prohibit the conversion of chr to factor
```

```{r}
source("install_dependencies.R")
library(TCGAbiolinks)
library(maftools)
library(trackViewer)
library(RColorBrewer)
```

## ä¸‹è½½TCGAæ•°æ®
## Download TCGA data

å¦‚æœä½ è¦ç”»çš„åŸºå› çªå˜ä¿¡æ¯å·²ç»ä¿å­˜åœ¨ç±»ä¼¼`easy_input.csv`çš„æ–‡ä»¶é‡Œï¼Œå°±å¯ä»¥è·³è¿‡è¿™æ­¥ï¼Œç›´æ¥è¿›å…¥â€œè¾“å…¥æ•°æ®â€ã€‚

ä¸‹è½½çªå˜ä¿¡æ¯ä½œä¸ºè¾“å…¥æ•°æ®ã€‚

If genetic mutation information you want to draw is already saved in a file like `easy_input.csv`, you can skip this step and go directly to `Input data`.

Download the mutation information as input data.

```{r,message=FALSE,warning=FALSE}
mutquery <- GDCquery(project = "TCGA-LIHC",
                     data.category = "Simple Nucleotide Variation",
                     data.type = "Masked Somatic Mutation",
                     workflow.type = "Aliquot Ensemble Somatic Variant Merging and Masking")
GDCdownload(mutquery)
mut<-GDCprepare(mutquery)

#è¿è¡Œä¸‹é¢è¿™è¡Œï¼Œä¼šæŠŠæ‰€æœ‰åŸºå› çš„çªå˜æ•°æ®ä¿å­˜åˆ°æ–‡ä»¶é‡Œï¼Œä½ å¯ä»¥ç”¨è¿™ä¸ªæ–‡ä»¶åšæ›´å¤šçš„äº‹æƒ…ã€‚
#run the line below will save the mutation data for all the genes to a file, which you can use to do more.
#write.csv(mut,"TCGA_SNV.csv")

#ç”»æ£’æ£’ç³–å›¾åªéœ€è¦å…¶ä¸­çš„ä¸€ä¸ªåŸºå› ï¼Œæˆ‘ä»¬å•ç‹¬æŠ½æå‡ºæ¥ï¼Œæ­¤å¤„æå–TP53ã€‚
#draw a lollipop plot requires only one of these genes, which we extract separately, here we extract TP53.
write.table(mut[mut$Hugo_Symbol == "TP53",],"easy_input.txt",row.names = F,quote = F,sep = "\t")
```

## è¾“å…¥æ•°æ®
## Input data

éœ€è¦ç»“æ„åŸŸå’Œçªå˜ä¿¡æ¯ã€‚

Structural domains and mutation information are required.

### çªå˜ä¿¡æ¯
### Mutation information

æ­¤å¤„ç›´æ¥ç”¨TCGAçš„çªå˜æ•°æ®ï¼Œæœ‰å¾ˆå¤šåˆ—ï¼Œå…¶ä¸­`RefSeq``HGVSp` `HGVSp_Short`ä¸‰åˆ—æ˜¯å¿…é¡»çš„ã€‚

å› æ­¤ï¼Œä½ ä¹Ÿå¯ä»¥åªå‡†å¤‡è¿™ä¸‰åˆ—ã€‚

Here, the mutation data from TCGA is used directly, with many columns, of which three columns `RefSeq` `HGVSp` `HGVSp_Short` are mandatory.

Therefore, you can also prepare only these three columns.

```{r}
df<-read.table("easy_input.txt",as.is = T,sep = "\t",header = T)

#ç­›æ‰æ°¨åŸºé…¸æœªçªå˜çš„è¡Œ
#screen out rows with unmutated amino acids
newdf<-df[!is.na(df$HGVSp),]
```

### ç»“æ„åŸŸ
### Structural domains

æ ¹æ®åŸºå› åï¼ˆæ­¤å¤„æ˜¯TP53ï¼‰å’Œrefseq IDä»maftoolsæä¾›çš„è›‹ç™½domainæ³¨é‡Šä¸­æå–æŸä¸ªåŸºå› çš„domainæ³¨é‡Š

Extract the domain annotation of a gene from the protein domain annotation provided by maftools according to the gene name (TP53 here) and refseq ID

```{r,message=FALSE}
gff = readRDS(file = system.file('extdata', 'protein_domains.RDs', package = 'maftools'))
refseqid <- strsplit(x = as.character(newdf$RefSeq), split = '.', fixed = TRUE)[[1]][1]
protein_inform <- gff[HGNC %in% "TP53"][refseq.ID == refseqid,]
```

### ä»çªå˜æ³¨é‡Šä¸­æå–æ°¨åŸºé…¸ä½ç½®
### Extract amino acid positions from mutation annotations

å…ˆå†™ä¸ªå‡½æ•°ï¼Œç”¨äºä»çªå˜æ³¨é‡Šä¸­æå–æ°¨åŸºé…¸ä½ç½®

First write a function for extracting amino acid positions from mutation annotations

```{r}
extractpos <- function(maf_aachange){
  prot.spl = strsplit(x = as.character(maf_aachange), split = '.', fixed = TRUE)
  prot.conv = sapply(sapply(prot.spl, function(x) x[length(x)]), '[', 1)
  pos = gsub(pattern = 'Ter.*', replacement = '',x = prot.conv)
  pos = gsub(pattern = '[[:alpha:]]', replacement = '', x = pos)
  pos = gsub(pattern = '\\*$', replacement = '', x = pos)
  pos = gsub(pattern = '^\\*', replacement = '', x = pos)
  pos = gsub(pattern = '\\*.*', replacement = '', x = pos)
  pos = as.numeric(sapply(X = strsplit(x = pos, split = '_', fixed = TRUE), FUN = function(x) x[1]))
  aa = paste0(unlist(regmatches(maf_aachange, gregexpr("p[.].[0-9]+",maf_aachange))),"X")
  mutpos = data.frame(position = pos, mutation = maf_aachange, aa = aa, stringsAsFactors = F)
  return(mutpos[order(mutpos$pos),])
}
```

ä»çªå˜æ³¨é‡Šä¸­æå–æ°¨åŸºé…¸ä½ç½®

Extract amino acid positions from mutation annotations

```{r}
pos <- extractpos(newdf$HGVSp_Short)
head(pos)
```

ç»Ÿè®¡ä¸åŒæ°¨åŸºé…¸çªå˜ç±»å‹çš„çªå˜é¢‘ç‡

Count the mutation frequency of different amino acid mutation types

```{r}
nrpos <- pos[!duplicated(pos),]
rownames(nrpos) <- nrpos$mutation
nrpos$rate <- 1
nrpos[names(table(pos$mutation)),]$rate <-table(pos$mutation)
head(nrpos)
```

## å¼€å§‹ç”»å›¾
## Start drawing
