---
title: "FigureYa157ChIPpvalue"
author: "Wen Wang"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Requirement

计算ChIP-seq样本间的p-value。
Calculate p-values between ChIP-seq samples.

![](example.png)

出自<https://www.nature.com/articles/s41588-018-0044-9>
From <https://www.nature.com/articles/s41588-018-0044-9>

Fig. 1 | Deregulated MYCN binds active chromatin and amplifies transcription in neuroblastoma.

g, RNA Pol II meta-gene across **all active genes upon MYCN shutdown**. The regions encompassing the TSS and gene body are enlarged, and significance is indicated. Bottom right, distribution plots of RNA Pol II traveling ratio (TR) for all active genes. Differences in the TR distributions at 0 and 2 h are significant (Welch’s two-tailed t test): ***P < 1 × 10−9, **P < 1 × 10−6, *P < 1 × 10–3.

# 应用场景
# Application Scenarios

用于判断全基因组范围的结合是否有global change，适用于校正过的ChIP-seq数据。
Used to determine whether there is a global change in genome-wide binding. Applicable to corrected ChIP-seq data.

- 加spike-in的定量ChIP-seq，例如这篇用到的ChIP-RX<https://mp.weixin.qq.com/s/tY-MunnBeFQ9A4_Svswu0g>
- 校正过的普通ChIP-seq，校正方法看这篇<https://www.notion.so/Paper-191102-ChIPseqSpikeInFree-8e225f58252144d0b1a0a15f99020b66>
- Quantitative ChIP-seq with spike-in, such as the ChIP-RX used in this article <https://mp.weixin.qq.com/s/tY-MunnBeFQ9A4_Svswu0g>
- Standard ChIP-seq with calibration. For calibration methods, see this article <https://www.notion.so/Paper-191102-ChIPseqSpikeInFree-8e225f58252144d0b1a0a15f99020b66>

# 环境设置
# Environment Setup

根据自己的系统选择下载链接
Select the download link based on your operating system.
