---
title: "FigureYa208FPI"
params:
  author: "Xiaofan Lu; Yasi Zhang"
  reviewer: "Ying Ge"
  date: "2025-05-20"
output: html_document
---

**Author(s)**: `r params$author`
**Reviewer(s)**: `r params$reviewer`
**Date**: `r params$date`

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Requirements Description

想众筹铁死亡指数（FPI）的计算方法，根据基因表达，算出ssGSEA富集分数，再计算FPI，并用GSE121689数据集验证。

I want to crowdfund the calculation method of the Ferroptosis Potential Index (FPI), which involves calculating ssGSEA enrichment scores based on gene expression, then deriving the FPI, and validating it using the GSE121689 dataset.

![](example.png)

图2. FPI与癌症组织学类型和分子亚型的关系。
(A) 不同癌症类型中肿瘤组织与正常组织间FPI的差异比较。

Figure 2. The Relations between FPI and Histological Types and Molecular Subtypes among Cancers.
(A) The different FPIs between tumor and normal tissues among cancers.

![](method1.png)

![](method2.png)

出自：<https://linkinghub.elsevier.com/retrieve/pii/S2589004220304892>

Source: <https://linkinghub.elsevier.com/retrieve/pii/S2589004220304892>

# 应用场景
# Application Scenario

例文通过查文献收集基因，用GSVA计算正负两个基因集的富集分数，然后计算FPI。

可批量计算多组（TCGA pancancer RNA-seq）、也可以单独计算一个数据集（GEO表达谱数据）。

另外，不仅可以计算铁死亡指数FPI，还可以按照这个方法，替换为自己收集的某一功能相关的基因集，自己自定义一个某某指数。

The example demonstrates collecting genes from literature, calculating enrichment scores for both positive and negative gene sets using GSVA, and then deriving the FPI.

This method can be applied to batch-process multiple datasets (e.g., TCGA pan-cancer RNA-seq) or analyze a single dataset (e.g., GEO expression profile data).

Additionally, it is not limited to calculating the Ferroptosis Potential Index (FPI). By replacing the gene sets with custom collections related to a specific function, users can define their own "XX Index" following the same approach.

# 环境设置
# Environment Setup


```{r}
source("install_dependencies.R")

```

# 加载包
# Loading packages=

```{r}
library(data.table)
library(GSVA)
library(ggplot2)

# 显示英文报错信息
# Show English error messages
Sys.setenv(LANGUAGE = "en")

# 禁止chr转成factor
# Prevent character-to-factor conversion
options(stringsAsFactors = FALSE)
```

# 输入文件
# Input Files

大文件已上传到微云：<https://share.weiyun.com/c8GQyxR4>

Large files have been uploaded to Weiyun: <https://share.weiyun.com/c8GQyxR4>

## 下载TCGA pan-cancer数据
## Download TCGA Pan-Cancer Data

GSVA的输入数据可以是microarray data，也可以是RNA-seq count、CPM、RPKM、TPM。

原文：We calculate now GSVA enrichment scores for these gene sets using first the microarray data and then the **RNA-seq integer count** data. Note that the only requirement to do the latter is to set the argument **kcdf="Poisson"** which is "Gaussian" by default. Note, however, that if our RNA-seq derived expression levels would be **continous, such as log-CPMs, log-RPKMs or log-TPMs**, the the default value of the **kcdf argument should remain unchanged**.

出自：<https://www.bioconductor.org/packages/release/bioc/vignettes/GSVA/inst/doc/GSVA.pdf>

- 表达矩阵和Gene mapping：
  - 这里跟FigureYa55panCancer_violin保持一致，从[XENA](https://xenabrowser.net/datapages/)下载UCSC Toil RNA-seq Recompute TPM：[TCGA Pan-Cancer (PANCAN) (41 datasets)
](https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)里的[TOIL RSEM tpm (n=10,535) UCSC Toil RNA-seq Recompute](https://toil.xenahubs.net/download/tcga_rsem_isoform_tpm.gz)，[Gene Mapping](https://toil.xenahubs.net/download/probeMap/gencode.v23.annotation.transcript.probemap)
  - 或者下载GDC pipeline的FPKM-UQ：[GDC Pan-Cancer (PANCAN) (17 datasets)](https://xenabrowser.net/datapages/?cohort=GDC%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443) 里的[HTSeq - FPKM-UQ (n=11,768) GDC Hub](https://gdc.xenahubs.net/download/GDC-PANCAN.htseq_fpkm-uq.tsv.gz)，[Gene Mapping](https://gdc.xenahubs.net/download/gencode.v22.annotation.gene.probeMap)

- TCGA_phenotype_denseDataOnlyDownload.tsv，样本注释。[TCGA Pan-Cancer (PANCAN) (41 datasets) phenotype](https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)里的[sample type and primary disease (n=12,804) Pan-Cancer Atlas Hub](https://pancanatlas.xenahubs.net/download/TCGA_phenotype_denseDataOnlyDownload.tsv.gz)，只有癌症全称，没有缩写。

- samplepair.txt，癌症缩写跟全称的对应关系

GSVA input data can be microarray data or RNA-seq counts, CPM, RPKM, TPM.

Original text: We calculate now GSVA enrichment scores for these gene sets using first the microarray data and then the **RNA-seq integer count** data. Note that the only requirement to do the latter is to set the argument **kcdf="Poisson"** which is "Gaussian" by default. Note, however, that if our RNA-seq derived expression levels would be **continuous, such as log-CPMs, log-RPKMs or log-TPMs**, then the default value of the **kcdf argument should remain unchanged**.

Source: <https://www.bioconductor.org/packages/release/bioc/vignettes/GSVA/inst/doc/GSVA.pdf>

- Expression matrix and Gene mapping:
  - To maintain consistency with FigureYa55panCancer_violin, download UCSC Toil RNA-seq Recompute TPM from [XENA](https://xenabrowser.net/datapages/): [TCGA Pan-Cancer (PANCAN) (41 datasets)
](https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443): [TOIL RSEM tpm (n=10,535) UCSC Toil RNA-seq Recompute](https://toil.xenahubs.net/download/tcga_rsem_isoform_tpm.gz)，[Gene Mapping](https://toil.xenahubs.net/download/probeMap/gencode.v23.annotation.transcript.probemap)
  - Or download GDC pipeline FPKM-UQ: [GDC Pan-Cancer (PANCAN) (17 datasets)](https://xenabrowser.net/datapages/?cohort=GDC%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443) : [HTSeq - FPKM-UQ (n=11,768) GDC Hub](https://gdc.xenahubs.net/download/GDC-PANCAN.htseq_fpkm-uq.tsv.gz)，[Gene Mapping](https://gdc.xenahubs.net/download/gencode.v22.annotation.gene.probeMap)

- TCGA_phenotype_denseDataOnlyDownload.tsv, sample annotation. [TCGA Pan-Cancer (PANCAN) (41 datasets) phenotype](https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443): [sample type and primary disease (n=12,804) Pan-Cancer Atlas Hub](https://pancanatlas.xenahubs.net/download/TCGA_phenotype_denseDataOnlyDownload.tsv.gz), contains only full cancer names without abbreviations.

- samplepair.txt, mapping between cancer abbreviations and full names
