---
title: "FigureYa88DAVIDkappa"
author: "Ying Ge, Taojun Ye"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# 设置knitr的全局代码块选项 / Set global chunk options for knitr
```

## 需求描述

富集分析得到太多term，其中好多还是相似的，怎样合理的合并？clusterProfiler有一个simplify函数，能给富集分析结果瘦身by removing redundancy of enriched GO terms，但有时瘦的不够多。

FigureYa80GOclustering用GOSemSim计算GO term之间的相似性，但是只能合并GO注释，对其他来源的注释无能为力。

想要合并来源于不同数据库的term，参考DAVID的方法。

## Requirement description

Enrichment analysis yields too many terms, many of which are still similar. How to merge them reasonably? ClusterProfiler has a simplify function that can thin down enriched GO terms by removing redundancy of enriched GO terms, but sometimes not enough weight is lost.

FigureYa80GOclustering uses GOSemSim to calculate the similarity between GO terms, but can only merge GO annotations and cannot handle annotations from other sources.

To merge terms from different databases, refer to DAVID's method.

![](example.png)

出自<https://www.nature.com/articles/nature22973>

from<https://www.nature.com/articles/nature22973>

**方法探讨：**

DAVID给gene或annotation做分类的原理：<https://david.ncifcrf.gov/helps/functional_classification.html>，包括kappa的计算和heuristic fuzzy partition algorithm。

H Ma等写了R代码来计算Kappa value：[Use of DAVID algorithms for clustering custom annotated gene lists in a non-model organism, rainbow trout](https://bmcresnotes.biomedcentral.com/articles/10.1186/s13104-018-3154-7)。作者在补充材料里提供了R代码：[Additional file 1. R_script for clustering.](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5781295/bin/13104_2018_3154_MOESM1_ESM.txt)，包含两部分，第一部分：计算Kappa value，第二部分：设置kappa的cutoff筛选。

基于这篇文章提供的代码里的第一部分来计算Kappa value。然后借助ggtree展示各个term之间的关系，并实现例文中相似term的标注效果。关于ggtree的问题，请前往Y叔知识星球biobabble讨论。

**Method exploration:**

The principle of DAVID classifying genes or annotations:< https://david.ncifcrf.gov/helps/functional_classification.html >Including kappa calculations and heuristic fuzzy partition algorithms.

H Ma and others wrote R code to calculate Kappa value：[Use of DAVID algorithms for clustering custom annotated gene lists in a non-model organism, rainbow trout]( https://bmcresnotes.biomedcentral.com/articles/10.1186/s13104-018-3154-7 ）The author provided R code in the supplementary materials: [Additional file 1. R_script for clustering.]（ https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5781295/bin/13104_2018_3154_MOESM1_ESM.txt ）It consists of two parts, the first part is to calculate the Kappa value, and the second part is to set the cutoff filter for Kappa.

Calculate Kappa value based on the first part of the code provided in this article. Then use ggtree to display the relationships between various terms and achieve the annotation effect of similar terms in the example text. Regarding the issue of ggtree, please go to Y Uncle's knowledge planet Biobabble for discussion.

## 应用场景

基于term里的基因间overlap来衡量term之间的相似性。

计算Kappa value，可用于找相似的注释、给相似的注释归类。

原理见：<https://david.ncifcrf.gov/helps/linear_search.html#kappa>

## Application scenarios

Measure the similarity between terms based on the overlap between genes in the term.

Calculate Kappa value, which can be used to find similar annotations and classify them.

The principle is as follows:< https://david.ncifcrf.gov/helps/linear_search.html#kappa >

## 环境设置


## Environment settings


```r

if (!requireNamespace("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
BiocManager::install("ggtree", version = "3.8")
```

加载包

library packages

```{r}
source("install_dependencies.R")
# 加载ape包，用于系统发育树分析和比较基因组学
# Load the ape package for phylogenetic tree analysis and comparative genomics
library(ape)

# 加载ggtree包，用于可视化和注释系统发育树
# Load the ggtree package for visualizing and annotating phylogenetic trees
library(ggtree)

# 加载scales包，用于ggplot2中的坐标轴缩放和格式化
# Load the scales package for axis scaling and formatting in ggplot2
library(scales)

# 加载ggplot2包，用于创建优雅的数据可视化
# Load the ggplot2 package for creating elegant data visualizations
library(ggplot2)

# 设置环境变量，使R显示英文错误信息
# Set environment variable to display error messages in English
Sys.setenv(LANGUAGE = "en")

# 禁止R自动将字符向量转换为因子类型
# Prevent R from automatically converting character vectors to factors
options(stringsAsFactors = FALSE)
```

## 输入文件

easy_input.csv，注释信息跟Gene ID的对应关系。至少包含两列：

- 第一列：GO、KEGG等无所谓什么来源的注释term；
- 第二列：Gene ID，此处是ENTREZ ID，也可以是gene symbol等无所谓什么来源的基因ID。
- 后面画热图时需要的logFC、Pvalue、zscore等信息也放在这里。如果要多组富集分析结果做对比，也都放到这里。多组合并的方法可参考FigureYa80GOclustering。

## Input file

easy_input.csv， The correspondence between annotation information and Gene ID. Include at least two columns:

-First column: Annotation terms such as GO, KEGG, etc. that do not matter what source they come from;
-Second column: Gene ID, here is ENTREZ ID, which can also be gene symbols or other gene IDs of no particular origin.
-The logFC, Pvalue, zscore and other information required for drawing heat maps later are also included here. If multiple sets of enrichment analysis results need to be compared, they should also be placed here. The method of merging multiple groups can refer to FigureYa80GOclustering.
