---
title: "FigureYa170ImmuLncRNA"
author: "Rongfang Shen"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(message = FALSE)
knitr::opts_chunk$set(warning = FALSE)
```

# 需求描述
# Requirement

能不能众筹这个算法的代码，希望在展示代码的时候能做一些讲解，讲的容易理解一些，原文写的有些看不懂。
Can you crowdfund the code for this algorithm? I'd appreciate some explanations when presenting the code to make it easier to understand, as the original paper is a bit difficult to understand.

![](example.png)

出自<https://www.nature.com/articles/s41467-020-14802-2>
From <https://www.nature.com/articles/s41467-020-14802-2>

Fig.1 Identification of immune-related lncRNAs across cancer types. d The number of immune-related lncRNAs identified in each cancer type. The top y-axis shows the number of lncRNAs and the bottom y-axis shows the proportion of lncRNAs. Source data are provided as a Source Data file.

利用偏相关系数和GSEA算法鉴定免疫相关的lncRNA，作者提供了R包，算法的主要精华其实已经给我们了。接下来我们就一步步解析，看文章里method部分如何对应算法R code。
The authors provide an R package for identifying immune-related lncRNAs using the partial correlation coefficient and the GSEA algorithm. The key points of the algorithm are already provided. Next, we will analyze it step by step to see how the method section in the article corresponds to the R code for the algorithm.

# 应用场景
# Application Scenarios

对lncRNA做功能注释，也可以用于甲基化数据等。
Functional annotation of lncRNAs can also be applied to methylation data, etc.

双坐标轴图通常用来对比展示数据之间的变化趋势相关关系：
Dual-axis charts are often used to compare and display trends and correlations between data:

- 散点图 + 线性回归，展示3组数据的相关性，A vs. B 和 A vs. C，可参考FigureYa62twoAxis。
- 柱形图 + 折线图，像本篇例文这样，一个是数量，另一个是比例。
- Scatter plot + linear regression, showing the correlation between three data sets, A vs. B and A vs. C. See FigureYa62twoAxis.
- Bar chart + line chart, like the example in this article, one showing quantity and the other showing proportion.

# 环境设置
# Environment Setup

下载R包[ImmuLncRNA](http://bio-bigdata.hrbmu.edu.cn/ImmLnc/download_loading.jsp?path=jt_download/ImmuLncRNA_0.1.0.tar.gz&name=ImmuLncRNA_0.1.0.tar.gz)后本地安装
Download the R package [ImmuLncRNA](http://bio-bigdata.hrbmu.edu.cn/ImmLnc/download_loading.jsp?path=jt_download/ImmuLncRNA_0.1.0.tar.gz&name=ImmuLncRNA_0.1.0.tar.gz) and install it locally.

```{r}
source("install_dependencies.R")
```

```{r, eval=FALSE}
install.packages("ImmuLncRNA_0.1.0.tar.gz", repos = NULL, type = "source")
install.packages("estimate")
```

加载包
Load package

```{r}
library(ComplexHeatmap)
library(TCGAbiolinks)
library(SummarizedExperiment)
library(dplyr)
library(tidyverse)
library(fgsea)
library(ggplot2)
library(ImmuLncRNA)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 #Display English error messages
options(stringsAsFactors = FALSE) #禁止chr转成factor #Disable conversion of chr values to factors
```

加载自定义函数，做标准化
Load custom functions for normalization

```{r}
Adjust_mrna <- function(mrna){
  mRow30 <- which(apply(mrna,1,function(v){return((sum(v==0)/length(v))>=0.3)}))
  mRemv <- mRow30
  if(length(mRemv)==0){
      mRNA_out0 <- mrna
    }else{
      mRNA_out0 <- mrna[-(mRemv),]
    }
    mRNA_exp_inter <- log2(mRNA_out0+0.001)
    return(mRNA_exp_inter)
}

Adjust_lncrna <- function(lncrna){
  lncRow50 <- which(apply(lncrna,1,quantile,probs=0.5)==0)
    lncRow90 <- which(apply(lncrna,1,quantile,probs=0.9)<=0.1)
    lncRemv <- union(lncRow50,lncRow90)
    if(length(lncRemv)==0){
      lncRNA_out0 <- lncrna
    }else{
      lncRNA_out0 <- lncrna[-(lncRemv),]
    }
    lncRNA_exp_inter <- log2(lncRNA_out0+0.001)
    return(lncRNA_exp_inter)
}
```

# 输入文件的准备
# Preparing the Input File

如果你只想画这种双坐标图，就把自己的数据整理成easy_input.csv的格式，直接跳到“开始画图”。
If you only want to draw this dual-coordinate graph, format your data as easy_input.csv and jump directly to "Start Drawing."

## 数据下载
## Data Download

中间文件已上传至微云<https://share.weiyun.com/5zZiFrs>
The intermediate file has been uploaded to Weiyun <https://share.weiyun.com/5zZiFrs>

以ACC为例，其余癌种同；下载数据后根据GENECODE的注释划分lncRNA表达矩阵和mRNA表达矩阵。数据处理为排除read数为0后进行Log转化。
Using ACC as an example, the same procedure applies to other cancer types. After downloading the data, divide the lncRNA expression matrix into an mRNA expression matrix based on GENECODE annotations. Data processing includes excluding reads with zero counts and performing log transformation.

数据处理方法按照文章所示：
The data processing method is as shown in the article:

RNA-Seqbased gene expression profile data were obtained from the TCGA project via the R package **“TCGAbiolinks”**56. We downloaded the **fragments per kilobase of transcript per million** mapped reads-based gene expression and the raw read count for 33 types of cancer. Based on the gene annotations in **GENCODE**, we divided the gene expression profiles into lncRNA and protein-coding gene expression for each cancer type. The lncRNAs were further classified into different subtypes based on the classification in GENCODE. In total, 19,663 coding genes and 15,513 lncRNAs were included. Next, we excluded the coding genes and lncRNAs with zero reads in all samples. The expression values of lncRNAs and protein-coding genes were logtransformed.

下载表达矩阵
Download expression matrix

```{r, eval=FALSE}
query <- GDCquery(project = "TCGA-ACC",
                  data.category = "Transcriptome Profiling",
                  data.type = "Gene Expression Quantification",
                  workflow.type = "HTSeq - FPKM", sample.type = c("Primary solid Tumor"))
GDCdownload(query, method = "api", files.per.chunk = 10)
acc <- GDCprepare(query)
#保存到文件，便于以后重复使用
#Save to file for future reuse
save(acc, file = "acc.rda")
```

## 提取protein coding RNA和lncRNA的gene_id和gene_name对应关系
## Extract protein coding Correspondence between gene_id and gene_name for RNA and lncRNA

根据[lncRNA的分类](https://asia.ensembl.org/info/genome/genebuild/biotypes.html)
lncRNA分为以下几种
According to the [lncRNA classification](https://asia.ensembl.org/info/genome/genebuild/biotypes.html)
lncRNAs are classified into the following categories:

![](lncRNAcategory.png)

lncRNA的注释文件可以从[genecode](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_22/gencode.v22.long_noncoding_RNAs.gtf.gz)下载，全部rna的基因注释可以从[这里](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_22/gencode.v22.annotation.gtf.gz)下载。
The lncRNA annotation file can be downloaded from [genecode](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_22/gencode.v22.long_noncoding_RNAs.gtf.gz). Gene annotations for all RNAs can be downloaded from [here](ftp://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_22/gencode.v22.annotation.gtf.gz).
