---
title: "FigureYa167TumorIndex"
author: "Rongfang Shen"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_chunk$set(warning = FALSE)
knitr::opts_chunk$set(message = FALSE)
```

# 需求描述：
作者通过不同肝癌发展模型的RNAseq和转录因子cluster计算出tumor index(`TI:-1~1`，正常~癌)。
# Requirement:
The author calculated the tumor index (TI: -1 to 1, normal to cancer) using RNAseq and transcription factor clusters from different liver cancer development models.

要求：①输入建模动物肝癌模型RNAseq数据算出TI②外部验证数据算出TI，重复出下图。
Requirements: ① Calculate the TI using RNAseq data from the animal liver cancer model. ② Calculate the TI using external validation data. Repeat the calculation to produce the following figure.

![](example.png)
出自<http://www.pnas.org/cgi/pmidlookup?view=long&pmid=31843886>
From <http://www.pnas.org/cgi/pmidlookup?view=long&pmid=31843886>

Fig.5 TI values of human HCC and precancer patients with steatosis, fibrosis, or NASH. (A) A total of 193 adjacent nontumor liver samples and 240 liver tumor samples were from the GSE36376 dataset. (B) A total of 10 nontumor liver samples, 17 dysplastic liver samples, 13 cirrhotic liver samples, and 35 HCC samples in GSE6764. (C) Cirrhotic liver (n = 34) and HCC (n = 35) samples in GSE56140. (D) Liver samples for control (n = 14), obese (n = 27), steatosis (n = 14), and NASH (n = 18) patients in GSE48452. (E) A total of 72 fibrotic liver samples (fibrosis stages 0–1, n = 40; fibrosis stages 3–4, n = 32) in GSE49541. (F) Hepatitis B virus-related liver fibrosis samples (n = 124) in GSE84044. The pathological Scheuer score of each sample was evaluated, score 0 (n = 43), score 1 (n = 20), score 2 (n = 33), and score 3 and 4 (n = 28). (G) A total of 371 liver tumor samples in the TCGA dataset. Of these, 347 tumor samples had well-documented tumor stages (stage I, n = 171; stage II, n = 86; stage III and more advanced, n = 90).

# 应用场景
Application Scenarios

1.利用此模型方法学训练自己多阶段/多点的RNAseq数据，或是应用到单细胞数据（如不同分化程度的T细胞）

2.验证数据：如癌和正常的区分度。

文章已经提供了[代码](https://github.com/wanyewang1/Index_model)，但是小伙伴说不会用。

我们这里重新包装了代码，提供比较简便的使用方法。

1. Use this model methodology to train your own multi-stage/multi-point RNAseq data, or apply it to single-cell data (e.g., T cells at different levels of differentiation).

2. Validate data: e.g., differentiation between cancer and normal cells.

The article already provides [code](https://github.com/wanyewang1/Index_model), but some users said they didn't know how to use it.

We've repackaged the code here to provide a simpler way to use it.

# 环境设置
# Environment Setup

加载包
Loading Packages

```{r}
source("install_dependencies.R")
library(GSA)
library(randomForest)
library(preprocessCore)
library(glmnet)
library(ggplot2)
library(tidyverse)
library(GEOquery)
library(ggpubr)
library(ggsci)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # Display English error messages
options(stringsAsFactors = FALSE) #禁止chr转成factor # Disable conversion of chr values to factors
```

# 训练模型
# Training Model

> 打开`model_training.R`文件查看详情，已添加必要的中文注释，可以用自己的数据来训练模型。
> Open the `model_training.R` file for details. Necessary Chinese annotations have been added. You can train the model using your own data.

这里概括一下训练模型的重要步骤：
Here is a summary of the key steps in model training:

- 1.WT adult liver and HCC的转录组数据用于训练模型。
- 2.鉴定到61个显著改变的TFcluster。
- 3.随机森林确定TFcluster里下游基因对结局变量的贡献度。
- 4.计算每个样本的TF activity.
- 5.计算每个样本的tumor-promoting and -inhibiting strengths.
- 6.广义线性模型拟合，计算index
- 1. Transcriptome data from WT adult liver and HCC were used to train the model.
- 2. 61 significantly altered TFclusters were identified.
- 3. Use a random forest to determine the contribution of downstream genes in the TFcluster to the outcome variable.
- 4. Calculate TF activity for each sample.
- 5. Calculate tumor-promoting and -inhibiting strengths for each sample.
- 6. Fit a generalized linear model and calculate the index
