---
title: "FigureYa166scCNV"
author: "Zongcheng Li"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Requirement Description

单细胞CNV的计算和画图。
Calculate and plot single-cell CNVs.

![](example1.png)

Figure 1. Characterizing Intra-tumoral Expression Heterogeneity in HNSCC by Single-Cell RNA-Seq. (B) Heatmap shows large-scale CNVs for individual cells (rows) from a representative tumor (MEEI5), inferred based on the average expression of 100 genes surrounding each chromosomal position (columns). Red: amplifications; blue: deletions.

出自<https://www.cell.com/cell/fulltext/S0092-8674(17)31270-9>
From <https://www.cell.com/cell/fulltext/S0092-8674(17)31270-9>

其他文章里类似的图：
Similar pictures from other articles:

![](example2.png)

Fig. 1. Dissection of melanoma with single-cell RNA-seq. (B) Chromosomal landscape of inferred large-scale CNVs allows us to distinguish malignant from nonmalignant cells. The Mel80 tumor is shown with individual cells (y axis) and chromosomal regions (x axis). Amplifications (red) or deletions (blue) were inferred by averaging expression over 100-gene stretches on the respective chromosomes.

出自<https://science.sciencemag.org/content/352/6282/189>

# 应用场景
# Application Scenarios

用单细胞RNA-seq数据计算CNV，对比展示多组之间的差异。
Calculate CNVs using single-cell RNA-seq data and compare differences between multiple groups.

# 环境设置
# Environment Setup


```{r}
source("install_dependencies.R")

```

加载包
Loading the package

```{r}
library(org.Hs.eg.db)
library(magrittr)
library(dplyr)
library(caTools)
library(pheatmap)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # Display English error messages
options(stringsAsFactors = FALSE) #禁止chr转成factor # Disable conversion of chr to factor
```

自定义函数
Custom Function

```{r}
minmax <- function(x, min, max ){
  x[x > max] <- max
  x[x < min] <- min
  return(x)
}
```

# 输入文件
# Input file

GSE103322_HNSCC_all_data.txt.gz，这里以原文的数据为例，其他单细胞RNA-seq数据与此类似，包含表达矩阵和样品分组等信息meta data。
GSE103322_HNSCC_all_data.txt.gz. This example uses the original data. Other single-cell RNA-seq datasets are similar and include metadata such as expression matrix and sample grouping.

Download GSE103322_HNSCC_all_data.txt.gz from <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103322>，已上传到微云，下载链接<https://share.weiyun.com/5gsuteR>
Download GSE103322_HNSCC_all_data.txt.gz from <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE103322>, uploaded to Weiyun, download link <https://share.weiyun.com/5gsuteR>
