---
title: "FigureYa122mut2expr"
author: "Long Zhao, Taojun Ye"
reviewer: "Ying Ge"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# 设置knitr代码块的全局选项 / Set global options for knitr code chunks
```

## 需求描述

虽然原文附件给出了代码，但自己看着费劲，请高手带我们理解并复现原文的Figure 1b建模。

## Requirement description

Although the original attachment provides the code, it seems difficult for us to understand and reproduce the Figure 1b modeling in the original text. Please help us understand and reproduce it.

![](example.png)

出自<http://www.nature.com/articles/ncomms6901>

from<http://www.nature.com/articles/ncomms6901>

代码在原文补充材料中有提供，里面有很多东西值得学习。同时我自己写了一些比原文精简的代码（一些数据命名尽量和原文一致，确保小伙伴跑的时候能够理解不同做法的异曲同工）。

**小丫碎碎念：**像这种附带代码的好文章越来越多，是非常好的学习资源。

这类文章怎么找到呢？小丫的经验是：在感兴趣的期刊主页的高级搜索里，搜code或script，同时限定research article和发表年份等信息。检索结果里会有上下文，一眼扫过去，就能看到哪些文章带代码。

怎样学习这类代码？这次众筹就从这篇带代码的文章带你上手，以后你就可以自己去挖宝藏了。推荐路线：

- 先尝试自己写，或者至少在脑子里猜测作者是用什么方法哪个包实现的
- 看原文代码，边看边实践
- 记笔记，思考这段代码能应用到哪些场景上，可以尝试按照FigureYa的格式来写，或者干脆写成R包（Y叔说写R包要趁早）
- 用在自己的数据上，争取能用到paper里，经历了这一过程，你会有质的提升

自己死磕一篇文章的代码是孤独的，遇到问题也没人讨论，以后有机会我们可以搞个学习小组，感兴趣的小伙伴一起讨论同一篇文章的代码。

The code is provided in the supplementary materials of the original text, and there are many things worth learning from it. At the same time, I wrote some code that was more concise than the original text (some data names should be as consistent as possible with the original text to ensure that my friends can understand the similarities of different approaches when running).

**Xiaoya's rambling: * * There are more and more good articles with accompanying code, which are very good learning resources.

How can I find such articles? Xiaoya's experience is to search for code or script on the advanced search page of the journal she is interested in, while limiting information such as research article and publication year. There will be context in the search results, and at a glance, you can see which articles have code.

How to learn this type of code? This crowdfunding campaign will take you through this article with code, so you can dig for treasure on your own in the future. Recommended route:

-Try writing it yourself first, or at least guess in your mind what method and package the author used to implement it
-Read the original code and practice while reading
-Take notes and think about which scenarios this code can be applied to. You can try writing it in the format of FigureYa, or simply write it as an R package (Uncle Y said it's better to write an R package as soon as possible)
-Apply it to your own data and strive to use it in a paper. Through this process, you will have a qualitative improvement

It's lonely to tinker with the code of an article by oneself, and no one will discuss it when encountering problems. In the future, we can form a study group where interested friends can discuss the code of the same article together.

## 应用场景

已有genotype和转录数据，通过线性模型解析突变对基因表达的贡献，即基因表达如何响应不同的genotype，不同的突变如何影响基因表达。

该模型的假设：

1. 每个突变位点只是影响特定的一些基因的表达；

2. 包含2种以上突变的个体，基因表达变化为两个或多个突变影响基因集合的并集；

缺点：该模型忽略了基因间的相互调控。

如果想要系统的研究基因之间的相互关系，就需要更多的基因样本容量，所以该线性模型方法是相对小样本量的很好的方式。

## Application scenarios

There is already genotype and transcription data available, and a linear model is used to analyze the contribution of mutations to gene expression, that is, how gene expression responds to different genotypes and how different mutations affect gene expression.

The assumption of this model is:

1. Each mutation site only affects the expression of specific genes;

2. Individuals with two or more mutations whose gene expression changes result in the union of gene sets affected by two or more mutations;

Disadvantage: This model ignores the mutual regulation between genes.

If you want to systematically study the interrelationships between genes, you need more gene sample sizes, so this linear model method is a good way for relatively small sample sizes.

## 环境设置


## Environment settings


```r

BiocManager::install("hgu133plus2.db")
```

加载包

library packages

```{r}
source("install_dependencies.R")
# 加载用于读取Excel文件的包
# Load the package for reading Excel files
library(openxlsx)
# 加载用于芯片数据处理的包
# Load the package for microarray data processing
library(affy)
# 加载用于芯片数据处理的包（GCRMA算法）
# Load the package for microarray data processing (GCRMA algorithm)
library(gcrma)
# 加载人类基因注释包
# Load the human gene annotation package
library(org.Hs.eg.db)
# 加载人类芯片(hgu133plus2)注释包
# Load the human microarray (hgu133plus2) annotation package
library(hgu133plus2.db)
# 加载线性模型分析包
# Load the package for linear model analysis
library(limma)
# 加载颜色方案包
# Load the package for color palettes
library(RColorBrewer)

# 设置环境变量，使R显示英文报错信息
# Set environment variable to display error messages in English
Sys.setenv(LANGUAGE = "en")
# 设置选项，禁止将字符串自动转换为因子类型
# Set option to prevent automatic conversion of strings to factors
options(stringsAsFactors = FALSE)
```

## 输入数据的预处理

需要两种输入数据：表达数据和基因型数据

表达数据：159个突变体和17个正常个体芯片数据，下载地址：< http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE58831>，点击Download里的http下载（早上下载速度更快），解压缩到当前文件夹。

genotype和表型数据：Supplementary Data 1—ncomms6901-s2.xlsx，其中部分个体数据缺失。

## Preprocessing of input data

Two types of input data are required: expression data and genotype data

Expression data: Chip data from 159 mutants and 17 normal individuals, download link:< http://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE58831 >Click on the HTTP download in Download (faster download speed in the morning), and extract it to the current folder.

Genotype and phenotype data: Supplementary Data 1- ncomms6901-s2.xlsx, with some individual data missing.

### 芯片数据处理：

以下是芯片数据处理过程，主要用到affy，gcrma包

下面这一步需要较长的运行时间，我把运行结果gset保存在文件里，便于再次运行时读取

### Chip data processing:

The following is the chip data processing process, mainly using the affy and gcrma packages

The following step requires a relatively long running time. I will save the running result gset in a file for easy retrieval when running again

```r
celFiles <- dir("GSE58831_RAW",full.names=T)  #读取文件夹下面文件
affyBatch <- read.affybatch(filenames = celFiles)  #Read CEL files into an Affybatch
gset <- gcrma(affyBatch)  #converts an AffyBatch into an ExpressionSet
save(gset, file = "gset.RData")
```
