---
title: "FigureYa185sciATAC"
author: "Long Zhao"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

# Requirement

用Cicero分析cis-regulation，用washu browser展示结果，搞定中间的衔接。
Analyze cis-regulation with Cicero, display the results with washu browser, and handle the intermediate connections.

![](example1.png)

Figure 3. Cicero Constructs cis-Regulatory Models Genome-wide from Sci-ATAC-Seq Data
(F) A summary of the Cicero co-accessibility links between the MYOG promoter and distal sites in the surrounding region. The height of connections indicates the magnitude of the Cicero co-accessibility score between the connected peaks. The top set of (red) links were constructed from cells in phase 1, while the bottom (in blue) were built from phase 2.
See also Figures S3 and S4.

![](example2.png)
Figure 5. Co-accessible DNA Elements Linked by Cicero Are Epigenetically Co-modified
(E) The Cicero map for the 755 kb region surrounding MYH3 along with called MYOD1 ChIP-seq peaks from (Cao et al., 2010). Sites opening in accessibility are colored by their opening pseudotime (see STAR Methods), sites not opening in accessibility are shown in gray. Inset: 60 kb region surrounding MYH3 along with MYOD1 ChIP-seq and H3K27ac ChIP-seq signal tracks from Cao et al. (2010) and the ENCODE Project Consortium (2012). Only protein-coding genes are shown.

出自<https://doi.org/10.1016/j.molcel.2018.06.044>
fromhttps://doi.org/10.1016/j.molcel.2018.06.044

# 应用场景

# Application Scenarios

用单细胞ATAC-seq数据找cis-regulation，为挖掘目的基因的远距离调控机制提供线索。
Search for cis-regulation using single-cell ATAC-seq data to provide clues for exploring the long-distance regulatory mechanism of the target gene.

还可参考这篇Cell paper的Figure 5和6深入挖掘，
You can also refer to Figures 5 and 6 of this Cell paper for in-depth exploration.

<https://www.sciencedirect.com/science/article/pii/S0092867418308559>，
需要的话，咱们再众筹。
If necessary, let's raise funds through crowdfunding.

# 环境设置

# Environment Setup

参考
Refer to
[Cicero for Monocle 3](https://cole-trapnell-lab.github.io/cicero-release/docs_m3/#loading-data-from-a-simple-sparse-matrix-format)

踩了软件安装的坑。软件安装有些麻烦，一些依赖不太好解决。所以这里直接用conda安装monocle3。
I fell into a trap in software installation. The software installation is a bit troublesome, and some dependencies are not easy to solve. So here, we directly use conda to install monocle3.

```bash
conda create -n py36 python=3.6
conda activate py36
#conda install -c r r-base=3.6.3
conda install -c bioconda r-monocle3
conda install -c bioconda bioconductor-cicero -y
```

cicero直接安装github版。不用conda安装，是因为conda的cicero不是最新版本，在处理10X cell ranger的数据没问题，但是在处理稀疏矩阵的时候有问题。
Install the github version of cicero directly. Not using conda for installation is because conda's cicero is not the latest version. It works fine with the data of 10X cell ranger, but has problems when dealing with sparse matrices.

```{r}
source("install_dependencies.R")
```

```{r, eval=FALSE}
BiocManager::install(c("Gviz", "GenomicRanges", "rtracklayer"))
#install.packages("devtools")
devtools::install_github("cole-trapnell-lab/cicero-release", ref = "monocle3")
#小丫用上面的命令安装不成功，改从github下载后，重新压缩，本地安装，成功
# Xiaoya failed to install it using the above command. Instead, she downloaded it from github, re-compressed it and installed it locally. It was successful
#install.packages("cicero-release-master.tar.gz", repos = NULL, type = "source", ref = "monocle3")

#MAC系统安装monocle时可能遇到问题，解决方法看这里https://cole-trapnell-lab.github.io/monocle3/docs/installation/
# MAC system installing a monocle may encounter problems, solution here at https://cole-trapnell-lab.github.io/monocle3/docs/installation/
devtools::install_github('cole-trapnell-lab/leidenbase')
devtools::install_github('cole-trapnell-lab/monocle3')
```

加载包
Loading package

```{r, eval=FALSE}
library(tidyverse)
library(cicero) #单细胞ATAC-seq数据分析  # Single-cell ATAC-seq Data analysis
library(monocle3) #单细胞  # Single cell

Sys.setenv(LANGUAGE = "en") #显示英文报错信息  # displays an English error message
options(stringsAsFactors = FALSE) #禁止chr转成factor   # prohibit chr from being converted to factor
```

# 输入文件

# Input File

需要单细胞ATAC-seq数据和基因组注释gtf文件。
Single-cell ATAC-seq data and genomic annotation gtf files are required.

这些文件已上传至微云
These files have been uploaded to Weiyun
<https://share.weiyun.com/gxepYhd9>

## 单细胞ATAC-seq数据

## Single-cell ATAC-seq data

从GEO下载单细胞ATAC-seq数据：<https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2970932>，以其中数据量小些的GSM2970932为例，包括两个细胞的mix，一个是HL60，另外一个是GM12878。
Download single-cell ATAC-seq data from GEO: < https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM2970932 >, in which the data volume small GSM2970932, for example, including the mix two cells, one is the HL60, another is GM12878.

![](download.png)

- GSM2970932_sciATAC_GM12878_HL60_counts.txt.gz，每个peak的reads数量
  GSM2970932_sciATAC_GM12878_HL60_counts.txt.gz, the number of reads per peak
- GSM2970932_sciATAC_GM12878_HL60_indextable.txt.gz，index

这是一种常见格式，好多文章都在GEO中提供这种格式的文件。感兴趣的小伙伴可以参考本文的方式挖掘一下自己感兴趣问题的相关数据。
This is a common format, and many articles provide files in this format in GEO. Interested friends can refer to the methods in this article to dig out the relevant data of the issues they are interested in.
