---
title: "FigureYa239ST_PDAC"
author: "Houshi Xu"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

丫姐好，我想众筹这篇单细胞空间转录组的Figure 2，作者已经把原始数据都提供了。

# Requirement Description

Hello Xiaoya, I want to crowdfund Figure 2 of this single-cell spatial transcriptome, and the author has provided all the original data.

![](example.png)

出自<https://www.nature.com/articles/s41587-019-0392-8>
from<https://www.nature.com/articles/s41587-019-0392-8>

Fig. 2 | St of PDAC and mapping of cell types.
a, Annotated PDAC-A tumor cryosection on the ST slide. The annotations indicate a region high in cancer cells and desmoplasia (red), duct epithelium (yellow) and normal pancreatic tissue (blue). Scale bar, 1 mm.
b, Annotated PDAC-B tumor cryosection on the ST slide. Annotated regions include a cancer cell-rich region (red), duct epithelium (yellow) and interstitium (green). Scale bar, 1 mm.
c,d, Standardized expression levels of three genes in the PDAC-A ST (c) and PDAC-B ST (d) datasets.
e,f, Clustering of the PDAC-A (e) and PDAC-B (f) ST spots. Color indicates the clustering assignments.
g, MIA. The hypergeometric distribution is used to infer the significance of the intersection of genes specifically expressed in a given cell type (fibroblasts in this case) and genes specifically expressed in a given tissue region (cancer region). Applying this analysis systematically for all pairs of cell types and tissue regions allows for insight into the spatial distribution of the cell types in the tumor. The numbers of genes used in the calculation are shown.
h, The PDAC-A **MIA map** of all scRNA-seq-identified cell types and ST-defined regions. Each element in the matrix is computed as described in g for all pairs of cell types and tissue regions using the same 19,738 background genes. The numbers of cell type- and tissue region-specific genes used in the calculation are shown. Red indicates enrichment (significantly high overlap); blue indicates depletion (significantly low overlap). The bar on top indicates the regions delineated in e.
i, The PDAC-B MIA map for the ST clusters identified in f. Matrix elements were computed as described in g and h. The bar on top indicates the regions delineated in f.

# 应用场景

单细胞测序和空间转录组数据各有优缺点，作者开发的MIA (multimodal intersection analysis)算法整合了单细胞数据以及空间转录组数据，解决了ST技术无法达到单细胞分辨率、scRNA-seq无法进行空间定位的问题，从而实现了优势互补。（出自这篇中文解读<https://mp.weixin.qq.com/s/52xYNJ-xprLLQ-ZfiB4fig>）

因为原文没有指出一些重要的参数，最后的图和原文有出入，但是算法一定是准确的，可以用到自己的数据中。

# Application Scenarios

Single-cell sequencing and spatial transcriptome data have their own advantages and disadvantages, and the MIA (multimodal intersection analysis) algorithm developed by the authors integrates single-cell data and spatial transcriptome data to solve the problem that ST technology cannot achieve single-cell resolution and scRNA-seq cannot be spatially localized, thus achieving complementary advantages. (From this Chinese interpretation <https://mp.weixin.qq.com/s/52xYNJ-xprLLQ-ZfiB4fig>)

Because the original text does not point out some important parameters, the final graph is different from the original text, but the algorithm must be accurate and can be used in its own data.

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")

```

加载包
load packages

```{r}
library(data.table)
library(Seurat)
library(tidyverse)
library(ggplot2)
library(tidyHeatmap)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入文件的下载

下载原文的数据<https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111672>，解压缩GSE111672_RAW.rar

# Download the input file

Download the data <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE111672> in the original text, unzip the GSE111672_RAW.rar

![](download.png)

# PDAC-A

## Figure 2c - 空间转录组

c,d, Standardized expression levels of three genes in the **PDAC-A ST (c)** and PDAC-B ST (d) datasets.

结论：In both sample datasets, we found that the spatial expression of many variably expressed genes (see Methods) matched the annotated histological regions.

这里需要注意，方法中提到 UMI counts in each spot were normalized by the total transcript count and then scaled by the median number transcript count across all spots. A pseudocount of 1 was added before log10 transformation.

# PDAC-A

## Figure 2c - Spatial transcriptome

c,d, Standardized expression levels of three genes in the **PDAC-A ST (c)** and PDAC-B ST (d) datasets.

Conclusion: In both sample datasets, we found that the spatial expression of many variably expressed genes (see Methods) matched the annotated histological regions.

It is important to note here that the method mentions UMI counts in each spot were normalized by the total transcript count and then scaled by the median number transcript count across all spots. A pseudocount of 1 was added before log10 transformation.
