---
title: "FigureYa224scMarker"
author: "Yangding Zheng"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

这篇NC单细胞展示比例和marker基因的图很好看啊，可以众筹下做这几个图吗？

# Requirement Description

This diagram of NC single-cell display ratio and marker gene is very good, can you make these figures under crowdfunding?

![](example.png)

出自<https://www.nature.com/articles/s41467-020-16164-1>
from<https://www.nature.com/articles/s41467-020-16164-1>

Fig. 3 Tumor endothelial cells and myofibroblasts promoting angiogenesis and tissue remodeling.
a tSNE plot of endothelial cells, color-coded by clusters and cell subsets as indicated. EPCs: endothelial progenitor cells.
b Three-layered complex heatmap of **selected endothelial cell marker genes** in each cell cluster.
Top: Mean expression of known lineage markers.
Middle: Tissue preference of each cluster;
Bottom: Relative expression map of known marker genes associated with each cell subset.
Mean expression values are scaled by mean-centering, and transformed to a scale from -2 to 2.
c **Average cell number and relative proportion** of EC subsets from tissues of each origin. nLung, n = 11 samples; tLung, n = 11 samples; tL/B, n = 3 samples; mLN, n = 2 samples; mBrain, n = 9 samples.

# 应用场景

单细胞数据展示细胞比例和marker基因。

# Application Scenarios

Single-cell data showing cell ratios and marker genes.

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")

```

加载包
load packages

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(reshape2)
library(RColorBrewer)
library(scales)
library(ggplot2)
library(ggpubr)
library(ggplotify)
library(pheatmap)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入文件及预处理

以Seurat官网PBMC数据为例，跟FigureYa111markerGene和FigureYa160scGSVA用的是同一套数据，出自不同作者之手，可相互参考。

下载<https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz>，解压到当前文件夹。

预处理过程参考<https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html>

# Input files and preprocessing

Taking the PBMC data on the official website of Seurat as an example, it uses the same set of data as FigureYa111markerGene and FigureYa160scGSVA, which are from different authors and can be cross-referenced.

Download the <https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz> and extract it to the current folder.

Refer to <https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html> for the pretreatment process

```bash
tar -zxvf pbmc3k_filtered_gene_bc_matrices.tar.gz
```

```{r}
plot_wd <- "plot_result/"
dir.create(plot_wd)

# Load the PBMC dataset
pbmc.data <- Read10X(data.dir = "filtered_gene_bc_matrices/hg19/")
# Initialize the Seurat object with the raw (non-normalized data).
pbmc <- CreateSeuratObject(counts = pbmc.data, project = "pbmc3k", min.cells = 3, min.features = 200)
pbmc
```
