---
title: "FigureYa210survivalScape"
author: "Xiaofan Lu"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

用survdiff复现文章里的图，把所有基因的HR、pvalue输出到文件。

# Requirement Description

Use survdiff to reproduce the graph in the article, and output the HR and pvalue of all genes to a file.

![](example.png)

出自<https://molecular-cancer.biomedcentral.com/articles/10.1186/s12943-019-1066-3>
from<https://molecular-cancer.biomedcentral.com/articles/10.1186/s12943-019-1066-3>

Fig. 3 Clinical relevance of m6A regulators across cancer types. a Summary of the correlation between expression of m6A regulators and patient survival. **Red** represents a higher expression of m6A regulator associated with **worse survival**, and **blue** represents an association with **better survival**. Only p values < 0.05 are shown.

基于这一结果，下一步的做法：

- 例文分析每个m6A基因在pancancer和GEO多个数据集中的hazard ratios，画出森林图Fig. 3b The distribution of hazard ratios across different cancer types. 可参考FigureYa66单因素cox。
- 筛选获得显著影响生存的基因/癌症，然后批量画生存曲线，可参考FigureYa35batch_bestSeparation

我们还复现过这篇文章的Fig. 2A，用`FigureYa199crosslink`带你实现连线自由。

Based on these results, the next steps are as follows:

-Analyze each m6A gene by calculating hazard ratios across pan-cancer and multiple GEO datasets, then generate a forest plot (Fig. 3b: The distribution of hazard ratios across different cancer types). Refer to FigureYa66 single Cox for implementation.

-Screen for genes/cancers with significant survival impact, then batch-plot survival curves. Refer to FigureYa35batch_bestSeparation for guidance.

We have also reproduced Fig. 2A from this paper—use FigureYa199crosslink to achieve flexible linkage visualization.

# 应用场景

场景一：像例文那样，用TCGA-pancancer表达数据计算某个基因集的survival landscape。

场景二：批量筛选。或许你可以计算基因组上所有基因的survival landscape，筛选出genes associated with the overall survival of patients (worse/better survival) in at least one cancer type或你关心的几种癌症或所有癌症。

这里提供cox和survdiff（logrank，中位数区分高低组）两种方式：

- 方法一：（推荐）用cox来说明protective和risky是比较合理的做法。
- 方法二：（需求者提出的）survdiff的logrank，对于输入FPKM和TPM都可以用。

结果的可视化。例文用红蓝白表示风险、保护和不显著。或许你想用颜色的深浅来展示具体的数值，或者用形状和颜色同时展示HR和pvalue，可参考的FigureYa97correlation画法。

# Application Scenarios

Scenario 1: Calculate the survival landscape of a gene set using TCGA-pancancer expression data as in the example.

Scenario 2: Batch filtering. Perhaps you can calculate the survival landscape of all genes on the genome and screen out genes associated with the overall survival of patients (worse/better survival) in at least one cancer type or a few or all cancers that you care about.

There are two ways to use cox and survdiff (logrank, median distinguishes between high and low groups):

- Method 1: (Recommended) It is reasonable to use COX to illustrate that Protective and Risky are used.
- Method 2: The logrank of survdiff (proposed by the demander) can be used for both FPKM and TPM.

Visualization of results. Examples are written in red, blue, and white for risk, protection, and non-significant. Maybe you want to use the shade of color to show specific values, or use shapes and colors to show HR and pvalue at the same time, you can refer to FigureYa97correlation.

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")

```

加载包
load package

```{r}
library(data.table)
library(survival)
library(pheatmap)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入文件的获得

需要表达矩阵和生存数据，如果你已经准备好easy_input_gene.txt、easy_input_expr.csv和easy_input_surv.csv文件，就可以跳过这步，直接进入“输入文件”。

- easy_input_gene.txt，例文中的m6A基因，用到`Symbol`列，将提取这些基因的表达矩阵。

- tcga_RSEM_gene_tpm和gencode.v23.annotation.gene.probemap，表达矩阵和Gene mapping，跟FigureYa208FPI的一样，微云链接：<https://share.weiyun.com/c8GQyxR4>。**建议自己动手，从xena网站下载**，下载方式：

  - 这里跟FigureYa55panCancer_violin保持一致，从[XENA](https://xenabrowser.net/datapages/)下载UCSC Toil RNA-seq Recompute TPM：[TCGA Pan-Cancer (PANCAN) (41 datasets)](https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)里的[TOIL RSEM tpm (n=10,535) UCSC Toil RNA-seq Recompute](https://toil.xenahubs.net/download/tcga_rsem_isoform_tpm.gz)，[Gene Mapping](https://toil.xenahubs.net/download/probeMap/gencode.v23.annotation.transcript.probemap)

  - 或者下载GDC pipeline的FPKM-UQ：[GDC Pan-Cancer (PANCAN) (17 datasets)](https://xenabrowser.net/datapages/?cohort=GDC%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443) 里的[HTSeq - FPKM-UQ (n=11,768) GDC Hub](https://gdc.xenahubs.net/download/GDC-PANCAN.htseq_fpkm-uq.tsv.gz)，[Gene Mapping](https://gdc.xenahubs.net/download/gencode.v22.annotation.gene.probeMap)

- pancancerSurvivalData.txt，样本注释及生存信息，从[XENA](https://xenabrowser.net/datapages/)下载后整理而成。

  - 生存信息的下载方式：[Curated clinical data (n=12,591)](https://xenabrowser.net/datapages/?dataset=Survival_SupplementalTable_S1_20171025_xena_sp&host=https%3A%2F%2Fpancanatlas.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)里的[Curated clinical data](https://pancanatlas.xenahubs.net/download/Survival_SupplementalTable_S1_20171025_xena_sp.gz)

  - 样本信息的下载方式：[TCGA Pan-Cancer (PANCAN) (41 datasets) phenotype](https://xenabrowser.net/datapages/?cohort=TCGA%20Pan-Cancer%20(PANCAN)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443)里的[sample type and primary disease (n=12,804) Pan-Cancer Atlas Hub](https://pancanatlas.xenahubs.net/download/TCGA_phenotype_denseDataOnlyDownload.tsv.gz)。

# Acquisition of input files

Matrix and survival data need to be expressed, and if you already have easy_input_gene.txt, easy_input_expr.csv, and easy_input_surv.csv files ready, you can skip this step and go straight to the "Input File".

- easy_input_gene.txt, the m6A gene in the example is used in the 'Symbol' column to extract the expression matrix of these genes.

- tcga_RSEM_gene_tpm and gencode.v23.annotation.gene.probemap, expression matrix and gene mapping, the same as FigureYa208FPI, microcloud link: <https://share.weiyun.com/c8GQyxR4>。 **It is recommended to do it yourself, download it from the Xena website**, download method:

- Consistent with FigureYa55panCancer_violin here, download UCSC Toil RNA-seq Recompute TPM from [XENA](https://xenabrowser.net/datapages/): [TCGA Pan-Cancer (PANCAN) (41 datasets)]( https://xenabrowser.net/datapages/?cohort=TCGA [TOIL RSEM tpm (n=10,535) UCSC Toil RNA-seq] in Pan-Cancer (PANCAN)&removeHub=https://xena.treehouse.gi.ucsc.edu:443). Recompute](https://toil.xenahubs.net/download/tcga_rsem_isoform_tpm.gz)，[Gene Mapping]( https://toil.xenahubs.net/download/probeMap/gencode.v23.annotation.transcript.probemap)

- Or download the FPKM-UQ of GDC pipeline:[gdc Pan-Cancer (PANCAN) (17 datasets)](https://xenabrowser.net/datapages/?cohort=GDC Pan-Cancer (PANCAN)&removeHub=https:// [HTSeq - FPKM-UQ (n=11,768) GDC Hub](https://gdc.xenahubs.net/download/GDC-PANCAN.htseq_fpkm-uq.tsv.gz) in xena.treehouse.gi.ucsc.edu:443), [Gene Mapping](https://gdc.xenahubs.net/download/gencode.v22.annotation.gene.probeMap)

- pancancerSurvivalData.txt, sample annotations and survival information, downloaded from [XENA](https://xenabrowser.net/datapages/)

- Survival information is available as a download method: [Curated clinical data (n=12, 591)]( https://xenabrowser.net/datapages/?dataset=Survival_SupplementalTable_S1_20171025_xena_sp&host=https://pancanatlas.xenahubs.net&removeHub=https://xena.treehouse.gi.ucsc.edu:443 [Curated clinical data](https://pancanatlas.xenahubs.net/download/Survival_SupplementalTable_S1_20171025_xena_sp.gz)

- How to download the sample information: [TCGA Pan-Cancer (PANCAN) (41 datasets) phenotype](https://xenabrowser.net/datapages/?cohort=TCGA Pan-Cancer (PANCAN)&removeHub=https://xena.treehouse.gi.ucsc.edu [sample type and primary disease (n=12,804) Pan-Cancer Atlas Hub](https://pancanatlas.xenahubs.net/download/TCGA_phenotype_denseDataOnlyDownload.tsv.gz)。
