---
title: "FigureYa296VIPER"
params:
  author: "Xiaofan Lu; Yasi Zhang"
  reviewer: "Ying Ge"
  date: "2025-05-20"
output: html_document
---

**Author(s)**: `r params$author`
**Reviewer(s)**: `r params$reviewer`
**Date**: `r params$date`

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Demand description

转录因子活性计算方法。

**转录因子活性推断**

每个样本的转录因子活性通过VIPER软件包（Alvarez et al., 2016）基于z-score标准化的RNA数据进行计算。转录因子的靶基因来源于DoRothEA数据库（Garcia-Alonso et al., 2019），分析时采用中等置信度的靶基因集合。肿瘤样本与正常样本的活性评分通过Student t检验进行比较，并使用Benjamini-Hochberg方法校正p值。校正后p值 < 0.05的转录因子被视为具有显著差异。

我们还利用VIPER中的ARACNe算法（Lachmann et al., 2016）构建基因调控网络，并根据转录因子蛋白丰度的相关性推断其靶基因，从而解析癌症特异的转录因子调控机制。此外，我们将标准化富集的蛋白活性评分与免疫评分进行了相关性分析。

出自：<https://linkinghub.elsevier.com/retrieve/pii/S1535610820306553>

Calculation methods for transcription factor activity.

**Transcription factor activity inference**

Transcription factor activity for each sample was inferred using the VIPER package (Alvarez et al., 2016) on z-score transformed RNA data. The transcription factor targets were collected from DoRothEA (Garcia-Alonso et al., 2019) and the medium confidence targets were used for analysis. Activity scores for tumor and normal samples were compared using Student’s t-test and the p values were adjusted using the Benjamini-Hochberg method. Transcription factors with an adjusted p value < 0.05 were considered significant.

We also used the ARACNe algorithm (Lachmann et al., 2016) in VIPER to construct gene regulatory networks and infer transcription factor targets based on correlation to the transcription factor protein abundance. This allows for cancer-specific transcription factor gene regulation. We correlated the normalized enrichment protein activity scores with immune scores.

Source: <https://linkinghub.elsevier.com/retrieve/pii/S1535610820306553>

# 应用场景
# Application scenarios

将转录组表达矩阵转换为调控因子活性矩阵，并画图。

这里用VIPER包计算转录因子活性（Protein-activity by Enriched Regulon analysis）。另外，如果要分析**转录调控网络regulon的活性**（注意：并非转录因子活性），可参考FigureYa249Regulon。

例文METHOD DETAILS的Integrated analysis部分描述了挖掘生物学意义时用到的整合分析方法，我们曾众筹过其中一些算法：

- ESTIMATE：可参考FigureYa211multiCohortImmSubtype
- ssGSEA：可参考FigureYa71ssGSEA
- NMF：可参考FigureYa110mutationSignature，FigureYa158MutationPattern，FigureYa169sigHeatmap。

Convert the transcriptomic expression matrix into a regulator activity matrix and visualize the results.

Here, the VIPER package is used to calculate transcription factor activity (Protein-activity by Enriched Regulon analysis). Alternatively, if **analyzing regulon activity** (note: not transcription factor activity), refer to FigureYa249Regulon.

The METHOD DETAILS section under Integrated analysis describes the computational methods used for biological interpretation, some of which were previously crowdsourced in our studies:

- ESTIMATE: Refer to FigureYa211multiCohortImmSubtype
- ssGSEA: Refer to FigureYa71ssGSEA
- NMF: Refer to FigureYa110mutationSignature, FigureYa158MutationPattern, and FigureYa169sigHeatmap

# 环境设置
# Environment Setup


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

BiocManager::install("dorothea")
BiocManager::install("mixtools")
BiocManager::install("viper")
```

# 加载包
# Loading packages

```{r}
library(mixtools)
library(dorothea)
library(viper)

# 显示英文报错信息
# Show English error messages
Sys.setenv(LANGUAGE = "en")

# 禁止chr转成factor
# Prevent character-to-factor conversion
options(stringsAsFactors = FALSE)
```

# 输入文件
# Input Files

viper/ARACNe需要TXT格式的转录因子列表和表达矩阵作为输入数据，输入文件保存在"InputData"文件夹中。

- tfs.txt，转录因子列表，这里是从dorothea包中提取的。
- HS_CPTAC_HNSCC_RNAseq_RSEM_UQ_log2_Normal.cct，HS_CPTAC_HNSCC_RNAseq_RSEM_UQ_log2_Tumor.cct，表达矩阵，下载自<http://linkedomics.org/data_download/CPTAC-HNSCC>。整理成easy_input_matrix.txt。

文件较大，请到百度网盘下载<https://pan.baidu.com/s/1czOsuUfywIZD_h11TY0zHQ?pwd=rai9>，如图保存到相应文件夹内：

VIPER/ARACNe requires input data in TXT format, including a transcription factor list and an expression matrix, stored in the "InputData" folder.

- tfs.txt: Transcription factor list, extracted from the DoRothEA package.
- HS_CPTAC_HNSCC_RNAseq_RSEM_UQ_log2_Normal.cct，HS_CPTAC_HNSCC_RNAseq_RSEM_UQ_log2_Tumor.cct: Expression matrices downloaded from <http://linkedomics.org/data_download/CPTAC-HNSCC> and reformatted into easy_input_matrix.txt.

Due to large file sizes, please download them from BaiduNetdisk<https://pan.baidu.com/s/1czOsuUfywIZD_h11TY0zHQ?pwd=rai9> and save them in the corresponding folder as shown in the figure:

![](content.png)

```{r}
work.path <- "."
data.path <- file.path(work.path, "InputData")
res.path <- file.path(work.path, "Results")
code.path <- file.path(work.path, "Codes")
fig.path <- file.path(work.path, "Figures")

if (!dir.exists(data.path)) dir.create(data.path)
if (!dir.exists(res.path)) dir.create(res.path)
if (!dir.exists(code.path)) dir.create(code.path)
if (!dir.exists(fig.path)) dir.create(fig.path)

# 从dorothea包中提取转录因子列表
# Extract transcription factor list from dorothea package
net <- dorothea::dorothea_hs
tfs <- unique(net$tf)
write.table(tfs, file.path(data.path, "easy_input_tfs.txt"), row.names = F, col.names = F, quote = F)

# 读取原文提供的转录组数据，并保存到InputData/matrix.txt中
# Read original transcriptomic data and save to InputData/matrix.txt
tmp1 <- read.table(file.path(data.path, "HS_CPTAC_HNSCC_RNAseq_RSEM_UQ_log2_Normal.cct"),
                   header = T, row.names = 1, check.names = F)
colnames(tmp1) <- paste0("Normal", "_", colnames(tmp1))

tmp2 <- read.table(file.path(data.path, "HS_CPTAC_HNSCC_RNAseq_RSEM_UQ_log2_Tumor.cct"),
                   header = T, row.names = 1, check.names = F)
colnames(tmp2) <- paste0("Tumor", "_", colnames(tmp2))
expr <- cbind(tmp1, tmp2)

## 去除在过多样本(>10%)中表达值为0的基因，以避免在bootstrap过程中出现报错
## Remove genes with zero expression in >10% samples to avoid bootstrap errors
expr <- expr[rowSums(expr>0) >= 0.1*ncol(expr), ]

## 整理样本分组信息
## Prepare sample grouping information
clin <- data.frame(
  row.names = colnames(expr),
  "IsTumor" = c(rep("Normal", ncol(tmp1)), rep("Tumor", ncol(tmp2))),
  "Patient" = gsub("Tumor_|Normal_", "", colnames(expr))
)
# 也可以直接读取样本分组信息文件：
# Alternatively read sample grouping info directly from file:
# clin <- read.delim(file.path(data.path, "HS_CPTAC_HNSCC_CLI.tsi"))
# clin <- clin[-1, ]; rownames(clin) <- clin$case_id
# expr <- expr[, c("Idx", clin$case_id)]

write.table(data.frame("gene" = rownames(expr), expr), file.path(data.path, "easy_input_matrix.txt"),
            row.names = F, col.names = T, quote = F, sep = "\t")
```

# 重建基因调控网络 - ARACNe-AP
# Reconstructing Gene Regulatory Networks - ARACNe-AP

ARACNe-AP是基于java的程序，需要安装1.8以上的java。java下载地址<https://www.java.com/>

ARACNe-AP可以从GitHub上下载后自行编译，存放在`Code/ARACNe-AP`文件夹里。具体过程参考GitHub指南。git clone <https://github.com/califano-lab/ARACNe-AP.git>，或者使用已经编译好的可执行文件：dist/aracne.jar。

ARACNe-AP is a Java-based program that requires Java 1.8 or higher to be installed. Java can be downloaded from: <https://www.java.com/>

ARACNe-AP can be downloaded from GitHub and compiled locally, then stored in the `Code/ARACNe-AP` directory. For detailed instructions, please refer to the GitHub guide: git clone <https://github.com/califano-lab/ARACNe-AP.git>. Alternatively, you may directly use the pre-compiled executable located at: dist/aracne.jar.
