---
title: "FigureYa307CNVHeatmap"
params:
  author: "Xiaofan Lu; Yasi Zhang"
  reviewer: "Ying Ge"
  date: "2025-05-20"
output: html_document
---

**Author(s)**: `r params$author`
**Reviewer(s)**: `r params$reviewer`
**Date**: `r params$date`

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Demand description

希望画出拷贝数变异的图。

I want to plot copy number variation (CNV) profiles.​

![](example.png)

出自：<https://www.nature.com/articles/s41467-021-22465-w>

图2：非肌层浸润性膀胱癌(NMIBC)的拷贝数变异情况。
a. 按基因组类别(GC 1-3)分层的473个肿瘤的全基因组拷贝数景观图。增益(增益+高平衡增益)和缺失(缺失+高平衡缺失)汇总显示在染色体条带面板左侧。EORTC - 欧洲癌症研究与治疗组织，EAU - 欧洲泌尿外科协会，MIBC - 肌层浸润性膀胱癌

Suorce: <https://www.nature.com/articles/s41467-021-22465-w>

Fig. 2: Copy number alterations in NMIBC.
a. Genome-wide copy number landscape of 473 tumors stratified by genomic class (GC) 1–3. Gains (gain + high balanced gain) and losses (loss + high balanced loss) are summarized to the left of the chromosome band panel. EORTC European Organisation for Research and Treatment of Cancer, EAU European Association of Urology, MIBC muscle-invasive bladder cancer.

# 应用场景
# Application scenarios

利用GISTIC算法的输出文件，绘制拷贝数热图（不包含底部饼图）。

我们众筹过一些突变相关的图 <https://k.youshop10.com/gZBI0YzR>以及富集分析的图 <https://k.youshop10.com/sm2p0xYn>，可灵活运用，画出文中的图：

- Figure 1ah共识聚类及热图，可参考FigureYa202consensusGene <https://k.youshop10.com/dNHdlWen>
- Figure 1b桑基图，画法可参考FigureYa25sankey <https://k.youshop10.com/1vm3In-4>
- Figure 1cd、2bcd、3c、4e、5c、多亚型生存分析，可参考FigureYa284pairwiseLogrank <https://k.youshop10.com/=nHJB0CE>或FigureYa1survivalCurve <https://k.youshop10.com/3GKrTHDG>
- Figure 2a，本文档将带你实现。只画左侧频率对比图，可参考FigureYa79CNV <https://k.youshop10.com/7N-Flfly>
- Figure 3e突变综合热图，可参考FigureYa248MutLandscape <https://k.youshop10.com/hm=bnmr6>
- Figure 5a森林图，可参考FigureYa90subgroup <https://k.youshop10.com/HD4mSqRP>

Copy Number Heatmap Generation Using GISTIC Output Files (Excluding Bottom Pie Chart).

We have previously crowdsourced several mutation-related plots <https://k.youshop10.com/gZBI0YzR> and enrichment analysis plots <https://k.youshop10.com/sm2p0xYn>, which can be flexibly adapted to reproduce the figures in this study:

- Figure 1a&h: Consensus clustering and heatmap (refer to FigureYa202consensusGene <https://k.youshop10.com/dNHdlWen>
- Figure 1b: Sankey diagram (refer to FigureYa25sankey <https://k.youshop10.com/1vm3In-4>
- Figures 1c-d, 2b-d, 3c, 4e, 5c: Multi-subtype survival analysis (refer to FigureYa284pairwiseLogrank <https://k.youshop10.com/=nHJB0CE> or FigureYa1survivalCurve <https://k.youshop10.com/3GKrTHDG>
- Figure 2a: This document will guide you through its implementation. For the left-side frequency comparison plot only, refer to FigureYa79CNV <https://k.youshop10.com/7N-Flfly>
- Figure 3e: Integrated mutation heatmap (refer to FigureYa248MutLandscape <https://k.youshop10.com/hm=bnmr6>)
- Figure 5a: Forest plot (refer to FigureYa90subgroup  <https://k.youshop10.com/HD4mSqRP>)

# 环境设置
# Environment Setup


```{r}
source("install_dependencies.R")

BiocManager::install("trackViewer")
```

# 加载包
# Loading packages

```{r}
library(ComplexHeatmap)

# 如果是用hg38的基因组请加载:
# Load if using hg38 genome reference:
#library(TxDb.Hsapiens.UCSC.hg38.knownGene)

# 如果是用hg38的基因组请加载:
# Load if using hg38 genome reference:
library(TxDb.Hsapiens.UCSC.hg19.knownGene)

# 用于获取基因位点信息
# For gene locus information
library(trackViewer)

library(org.Hs.eg.db)
library(ggplot2)

# 用于获取TCGA数据的分子亚型信息
# For TCGA molecular subtype data
library(TCGAbiolinks)

library(dplyr)
library(RColorBrewer)
library(circlize)

# 显示英文报错信息
# Show English error messages
Sys.setenv(LANGUAGE = "en")

# 禁止chr转成factor
# Prevent character-to-factor conversion
options(stringsAsFactors = FALSE)
```

自定义函数，用于获取基因位点

Custom function for retrieving gene loci

```{r}
getGeneLocation <- function(symbol, txdb, org){
  # 获取基因位置
  # trackViewer::getLocation的改进版，避免报错，增加GeneSymbol列
  # Get gene locations
  # Improved version of trackViewer::getLocation with error handling and added GeneSymbol column
  dbPrefix <- sub(".db", "", org)
  eg <- mget(symbol, get(paste0(dbPrefix, "SYMBOL2EG")), ifnotfound = NA)
  eg <- lapply(eg, `[`, i = 1)
  eg <- unlist(eg, use.names = TRUE)
  eg <- eg[!is.na(eg)]
  if (length(eg) < 1) {
    return(NULL)
  }
  genes <- genes(txdb)
  eg <- eg[eg%in%genes$gene_id]
  gr <- genes[eg]
  gr$symbol <- names(eg)
  seqlevelsStyle(gr) <- "UCSC"
  seqlevels(gr) <- seqlevels(gr)[seqlevels(gr) %in% as.character(seqnames(gr))]
  if (length(seqinfo(gr)) > 0) {
    seqinfo(gr) <- seqinfo(gr)[seqlevels(gr)]
  }
  return(gr)
}
```

# 输入文件
# Input files

这里以TCGA数据为例。TCGA_BLCA.all_thresholded.by_genes.txt，该文件为GISTIC算法的输出文件之一，也可在<http://firebrowse.org/>中获取已经分析好的GISTIC结果。

Using TCGA data as example. TCGA_BLCA.all_thresholded.by_genes.txt is one of the output files from GISTIC algorithm. Pre-analyzed GISTIC results can also be obtained from <http://firebrowse.org/>

```{r}
cnv <- read.table("TCGA_BLCA.all_thresholded.by_genes.txt", sep = "\t", row.names = 1, header = T, check.names = F)

## 构建基因*样本矩阵
## Construct gene × sample matrix
cnv$`Locus ID` = cnv$Cytoband = NULL
colnames(cnv) <- substr(colnames(cnv)[4:ncol(cnv)], 1, 12)
```

# 准备绘图数据
# Prepare plotting data
