---
title: "FigureYa234panImmune"
author: "Houshi Xu"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

小丫姐，能众筹利用USCS上的pan-cancer数据分别通过EPIC和Cibersort算法，计算各个肿瘤的免疫细胞浸润信息，并输出结果，同时计算某个指定基因在各个肿瘤中分别与免疫细胞的相关性，并分别输出相关性热图和相关性信息的excel表？

例文只使用了一种方法，我想用多种方法来证明相关性

# Requirement Description

Sister Xiaoya, can you use the pan-cancer data on USCS to calculate the immune cell infiltration information of each tumor through the EPIC and Cibersort algorithms, and output the results, and at the same time calculate the correlation between a specified gene and immune cells in each tumor, and output the correlation heat map and correlation information in the excel sheet?

The example uses only one method, and I want to use multiple methods to prove correlation

![](example.png)

出自<https://www.frontiersin.org/articles/10.3389/fonc.2021.634617/full>
from<https://www.frontiersin.org/articles/10.3389/fonc.2021.634617/full>

FIGURE 7 | Associations of CD96 expression to tumor purity and immune infiltration.
(B) The correlations of CD96 expression and immune infiltration in cancers.

Besides, we employed **TIMER2.0** to exhibit the landscape of CD96 correlating with various immune infiltrates in human cancers (Figure 7B).
# 应用场景

分别用TIMER、EPIC、Cibersort计算各个肿瘤的免疫细胞浸润信息。

热图可用于展示目的基因在泛癌中与免疫浸润之间的关系。

下面分别展示TIMER、EPIC、Cibersort这三种计算方法并画图（例文用的是TIMER2.0）。三种方法的细胞不同，可相互参照，结合研究兴趣，讨论自己的结果。

# Application Scenarios

TIMER, EPIC, and Cibersort were used to calculate the immune cell infiltration information of each tumor.

Heatmaps can be used to demonstrate the relationship between genes of interest and immune invasion in pan-cancer.

The following three calculation methods are shown and plotted respectively (TIMER 2.0 is used in the example). The cells of the three methods are different, and they can be cross-referenced, combined with research interests, and discuss their own results.

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

install.packages("showtext")

# 安装IOBR，参照https://github.com/IOBR/IOBR
# Install the IOBR, refer to https://github.com/IOBR/IOBR
if (!requireNamespace("BiocManager", quietly = TRUE)) install.packages("BiocManager")
depens <- c('tibble', 'survival', 'survminer', 'sva', 'limma', "DESeq2","devtools",
          'limSolve', 'GSVA', 'e1071', 'preprocessCore', 'ggplot2', "biomaRt",
          'ggpubr', "devtools", "tidyHeatmap", "caret", "glmnet", "ppcor", "timeROC","pracma")
for(i in 1:length(depens)){
  depen<-depens[i]
  if (!requireNamespace(depen, quietly = TRUE))
    BiocManager::install(depen,update = FALSE)
}
if (!requireNamespace("remotes", quietly = TRUE)) install("remotes")
if (!requireNamespace("EPIC", quietly = TRUE))
  remotes::install_github("GfellerLab/EPIC", build_vignettes=TRUE)
if (!requireNamespace("MCPcounter", quietly = TRUE))
  remotes::install_github("ebecht/MCPcounter",ref="master", subdir="Source")
if (!requireNamespace("estimate", quietly = TRUE)){
  rforge <- "http://r-forge.r-project.org"
  install.packages("estimate", repos=rforge, dependencies=TRUE)
}
devtools::install_github("IOBR/IOBR")
# 如果网络不好，可以从github下载后本地安装，安装包已上传至https://share.weiyun.com/g2BTNgVX
# If the network is not good, you can download it from github and install it locally, and the installation package has been uploaded to the https://share.weiyun.com/g2BTNgVX
install.packages("IOBR-master.tar.gz", repos = NULL, type = "source")
```

加载包
load packages

```{r}
library(ggplot2)
library(ggpubr)
library(patchwork)
library(showtext)
library(EPIC)
library(IOBR)

showtext.auto(enable = TRUE)
font.add('arial', 'arial.ttf') #设置字体，windows用户  # Set fonts for Windows users

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 #error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# TIMER

## 从TIMER下载数据

Methods: Tumor IMmune Estimation Resource 2.0 (TIMER2.0; http://timer.cistrome.org/) web server is a comprehensive resource forsystematical analysis of immune infiltrates across diverse cancertypes. At first, we used it to study the differential expressionof CD96 between tumor and adjacent normal tissues across all TCGA cohorts.

We then explored the association between CD96 expression and immune infiltration based on several immune deconvolution algorithms

进入 [TIMER2.0](http://timer.cistrome.org/) 或者 [TIMER2.0 (comp-genomics.org)](http://timer.comp-genomics.org/)，按一下步骤获得分析结果文件：

- 点击Immune
- 在`Gene Expression`里选择`CD96`
- 在`Immune Infiltrates`里每次选一种免疫细胞
- 点击`Submit`
- 点击`Table`下载分析结果
- 重复以上操作，直到获得所有免疫细胞的分析结果，放到data文件夹下。（虽然一次可以选择多个细胞，但网站对数量有限制，所以我们还是一次只选一种细胞。 !!!Querying 115 interested features exceeds the limited amount (n=20). Please reduce the number. ）

# TIMER

## Download data from TIMER

Methods: Tumor IMmune Estimation Resource 2.0 (TIMER2.0; http://timer.cistrome.org/) web server is a comprehensive resource forsystematical analysis of immune infiltrates across diverse cancertypes. At first, we used it to study the differential expressionof CD96 between tumor and adjacent normal tissues across all TCGA cohorts.

We then explored the association between CD96 expression and immune infiltration based on several immune deconvolution algorithms

Go to [TIMER2.0](http://timer.cistrome.org/) or [TIMER2.0 (comp-genomics.org)](http://timer.comp-genomics.org/) and follow the steps below to obtain the analysis result file:

- Click Immune
- Select 'CD96' in 'Gene Expression'
- Select one type of immune cell at a time in 'Immune Infiltrates'
- Click 'Submit'
- Click 'Table' to download the analysis results
- Repeat the above until you have the results of the analysis of all immune cells and put them in the data folder. (Although you can select more than one cell at a time, the website has a limit on the number, so we still only select one cell at a time.) !! Querying 115 interested features exceeds the limited amount (n=20). Please reduce the number. ）

![](TIMER2.png)

## 批量读入TIMER结果
## Batch reading TIMER results

data文件夹里除了从TIMER下载的Table表以外，不要放其他任何文件。
Do not put any files in the data folder except for the table table downloaded from the TIMER.

```{r}
setwd("./data")
file <- dir()
data <- list()
for (i in (1:length(file))) {
  data[[i]] <- read.csv(file[i],header = T)
}
```

## 画图前的数据预处理

TIMER的免疫细胞分类细致，有19类，例文给每一类画成一个小热图，排成了两行，因此画图时麻烦些。

## Pre-processing of data before drawing

There are 19 types of immune cells in TIMER, and each type is drawn into a small heat map arranged in two rows, so it is more troublesome to draw the picture.
