---
title: "FigureYa237circVenn"
author: "Yandong Zheng"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

丫姐好，我想众筹这幅韦恩图，觉得蛮有意思的

# Requirement Description

Hello sister Ya, I want to crowdfund this Venn diagram, I think it's quite interesting

![](example.png)

出自<https://www.mdpi.com/2073-4425/11/5/544>
from<https://www.mdpi.com/2073-4425/11/5/544>

Figure 1. Venn diagram showing the gene expression profile of plant organs. Each compartment indicates the included organs and the number of genes expressed in it. A gene is considered expressed in an organ when its expression is > -2 log2 reads per kilobase per million mapped reads (RPKM) and “not expressed” when its expression is ≤ -2 log2 RPKM.

# 应用场景

展示多组之间的overlap genes、以及每组specific genes，通常会用到venn图（可参考 [FigureYa112venn](https://www.yuque.com/figureya/figureyaplus/figureya112p)），有太多组需要对比的话还可以像例文这样画成环形图。

# Application Scenarios

To show the overlap genes between multiple groups and each specific gene, a venn diagram is usually used (see [FigureYa112venn](https://www.yuque.com/figureya/figureyaplus/figureya112p)), and if there are too many groups to compare, you can draw a donut chart like the example text.

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

install.packages("ggVennDiagram")
```

加载包
load packages

```{r}
library(ggVennDiagram) # 找overlap基因 # Find overlap genes
library(ggplot2)
library(magrittr)
library(RColorBrewer)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # An error message is displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入差异表达分析结果

如果你的数据已经找完overlap，并整理成easy_input.csv的格式，就可以跳过这步，直接进入“输入overlap”。

# Enter the results of the differential expression analysis

If your data has already been found for overlapping and organized into a easy_input.csv format, you can skip this step and go directly to "Enter Overlap".

## 输入文件

SKCM_\*_test_result.\*.txt，差异表达分析结果，出自FigureYa118MulticlassDESeq2和FigureYa120MulticlassedgeR。

## Input files

SKCM_*_test_result.*.txt, differential expression analysis results from FigureYa118MulticlassDESeq2 and FigureYa120MulticlassedgeR.

并不局限于差异表达分析结果文件。只要每个文件中有一列ID即可。

It is not limited to differential expression analysis result files. As long as there is a column ID in each file, it is sufficient.

> 怎样获得差异表达分析结果？

- 用[FigureYa59Plus_GEO2DEG](https://www.yuque.com/figureya/figureyaplus/figureya59p)获得芯片的差异表达分析结果文件`easy_input_limma.csv`

- 用FigureYa118MulticlassDESeq2、FigureYa119Multiclasslimma和FigureYa120MulticlassedgeR获得RNA-seq的差异表达分析结果。

> How do I get the results of differential expression analysis?

- Obtain the differential expression analysis result file 'easy_input_limma.csv' of the chip with [FigureYa59Plus_GEO2DEG](https://www.yuque.com/figureya/figureyaplus/figureya59p)

- Differential expression analysis of RNA-seq was obtained with FigureYa118MulticlassDESeq2, FigureYa119Multiclasslimma, and FigureYa120MulticlassedgeR.

```{r}
fnames <- Sys.glob("*.txt") # 读取当前文件夹下的所有txt文件 # Read all txt files in the current folder
fnames # 检查一下读取的文件名 # Check the file name of the read

#根据自己的数据修改阈值，不根据FDR筛选基因的话就删掉这行
# Modify the threshold according to your own data, and delete this line if you do not screen genes based on FDR
adjcutoff <- 0.05
# 自定义函数读取文件并筛选差异表达基因，只保留基因名
# The custom function reads the file and screens the differentially expressed genes, keeping only the gene name
read_cut_ID <- function(file){
  df <- read.table(file, header = T)
  DEG <- df[df$FDR <= adjcutoff, ]$id #不想根据FDR筛选基因的话就删掉这行，改用下一行 # If you don't want to screen genes based on FDR, delete this line and use the next line instead
  # DEG <- df$id
}
fdataset <- lapply(fnames, read_cut_ID)

names(fdataset)
# 把组名改为字母ABCD等等
# Change the group name to the letters ABCD, etc
names(fdataset) <- LETTERS[1:length(fnames)]
names(fdataset)
```

## 找overlap

使用ggVennDiagram包里的process_region_data处理数据，统计任意并集元素。

## Find overlap

Use the process_region_data in the ggVennDiagram package to process data and count arbitrary union elements.

```{r}
set <- process_region_data(venn = Venn(fdataset)) %>% as.data.frame(.)

dt_all <- set[,c("count", "name")]
colnames(dt_all) <- c("num", "name")

# 将ggVennDiagram包里面name列两个元素之间的..删除
# Delete the two points between the two elements in the name column in the ggVennDiagram package
dt_all$name <- gsub("\\..", "\\", dt_all$name)

# 根据自己情况决定是否删除没有overlap的区域（即overlap基因数为0），
# Decide whether to delete the non-overlapping region (i.e., the number of overlapping genes is 0) according to your own situation.
#dt_all %<>% subset(., num!=0) # 删除overlap基因数为0的区域
```

## 画图
## Drawing
