---
title: "FigureYa223scNMF"
author: "Rongfang Shen; Ying Ge, Yijing Chen"
date: "2025-5-20"
output: html_document
---

![](IOBR.png)

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Requirement description

对单细胞癌上皮数据每个样品进行NMF非负矩阵分解后得到的programs后进行无监督聚类，得到主要cluster后，定义其meta-signature，复现出下图。

After performing NMF (Non-negative Matrix Factorization) on each sample of single-cell carcinoma epithelial data to obtain the programs, unsupervised clustering is conducted. After obtaining the main clusters, their meta-signatures are defined, and the following figure is reproduced.

![](example1.png)

出自<https://www.nature.com/articles/s41422-020-0374-x>

图2 鼻咽癌中揭示的恶性细胞簇与常见恶性特征
c 热图展示了源自11个肿瘤的44个代谢基因的成对相关性。聚类分析在各肿瘤中识别出**五种一致的恶性基因表达特征**。

提需求的小伙伴还找到2017年的这篇Cell文章里也有这图：

from<https://www.nature.com/articles/s41422-020-0374-x>

Fig. 2 Malignant cell clusters and common malignant signatures revealed in NPC.
c A heatmap depicts the pairwise correlations of 44 metagenes derived from 11 tumors. Clustering identified **five coherent malignant gene expression signatures** across the tumors.

The colleague who raised the request also found this figure in this 2017 Cell article:

![](example2.png)

出自<https://linkinghub.elsevier.com/retrieve/pii/S0092867417312709>

图3. 无偏聚类揭示头颈部鳞状细胞癌中p-EMT的共同表达程序
(B) 热图展示了来自十个肿瘤的60个肿瘤内表达程序的成对相关性（如(A)所示）。聚类分析识别出七个跨肿瘤的协调表达程序。热图中对应MEEI25样本的程序行以箭头标注，并按(A)中的编号标示。

from<https://linkinghub.elsevier.com/retrieve/pii/S0092867417312709>

Figure 3. Unbiased Clustering Reveals a Common Program of p-EMT in HNSCC Tumors
(B) Heatmap depicts pairwise correlations of 60 intra-tumoral programs derived from ten tumors, as in (A). Clustering identifies seven coherent expression programs across tumors. Rows in the heatmap that correspond to programs derived from MEEI25 are indicated by arrows and numbered as in (A).

# 应用场景
# Application scenario

**提出问题：**像例文这种鼻咽癌的肿瘤样本，异质性较大。如果按照Seurat的降维聚类流程所聚集的类别，主要变异来源于样品，而不是其生物学特性，这不是我们想要的。

**怎样解决这个问题呢？**例文利用NMF找出每个样品不同的生物学功能program后，对program进行聚类，从而找到上皮样品间共有的生物学功能。具体做法：

- 例文共有11个肿瘤组织，7581个恶性上皮细胞。作者做NMF的时候对每个肿瘤都固定了k = 4，总共得到11 X 4个program。
- 然后对这些program进行聚类，找出在癌上皮中共有的生物学功能改变。
- 最后找出这些signature的metagene，对每个MalignantSig里的每个program前100个基因算其平均的gene loading，选择前30个最高loading的基因作为该signature的marker genes；
- 定义metagene这一块大家可以借助以往处理bulk数据的经验，合理即可。

原文：We then applied dimension reduction analyses, specifically non-negative matrix factorization (NMF), and identified a total of 44 **metagenes** that were preferentially co-expressed  by subpopulations of malignant cells across tumors (Supplementary information, Table S4). Then, hierarchical clustering was used to characterize these 44 metagenes into gene expression signatures, and high concordance was shown among five **signatures** (Fig. 2c; Supplementary information, Table S5). For each signature, we then combined the **top 100 genes of each metagene and calculated the average loadings** for each gene. We summarized the total loadings for repetitive genes, retained the original loadings for exclusive genes, and divided the loadings of each gene by the number of metagenes within the signature. Finally, the top 30 genes with the highest loading were defined as the **marker genes for the signature**.

仔细读全文，小丫发现例文的很多图我们都有[相应的FigureYa](https://k.koudai.com/McawGY6=)啦！例如：

- `FigureYa166scCNV`能帮你实现例文1的Figure 1c 和例文2的Figure 1B。用 单细胞RNA-seq数据计算CNV，用来separate the malignant cells from the non-malignant cells with normal karyotypes.
- `FigureYa194pySCENIC`能帮你实现例文1的Figure 4i，用SCENIC找出potential TFs underlying the regulation across subtypes.
- `FigureYa178receptorLigand`能帮你实现例文1的Figure 6，用CellPhoneDB找出ligand–receptor pairs and molecular interactions among the major cell types.

**Raising the issue:** Tumour samples from nasopharyngeal carcinoma, such as those in the example text, exhibit considerable heterogeneity. If the categories clustered according to Seurat's dimensionality reduction clustering process, the main variation derive from the samples rather than their biological characteristics, this is not what we desire.

**How to address this issue?** The example utilizes NMF to identify distinct biological functional programs in each sample, and then clusters these programs to discover shared biological functions among epithelial samples. The specific approach is as follows:

- The example includes 11 tumor tissues with 7,581 malignant epithelial cells. The author fixed k = 4 for each tumor when performing NMF, resulting in a total of 11 X 4 programs.
-These programs are then clustered to identify shared biological functional changes in cancerous epithelium.
- Finally, the metagenes of these signatures are determined. For each program within each MalignantSig, the average gene loading of the top 100 genes is calculated, and the top 30 genes with the highest loading are selected as the marker genes for that signature.
- In defining the metagene, you can draw on your past experience with bulk data processing; as long as it is reasonable.

Original text: We then applied dimension reduction analyses, specifically non-negative matrix factorization (NMF), and identified a total of 44 **metagenes** that were preferentially co-expressed  by subpopulations of malignant cells across tumors (Supplementary information, Table S4). Then, hierarchical clustering was used to characterize these 44 metagenes into gene expression signatures, and high concordance was shown among five **signatures** (Fig. 2c; Supplementary information, Table S5). For each signature, we then combined the **top 100 genes of each metagene and calculated the average loadings** for each gene. We summarized the total loadings for repetitive genes, retained the original loadings for exclusive genes, and divided the loadings of each gene by the number of metagenes within the signature. Finally, the top 30 genes with the highest loading were defined as the **marker genes for the signature**.

Read the entire text carefully, and Xiao Ya found that we have [corresponding FigureYa](https://k.koudai.com/McawGY6=) for many of the figures in the example text! For instance:

- `FigureYa166scCNV` can help you achieve Figure 1c from Example Text 1 and Figure 1B from Example Text 2. It uses single-cell RNA-seq data to calculate CNV to separate the malignant cells from the non-malignant cells with normal karyotypes.
- `FigureYa194pySCENIC` can help you achieve Figure 4i from Example Text 1, using SCENIC to identify potential TFs underlying the regulation across subtypes.
- `FigureYa178receptorLigand` can help you achieve Figure 6 from Example Text 1, using CellPhoneDB to identify ligand–receptor pairs and molecular interactions among the major cell types.

# 环境设置
# Environment setting


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

install.packages("Seurat")
BiocManager::install("AUCell")
BiocManager::install("GEOquery")
```

加载包

load package

```{r}
library(NMF)
library(data.table)
library(ggsci)
library(pheatmap)
library(AUCell)
library(GEOquery)
library(tidyverse)
library(Seurat)
library(RColorBrewer)
library(corrplot)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 display English error messages
options(stringsAsFactors = FALSE) #禁止chr转成factor prohibit the conversion of chr to factor
```

自定义函数，用[clusterProfiler](https://yulab-smu.top/clusterProfiler-book/)做富集分析。

这里用最常用的两种注释数据库做富集分析：KEGG和GO的biological process。
[GO ORA](https://yulab-smu.top/clusterProfiler-book/chapter5.html#go-over-representation-test)
[KEGG ORA](https://yulab-smu.top/clusterProfiler-book/chapter6.html#kegg-over-representation-test)

也可以修改函数，用自己喜欢的注释数据库做富集分析。可参考FigureYa209batchEnrich，找到适合你的数据的注释数据库。

Customised functions for enrichment analysis using [clusterProfiler](https://yulab-smu.top/clusterProfiler-book/).

Here, enrichment analysis is performed using the two most commonly employed annotation databases: KEGG and GO biological processes.
[GO ORA](https://yulab-smu.top/clusterProfiler-book/chapter5.html#go-over-representation-test)
[KEGG ORA](https://yulab-smu.top/clusterProfiler-book/chapter6.html#kegg-over-representation-test)

Alternatively, you may modify the function to perform enrichment analysis using your preferred annotation database. Refer to FigureYa209batchEnrich to identify an annotation database suitable for your data.

```{r}
Myenrich <- function(genes, category = c("kegg", "gobp"),
                     geneid = c("SYMBOL", "ENTREZID", "ENSEMBL", "UNIPROT")){
  library(clusterProfiler)
  category <- match.arg(category)
  geneid <- match.arg(geneid)
  if (geneid != "ENTREZID"){
    genes <- bitr(genes, fromType = geneid,toType ="ENTREZID",OrgDb="org.Hs.eg.db") %>% .[, 2] %>% as.character()
  }else{genes <- genes}
  if (category == "kegg"){
    enrich <- enrichKEGG(genes, organism = "hsa",keyType = "kegg",
                            pAdjustMethod = "BH",pvalueCutoff  = 0.05,
                            qvalueCutoff  = 0.05, maxGSSize = 5000)
  }
  if (category == "gobp"){
    enrich <- enrichGO(genes, OrgDb="org.Hs.eg.db",ont= "BP",pAdjustMethod = "BH",
                          pvalueCutoff  = 0.05,qvalueCutoff  = 0.05, readable = TRUE)
  }
  return(enrich)
}
```

# 输入文件
# Input file

文件较大，已上传至微云<https://share.weiyun.com/mpXIOggi>

数据下载：从GEO数据库下载GSE150430的GSE150430_npc_scRNA_hg19_processed_data.txt.gz文件，<https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150430>，解压、预处理。

The file is rather large and has been uploaded to Weiyun: <https://share.weiyun.com/mpXIOggi>

Data download: Download the GSE150430_npc_scRNA_hg19_processed_data.txt.gz file from the GEO database (GSE150430), <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE150430>, decompress and preprocess it.
