---
title: "FigureYa304MAGIC"
author: "小丫画图出品"
date: "2023-2-10"
output: html_document
---
欢迎关注“小丫画图”公众号，回复“小白”，看小视频，实现点鼠标跑代码。

小丫微信: epigenomics  E-mail: figureya@126.com

作者：Jarning，他的更多作品看这里<https://k.koudai.com/BXEtOldY>

小丫编辑校验

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

对于单细胞数据，使用MAGIC-knnDREMI计算与某个基因协同表达的基因模块，这篇文章通篇都用了MAGIC，将MAGIC结果在umap图上展示等。

![](example.png)

出自<https://linkinghub.elsevier.com/retrieve/pii/S1535610821004979>

Figure S3. Related to Figures 3 and 4 and to Table S2.
(E) Scaled expression (z-score, imputed by MAGIC with k = 30, t = 3) of genes with high knnDREMI conditioned on PLCG2 > 1 (rows), with cells ordered by PLCG2 expression (columns).
For visualization, expression was smoothed over the ordered cells with a rolling window of 100 cells.
Hierarchical clustering of genes on the unsmoothed, imputed expression was performed with complete linkage and Pearson correlation as a distance metric, identifying 3 gene modules that predict low, medium, and high PLCG2 (purple, gray, and yellow respectively) (Table S13).
Top annotations include PLCG2 expression and SCLC subtype.

# 应用场景

MAGIC特别有利于找时序变化基因、度量基因间的关系。发表MAGIC算法的Cell是更加高级的应用，可参考FigureYa177RNAvelocity <https://k.youshop10.com/RNHFyO4a>，有讲解视频。强烈推荐拖到视频的32分，听Jarning讲解那篇Cell文章使用MAGIC的思路，帮你找单细胞的时序变化基因、度量基因间的关系<https://mp.weixin.qq.com/s/2HRXZicyzHmrsFWUJ9JKdQ>。

# 环境设置


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

BiocManager::install("biomaRt")
install.packages("arrow") # for store large matrix, data.frame or data.table

#
install.packages("reticulate")
reticulate::install_miniconda() # if no conda env in your PC or server
reticulate::py_install("magic-impute") # this will install py packages: `magic` and `scprep`
install.packages("Rmagic") # R interface for magic package
```

加载包

```{r}
library(biomaRt)
library(Seurat)
library(tidyverse)
library(ComplexHeatmap)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

# 输入文件

SCLC_epithelial_cells.rds，Seurat对象。作者已将例文的数据上传至<https://data.humantumoratlas.org/>，点击链接下载<https://cellxgene.cziscience.com/collections/62e8f058-9c37-48bc-9200-e767f318a8ec>，然后保存至data文件夹。

![](download1.png)

![](download2.png)

1. 将ENSEMBL GENE ID转换为GENE NAME

2. 读取Seurat对象，将基因表达矩阵转存为arrow-feather格式，作为MAGIC的输入

```{r}
seu <- readRDS("data/SCLC_epithelial_cells.rds")
dir.create("tmp.data")
saveRDS(seu@meta.data, "tmp.data/SCLC_epithelial_cells.cellmeta.rds")
```

```{r eval=FALSE}
ensembl <- useEnsembl(biomart = "ENSEMBL_MART_ENSEMBL", version = 98)
hg38.ens <- useDataset(dataset = "hsapiens_gene_ensembl", mart = ensembl, verbose = 98)

results <- getBM(attributes = c('ensembl_gene_id', "external_gene_name", 'gene_biotype', 'chromosome_name'),
                 filters = 'ensembl_gene_id',
                 values = rownames(seu),
                 mart = hg38.ens)

results$uniq.name <- ifelse(duplicated(results$external_gene_name),
                            paste(results$ensembl_gene_id, results$external_gene_name, sep = "-"),
                            results$external_gene_name)
rownames(results) <- results$ensembl_gene_id
results <- results[rownames(seu), ]

saveRDS(results, "data/gene.info.rds")
```

```{r}
results <- readRDS("data/gene.info.rds")
all(results$ensembl_gene_id == rownames(seu)) # should be TRUE

counts <- seu[["RNA"]]@counts
rownames(counts) <- results$uniq.name

# !important: the unexpressed genes should be removed.
expr.in.cells <- Matrix::rowSums(counts > 0)
summary(expr.in.cells)
selected.genes <- expr.in.cells > 500 # only genes expressed > 500 cells were kept.
table(selected.genes)
counts <- counts[selected.genes, ]

seu <- CreateSeuratObject(counts, meta.data = seu@meta.data)
rm("counts")
gc()
```

# MAGIC-knnDREMI
