---
title: "FigureYa229PCOA"
author: "Xiaofan Lu"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

把PCOA的散点图变成这样的：

# Requirement Description

Turn the scatter plot of PCOA like this:

![](example.png)

出自<https://www.cell.com/cell-metabolism/fulltext/S1550-4131(20)30312-0?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS1550413120303120%3Fshowall%3Dtrue>

from<https://www.cell.com/cell-metabolism/fulltext/S1550-4131(20)30312-0?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS1550413120303120%3Fshowall%3Dtrue>

Figure 1. Shifts in the Gut Microbiota Composition in Parallel with Impaired Glycemic Status
(A) The **principle coordinate analysis** of Bray-Curtis dissimilarities obtained for the metagenomic species profiles in the discovery (closed circles) and validation (open circles) cohorts. **The circles and error bars indicate the mean and standard errors of the mean**. p values for the comparison of gut microbiota composition in the high-risk NGT (hrNGT), IFG, IGT, CGI, and T2D groups versus the low-risk NGT (lrNGT) group in the discovery cohort are 0.086, 0.069, 0.00070, 0.00010, and 0.00030, respectively, and in the IGT, CGI, and T2D groups versus the NGT group in the validation cohort are 0.009, 0.015, and 0.0001, respectively (PERMANOVA test with 10,000 permutations).

Principal coordinate analysis of the gut microbiota profiles in the discovery cohort showed that **significant shifts separated** the gut microbiota composition of the IGT, CGI, and T2D groups from that of the low-risk NGT group (p = 0.0007, 0.00010, and 0.00030, respectively; PERMANOVA test with 10,000 permutations), but no significant changes in composition were observed between the high- and low-risk NGT groups (p = 0.086), between the IFG and low-risk NGT group (p = 0.069), or between the CGI and T2D groups (p = 0.14; Figure 1A) ... A **similar pattern** was observed for the shifts in microbial composition **in the validation cohort** (Figures 1A and S3A).

# 应用场景

转录组的表达矩阵都有成千上万个基因，这类高维度数据很难直接进行可视化展示。

为了能够在二维平面上展示，就需要对其进行降维，PCA分析是目前最常用的降维工具。

实际应用时，想突出强调自己想要展示的特征，就需要变换展示方式。我们画过各种图来展示PCA结果<https://k.koudai.com/MnKusbPT>：

- 样本量不太大、区分比较好，可以画经典的带圈（置信区间）、带箭头的PCA，画法可参考FigureYa38PCA；
- 大部分情况两个主成分就够了，有时你可能需要3个主成分才能区分开，就需要画3D PCA展示三个主成分，可参考FigureYa164PCA3D；
- 想看多次生物学重复的重复效果，需要用不同形状和颜色同时展示分组和重复，可参考FigureYa101PCA；
- 当分组（亚型）较多且每组样本量较大时，画散点图相互覆盖很难看清，就可以像例文这样计算相同组的均值以及标准误后画图。这样能够看出各亚型距离较远，两个队列的结果相近。这种眼睛看出来的差异有统计学意义吗？作者用PERMANOVA test来做组间差异显著性分析，做法见“附：PERMANOVA test (2-way adonis)”。

# Application Scenarios

The transcriptome expression matrix has thousands of genes, and such high-dimensional data is difficult to visualize directly.

In order to be able to display it on a two-dimensional plane, it is necessary to reduce the dimensionality, and PCA analysis is the most commonly used dimensionality reduction tool at present.

In practice, if you want to highlight the features you want to show, you need to change the way you display them. We have drawn various graphs to illustrate the PCA results<https://k.koudai.com/MnKusbPT>:

- If the sample size is not too large and the discrimination is good, you can draw the classic PCA with circles (confidence intervals) and arrows, and refer to FigureYa38PCA for the drawing method;
- In most cases, two principal components are enough, sometimes you may need 3 principal components to distinguish them, you need to draw 3D PCA to show the three principal components, you can refer to FigureYa164PCA3D;
- If you want to see the repetition effect of multiple biological replicates, you need to display the grouping and repetition with different shapes and colors at the same time, please refer to FigureYa101PCA;
- When there are many groups (subtypes) and the sample size of each group is large, and it is difficult to see that the scatter plots cover each other, you can calculate the mean of the same group and the standard error after plotting as in the example. This allows for the distantness of the subtypes and the similar results of the two cohorts. Is there a statistically significant difference in what the eye sees? The authors used the PERMANOVA test to analyze the significance of differences between groups, as described in "Appendix: PERMANOVA test (2-way adonis)".

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

install.packages("ape")
install.packages("usedist")
```

加载包
load packages

```{r}
library(vegan)        #用于计算距离矩阵 # Used to calculate the distance matrix
library(ape)          #做PCoA # Do PCoA
library(RColorBrewer) #配色 # Color matching
library(usedist)      #PERMANOVA test时处理距离矩阵 # Handle distance matrices when doing PERMANOVA test

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入文件

easy_input_expr.txt，表达矩阵。

easy_input_sinfo.txt，样本信息，第一列是sample ID，跟表达矩阵一致，第二列是第一种分组方式（这里是亚型），第三列是二种分组方式（这里是队列），这里对比两个队列。

# Input files

easy_input_expr.txt, the expression matrix.

easy_input_sinfo.txt, the sample information, the first column is the sample ID, which is consistent with the expression matrix, the second column is the first grouping method (here the subtype), and the third column is the two grouping methods (here the cohort), and here the two cohorts are compared.
