---
title: "FigureYa231scScore"
author: "Yi Xiong"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

tsne降维，标记GO功能

# Requirement Description

tsne dimensionality reduction, marking GO function

![](example.png)

出自<https://www.frontiersin.org/articles/10.3389/fimmu.2020.606164/full>
from<https://www.frontiersin.org/articles/10.3389/fimmu.2020.606164/full>

FIGURE 2 | Functional enrichment analysis.
(D) The distribution of GBM and Control cells in GSE135437.
(E) The distribution of status of macrophages in GSE135437.
(F) The distribution of top-six biological processes in GSE135437.

Almost all **M2 macrophages were in the GBM cells**,
while **M1 macrophages were in the control cells** (Figures 2D, E).
Furthermore we explored the **relationship between the biological process and cell types** and found that the signal transduction and angiogenesis enriched in a subgroup of M2 macrophages, however the cell adhesion widely distributed in both control and GBM cells (Figure 2F).

# 应用场景

单细胞数据，从matrix到分组，用Single-Cell Signature Explorer 软件计算pathway的表达量，相比于FigureYa160scGSVA所用的GSVA运行速度更快。

# Application Scenarios

Single-cell data, from matrix to grouping, was calculated using Single-Cell Signature Explorer software, which was faster than the GSVA used in the FigureYa160sc GSVA.

> 怎样从t-SNE图中解读出有意义的信息？

需要从各个角度对比展示，从而找到能讲故事的角度：

- 用颜色标注各种细胞类群、突出显示其中某一/几种细胞类型、显示分组/分类信息等等。总之把你想要展示的细胞信息放在metadata里，就可以了。方法见下文“ 根据metadata里的信息分组”。
- 展示某一/几个基因（例如marker基因）的表达量跟细胞类型之间的关系，可参考FigureYa27t-SNE。
- 展示pathway的表达量跟细胞类型直接的关系。本文档“用渐变色展示pathway的表达量”部分将用Single-Cell Signature Explorer计算 score。另外，pathway的表达量还可以并用小提琴图的方式展示，并计算pathway在组间的差异显著性，可参考FigureYa160scGSVA。

> How can meaningful information be deduced from t-SNE plots?

You need to compare and contrast the presentation from various angles to find the angle from which the story can be told:

- Color-code various cell populations, highlight one/several cell types, display grouping/taxonomic information, and more. In short, put the cell information you want to display in metadata, and you're good to go. See below for the method, "Grouping by information in metadata".
- To show the relationship between the expression of one or several genes (e.g., marker gene) and cell type, refer to FigureYa27t-SNE.
- Demonstrate a direct relationship between pathway expression and cell type. In the section of this document, "Showing the Expression of a Pathway with Gradient Colors", the score will be calculated using the Single-Cell Signature Explorer. In addition, the expression of pathway can also be displayed in the form of violin diagram, and the significance of the difference of pathway between groups can be calculated, which can be referred to FigureYa160scGSVA.

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")

```

加载包

```{r}
library(tidyverse)
library(Seurat)
library(GO.db)
library(biomaRt)
library(RColorBrewer)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息  # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 标准的seurat流程

GSE135437，从<https://github.com/rsankowski/sankowski-et-al-microglia>下载数据并解压到当前文件夹

# Standard Seurat process

GSE135437, download the data from the <https://github.com/rsankowski/sankowski-et-al-microglia> and extract it to the current folder
