---
title: "FigureYa38 PCA"
author: "Zhougeng Xu; Ying Ge, Yijing Chen"
date: "2025-5-20"
output: html_document
---

![](http://oex750gzt.bkt.clouddn.com/18-9-3/64881053.jpg)

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述
## Requirement description

用TCGA数据画主成分分析(pricinple components analysis，PCA)的图。

例如：[画个小圈圈](https://mp.weixin.qq.com/s/8kqTVSDjuHpivGaim6ukNg)里的这个图：

Draw plot of principal components analysis (PCA) with TCGA data.

For example, this graph in [draw a small circle](https://mp.weixin.qq.com/s/8kqTVSDjuHpivGaim6ukNg):

![](小圈圈.png)

![](example1.png)

出自<http://www.haematologica.org/content/102/7/e245>

from<http://www.haematologica.org/content/102/7/e245>

![](example2.png)

出自<http://clincancerres.aacrjournals.org/content/23/8/2105.figures-only>

from<http://clincancerres.aacrjournals.org/content/23/8/2105.figures-only>

## 应用场景
## Application scenario

通常我们拿到的数据集，例如转录组的表达量矩阵都有成千上万个基因，这类高维度数据很难直接进行可视化展示。

为了能够在二维平面上展示，就需要对其进行降维，而PCA分析是目前最常用的降维工具。

然后你就能在图上一眼看出哪些样品之间更像，哪些样品属于/不属于同一组。不仅适用于测序数据。

场景一：多个分组，每组三次生物学重复。做PCA，看到某个样品远离了组内其他样品，它有什么问题？

场景二：多个分组，分三批送去测序。做PCA，看不出分组的规律，却看到同一批次的样品间距离更近，批次效应？

Usually we get datasets, for example, the expression matrix of transcriptome has thousands of genes, and this kind of high-dimensional data is difficult to visualize and display directly.

In order to be able to display it on a two-dimensional plane, it needs to be downscaled, and PCA analysis is currently the most commonly used downscaling tool.

Then you can see at a glance on the graph which samples are more similar to each other and which samples belong/do not belong to the same group. Not only applicable to sequencing data.

Scenario 1: Multiple subgroups, three biological replicates per group. Do PCA and see that one sample is far away from the other samples in the group, what's wrong with it?

Scenario 2: Multiple subgroups, sent for sequencing in three batches. Do PCA, don't see a pattern in the grouping, but see samples in the same batch closer to each other, batch effect?

## 输入数据
## Input data

两个输入文件:

- 表达量矩阵：easy_input_expr.csv，第一列是样品名，后续几列是各个基因的表达量，或者其他属性
- 样品分组：easy_input_meta.csv，第一列是样品名，第二列是分组

此处输入数据为3种癌症BRCA、OV、LUSC里，5个基因"GATA3", "PTEN", "XBP1","ESR1", "MUC1"的表达量。

数据来源：用R包`RTCGA.mRNA`提取TCGA数据。也可参考FigureYa18、21（临床信息）、FigureYa22、23、34（表达矩阵）里的数据下载方法。

Two input files:

- Expression matrix: easy_input_expr.csv, the first column is the sample name, the subsequent columns are the expression of each gene, or other attributes
- Sample grouping: easy_input_meta.csv, the first column is the sample name, the second column is the grouping.

The input data here is the expression of 5 genes "GATA3", "PTEN", "XBP1", "ESR1", "MUC1" in 3 cancers BRCA, OV, LUSC.

Data source: using the R package `RTCGA.mRNA` extracts TCGA data. See also the data download methods in FigureYa18, 21 (clinical information) and FigureYa22, 23, 34 (expression matrix).

```{r}
source("install_dependencies.R")
expr_df <- read.csv(file='easy_input_expr.csv',row.names = 1,
                    header = TRUE, sep=",", stringsAsFactors = FALSE)
meta_df <- read.csv(file='easy_input_meta.csv', row.names = 1,
                    header = TRUE, sep=",",stringsAsFactors = FALSE)
#查看前3个基因在前4个sample中的表达矩阵
#view the expression matrix of the first 3 genes in the first 4 samples
expr_df[1:3,1:4]
#查看样本信息前3行
#view the first 3 lines of sample information
head(meta_df, n=3)
```

## 做PCA
## Do PCA
