---
title: "FigureYa68friendsV2"
author: "Zhongtian Xu, Taojun Ye"
reviewer: "Ying Ge"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# 设置knitr的全局代码块选项 / Set global chunk options for knitr
```

# 需求描述

输入基因列表，用GO注释的相似性找它的好朋友，画出云雨图。

# Requirement Description.

Input a list of genes and use GO annotated similarity to find its best friends and draw a cloud and rain diagram.

![](example.png)

出自<https://www.sciencedirect.com/science/article/pii/S1874391912007567>

from<https://www.sciencedirect.com/science/article/pii/S1874391912007567>

# 应用场景

原理看这篇：<https://mp.weixin.qq.com/s/_Wt_GmC8yjcvEdXBNRUQTw>

在⽣信分析中，⼀个常⻅的应⽤场景就是：从特定蛋⽩质的功能信息出发，查找与其功能相似或相关的蛋⽩质，并对这些蛋⽩质间的关联程度进⾏⽐较、量化。这是生物信息学中经常遇到的问题，GO语义相似度为这种分析提供了可能。

通常认为，如果两个基因产物的功能相似，那么他们在GO中注释的术语（term），在GOtree中所处的位置就⽐较相近，反映在语义相似度上，就是他们的语义相似度比较⾼。

在我们⽇常的分析实践中，经常能够拿到⼀大堆的差异表达基因，从⾥面挑选哪⼀个基因出来进行验证常常让我们感到困扰。通常我们会对差异基因的条⽬进⾏富集分析，看看是否富集在某个GOterm或者KEGG通路当中。这时候已经对结果进⾏了相当程度上的清晰化。

但是如果富集到的某个我们感兴趣的通路中基因数⽬依然⽐较多，那么如何从这⼀堆基因中挑选最重要的那个就是⼀个问题。哪些基因会比较重要呢？

- 第⼀个线索是基因的差异改变的程度⽐较大，但差异改变程度⼤并不一定代表重要。
- 第⼆个线索就是该基因的产物与通路上的其它基因产物都有互作的话。简⽽言之，该基因编码蛋⽩的“朋友”⽐较多的话，那么该基因就可能⽐较重要。

跳出差异基因的范畴，任何⼀个涉及到基因互作的场景，例如打质谱拉下来的⼀堆蛋⽩、酵⺟双杂交得到的⼀堆蛋⽩，都可以通过类似⽅法找到最重要的那个hub基因。

# Application scenarios

See this for the principle: <https://mp.weixin.qq.com/s/_Wt_GmC8yjcvEdXBNRUQTw>

A common application in the analysis of growth information is to look for similar or related proteins from the functional information of a particular protein, and to compare and quantify the degree of association between these proteins. This is a frequently encountered problem in bioinformatics, and GO semantic similarity makes such analysis possible.

It is generally believed that if two gene products are functionally similar, the terms they annotate in GO and their positions in the GOtree are relatively similar, which is reflected in their semantic similarity.

In our daily analysis practice, we often get a large number of differentially expressed genes, and it is often a problem for us to choose which gene to verify. Often, we will perform enrichment analysis on the differentially expressed gene entries to see if they are enriched in a certain GOterm or KEGG pathway. At this point, the results have been clarified to a considerable degree.

However, if the number of genes enriched for a pathway of interest is still relatively large, then it is a question of how to select the most important genes from the pile. Which genes are important?

- The first clue is that the gene has a relatively large degree of differential alteration, but a large degree of differential alteration does not necessarily mean that it is important.
- The second clue is if the gene's product interacts with the products of other genes in the pathway. In short, if the gene has more “friends” encoding protein, then the gene is likely to be more important.

Going beyond differential genes, any scenario that involves gene interactions, such as a pile of egg white from mass spectrometry or a pile of egg white from a double hybrid, can be used in a similar way to find the most important hub gene.

Translated with DeepL.com (free version)

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")
```
