---
title: "FigureYa147interaction"
author: "Zongcheng Li"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述
## Requirements description

用interaction文件作为输入画调控网络图，线的长短与相互作用score对应，线的数量和颜色与证据对应。
Use the interaction file as input to draw a regulatory network diagram. The length of the line corresponds to the interaction score, and the number and color of the line correspond to the evidence.

![](example1.png)

出自<https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4951236>
From <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC4951236>

这篇文章用matlab写了个脚本OncoBinder.m，free online，能根据多种数据找interaction、打分、画图，感兴趣的小伙伴写信给文章作者吧～
This article uses Matlab to write a script OncoBinder.m, which is free online and can find interactions, score, and draw diagrams based on a variety of data. Interested friends can write to the author of the article~

![](example2.png)

出自<https://www.nature.com/articles/s41589-018-0161-x>

Fig. 1 | Identification of HIP1R as a regulator of PD-L1. a, Schematic representation of the identification process. A set of low-confidence interactors of PD-L1 according to high-throughput assays (left panel) were analyzed by the OncoBinder model (middle panel) to identify functionally relevant interactions. The ‘co-activation’ scores of candidate interactors are plotted in the right panel, with statistically significant associations marked in red (P < 0.01, Pearson correlation, two-sided). The modeling experiment was replicated for three times with the same results obtained.

**图的解析**
**Analysis of the graph**

用改写的OncoBinder（没有提供脚本）找interaction并画图。中央是PD-L1，周围是跟它有相互作用关系的蛋白质。
Use the rewritten OncoBinder (no script provided) to find the interaction and draw the graph. PD-L1 is in the center, and the proteins that interact with it are around it.

Figure legend和正文都没有描述线的长短和数量与哪个特征相对应。
Neither the figure legend nor the text describes which feature the length and number of lines correspond to.

我们结合这两幅图设计了下面的图，设计思路看这篇：<https://mp.weixin.qq.com/s/cQzsw1DKTUqooBmE078k2A>
We combined these two graphs to design the following graph. For the design ideas, see this article: <https://mp.weixin.qq.com/s/cQzsw1DKTUqooBmE078k2A>

## 应用场景
## Application scenarios

展示相互作用关系，同时展示相互作用证据的来源和打分。
Show the interaction relationship, and at the same time show the source and score of the interaction evidence.

像例文中的蛋白相互作用，或转录因子-靶基因，或miRNA-靶基因的关系。
Like the protein interaction in the example text, or the relationship between transcription factor-target gene, or miRNA-target gene.

其中转录因子-靶基因的关系可以这样设计：中央画转录因子，周围画靶基因；连线代表各种证据例如有motif、有ATAC-seq的peak、有H3K27ac的peak等等；自己给这些证据权重，计算一个总体的打分作为连线的长度。
The relationship between transcription factor-target gene can be designed as follows: draw the transcription factor in the center and the target gene around it; the connecting line represents various evidences such as motif, ATAC-seq peak, H3K27ac peak, etc.; give these evidence weights by yourself and calculate an overall score as the length of the connecting line.

## 环境设置
## Environment settings


```{r}
source("install_dependencies.R")

```

加载包
Load package

```{r}
library(org.Hs.eg.db)
library(clusterProfiler)
library(ggplot2)
library(reshape2)
library(igraph)
library(magrittr)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 #Display English error message
options(stringsAsFactors = FALSE) #禁止chr转成factor #Disable chr conversion to factor
```

自定义画图函数
Custom drawing function

`fcircle` is built for the white background of PD-L1

```{r}
mycircle <- function(coords, v=NULL, params) {
  vertex.color <- params("vertex", "color")
  if (length(vertex.color) != 1 && !is.null(v)) {
    vertex.color <- vertex.color[v]
  }
  vertex.size  <- 1/200 * params("vertex", "size")
  if (length(vertex.size) != 1 && !is.null(v)) {
    vertex.size <- vertex.size[v]
  }
  vertex.frame.color <- params("vertex", "frame.color")
  if (length(vertex.frame.color) != 1 && !is.null(v)) {
    vertex.frame.color <- vertex.frame.color[v]
  }
  vertex.frame.width <- params("vertex", "frame.width")
  if (length(vertex.frame.width) != 1 && !is.null(v)) {
    vertex.frame.width <- vertex.frame.width[v]
  }

  mapply(coords[,1], coords[,2], vertex.color, vertex.frame.color,
         vertex.size, vertex.frame.width,
         FUN=function(x, y, bg, fg, size, lwd) {
           symbols(x=x, y=y, bg=bg, fg=fg, lwd=lwd,
                   circles=size, lheight = 30, add=TRUE, inches=FALSE)
         })
}

add.vertex.shape("fcircle", clip=igraph.shape.noclip,
		 plot=mycircle, parameters=list(vertex.frame.color=1,
                                  vertex.frame.width=1))
```

## 参数设置
## Parameter settings
