---
title: "FigureYa260CNV"
params:
  author: "Shuwen Cheng; Yasi Zhang"
  reviewer: "Ying Ge"
  date: "2025-05-20"
output: html_document
---

**Author(s)**: `r params$author`
**Reviewer(s)**: `r params$reviewer`
**Date**: `r params$date`

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Demand description

画出CNV的这个图。

Draw this CNV plot.

![](example.png)

出自：<https://molecular-cancer.biomedcentral.com/articles/10.1186/s12943-021-01322-w>，跟FigureYa259circLink和FigureYa262GDC出自同一篇文章。

图1. CRC中RNA修饰“书写器”的遗传及转录改变。
c, 柱状图显示TCGA-COAD/READ队列中RNA修饰“书写器”的CNV增益（红色）、缺失（蓝色）和非CNV（绿色）的频率。每根柱子的高度代表改变频率。

Source: <https://molecular-cancer.biomedcentral.com/articles/10.1186/s12943-021-01322-w>, it is from the same article as FigureYa259circLink and FigureYa262GDC.

Fig. 1 Genetic and transcriptional alterations of RNA modification “writers” in CRC.
c, Bar graphs showing the frequency of CNV gain (red), loss (blue) and non_CNV (green) of RNA modification “writers” in the TCGA-COAD/READ cohort. The height of each bar represents the alteration frequency.

# 应用场景
# Application scenarios

外显子测序数据可以画这样的图。

RNA-seq数据也可以借助这样的分析来深挖。例如RNA-seq筛出成百上千个差异基因，谁才是关键基因？它通过什么方式影响了下游基因的表达？或许是某些基因发生了高频copy number variation（gain/loss）。

Exome sequencing data can generate similar plots.

RNA-seq data can also be leveraged for in-depth analysis using such approaches. For example, when RNA-seq identifies hundreds or thousands of differentially expressed genes, how can we pinpoint the key drivers? By what mechanisms do they regulate downstream gene expression? One possible explanation could be high-frequency copy number variations (gain/loss) in certain genes.

# 环境设置
# Environment Setup


```{r}
source("install_dependencies.R")

```

# 加载包
# Loading packages

```{r}
library(magrittr)
library(tidyverse)
library(org.Hs.eg.db)
library(TCGAbiolinks)
library(clusterProfiler)
library(SummarizedExperiment)

# 显示英文报错信息
# Show English error messages
Sys.setenv(LANGUAGE = "en")

# 禁止chr转成factor
# Prevent character-to-factor conversion
options(stringsAsFactors = FALSE)
```

# 输入数据
# Input Files

从GDC下载hg38的CNV数据，以COAD/READ为例。

easy_input_gene.txt，要看哪些基因的CNV，就写到这个文件里。甚至可以把所有差异基因或参与某个通路的基因都拿来计算，查看输出文件`output_cnv.csv`。然后筛选gain或loss频率高的前十几个基因，进行画图展示。

Download hg38 CNV data from GDC, using COAD/READ as an example.

The file easy_input_gene.txt should contain the list of genes for CNV analysis. You may include all differentially expressed genes or pathway-related genes for calculation, then check the output file `output_cnv.csv`. Subsequently, select the top ~10 genes with high-frequency gains or losses for visualization.
