---
title: "FigureYa179AMDAplot"
author: "Enyu Lin"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      tidy=TRUE,
                      dev = "svglite",
                      fig.ext = ".svg")
```

# 需求描述

# Requirement

重复文章中ageing-mediated disease alignment score (AMDA score)的算法，并复现原文的Figure3a。
Repeat the algorithm of aging-mediated disease alignment score (AMDA score) in the article and reproduce Figure3a of the original text.

![](example1.jpg){width=45%}

![](example2.jpg){width=45%}

Fig. 3 Association of ageing-associated gene expression changes with ageing diseases and lifespan. a Ageing-mediated disease alignment scores (AMDA scores) for different ageing (x axis) and disease data sets (y axis). Abbreviations: AD Alzheimer’s disease, CAD coronary artery disease, CPI cancer proliferation index, cross-cohort cross-cohort study of blood ageing, HMP human metabolic pathways, IR insulin resistance, LL longitudinal ageing data, MCI mild cognitive impairment; SAFHS, San Antonio Family Heart Study.

出自<https://www.nature.com/articles/s41467-017-02395-2>
fromhttps://www.nature.com/articles/s41467-017-02395-2

# 应用场景

# Application Scenarios

把自己的数据跟衰老联系起来，肿瘤、非肿瘤领域都可以用起来。具体原理参考下文的“AMDA score算法讲解”。
Linking one's own data to aging can be applied in both oncology and non-oncology fields. The specific principle can be referred to in the "Explanation of AMDA score Algorithm" below.

适用于转录组数据，用文中的公式计算出样本的**AMDA score（量化衰老与疾病相关的指标）**，进行组间对比。
Applicable to transcriptome data, calculate the **AMDA score (an indicator quantifying aging and disease-related indicators) ** of the samples using the formula in the text, and conduct inter-group comparisons.

也可以用自己的数据画这种**带有局部二级标签**的热图。
You can also use your own data to draw such a heat map with local secondary labels.

# 环境设置

# Environment Setup


```{r message=FALSE, warning=FALSE}
source("install_dependencies.R")

if (!require("pacman")) install.packages("pacman")
pacman::p_load(svglite, ggplot2, magrittr,patchwork,Rmisc,Cairo,stringr,tidyr,tidyfst,data.table,rlist,Hmisc,grid,gtable,gridExtra,ggpubr)
if (!require("DESeq2")) BiocManager::install("DESeq2")
if (!require("S4Vectors")) BiocManager::install("S4Vectors")
if (!require("BiocParallel")) BiocManager::install("BiocParallel")
```

加载R包
Load the R package

```{r message=FALSE, warning=FALSE}
library(DESeq2)
library(S4Vectors)
library(BiocParallel)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # Display an English error message
options(stringsAsFactors = FALSE) #禁止chr转成factor  # prohibit chr from being converted to factor
```

> 如果你对AMDA score算法不感兴趣，只想画带二级分类标签的热图，就直接跳到“开始画图”。
> If you are not interested in the AMDA score algorithm and only want to draw a heat map with secondary classification labels, just jump directly to "Start Drawing".

# AMDA score算法讲解

# Explanation of the AMDA score Algorithm

要明白AMDA score得先看原文中的Figure 3a，其中分为两部分。
To understand AMDA score, one must first look at Figure 3a in the original text, which is divided into two parts.

- Part-1属于衰老基因集，是作者自己通过不同物种和组织经过年龄分层的多因素ANOVA进行计算所得的。衰老基因集包含了体现衰老信息的差异基因集，里面分为up和down（详见原文Supplementary Data）。衰老基因集是计算AMDA score的基础，但这个无法复现。因为衰老基因集是作者使用包括斑马鱼、非洲青鳉鱼、小鼠和人的各种组织以及纤维细胞系的基因组所计算得出的。所以，Part-1的基因集在Supplementary Data有提供，直接使用即可。
  Part-1 belongs to the aging gene set and was calculated by the author himself through multi-factor ANOVA with age stratification of different species and tissues. The aging gene set contains a set of differentially expressed genes that reflect aging information, which is divided into up and down (for details, see the original Supplementary Data). The aging gene set is the basis for calculating the AMDA score, but this cannot be reproduced. Because the aging gene set was calculated by the authors using the genomes of various tissues including zebrafish, African medaka, mice and humans, as well as fibroblast cell lines. So, the gene set of Part-1 is available in Supplementary Data and can be used directly.

- Part-2是包括cancer、cvd等各种疾病的病人基因组数据，通过不同的分组策略（见原文Supplementary Table 5）后进行差异基因分析（Deseq2)。然后将差异结果里的log2 Foldchange（原文里描述的是Foldchange，其实实际上是log2FC）在进行normalization（详见下文**"计算AMDA score"**），最后得出取值范围为[-1,1]的log2FC值。
  Part-2 is genomic data of patients with various diseases including cancer and cvd, and differential gene analysis (Deseq2) was conducted after different grouping strategies (see Supplementary Table 5). Then, the log2 Foldchange in the difference results (described as Foldchange in the original text, but in fact it is log2FC) is subjected to normalization (for details, see below **" Calculating AMDA score"**). Finally, a log2FC value with a range of [-1,1] is obtained.

**AMDA score公式的分子部分表示某个病种中匹配到某个基因集中up组的差异基因的rank normalization后的log2FC值的和，减去匹配到down组的差异基因的log2FC的和。分母则是这个基因集中up和down组基因的总数量。**
The molecular part of the AMDA score formula represents the sum of log2FC values after rank normalization of the differentially expressed genes matched to the up group in a certain gene set in a certain disease, minus the sum of log2FC values of the differentially expressed genes matched to the down group. The denominator represents the total number of genes in the up and down groups of this gene set.

![](example3.jpg){width=50%}

下面以Part-2里的一种疾病为例，展示AMDA score 的计算过程。这里使用**TCGA-BLCA**的转录组数据进行计算（**为什么不用原文中所用的ICGC-BLCA数据，见文末 Q & A**）
The following takes a disease in Part-2 as an example to demonstrate the calculation process of the AMDA score. Here, the transcriptome data of **TCGA-BLCA** is used for calculation (**Why not use the ICGC-BLCA data used in the original text, see Q&A at the end of the text** )

# 输入数据预处理

# Input data preprocessing

实际应用时，如果已经准备好了基因表达矩阵，可跳过这步，直接进入“差异表达分析”。如果已经做好差异表达分析，就直接进入“输入文件”。
In practical application, if the gene expression matrix has already been prepared, this step can be skipped and you can directly enter "Differential expression Analysis". If the differential expression analysis has been completed, directly enter the "Input File".

中间文件已上传至微云<https://share.weiyun.com/c1ToPq7z>
Cloud is a kind of intermediate files uploaded to < https://share.weiyun.com/c1ToPq7z >

## 清洗TCGA-BLCA数据

## Clean the TCGA-BLCA data

这里为了复现原文，因此提取与原文匹配的样本名，用到ICGC-BLCA的转录组数据，但不用里面的raw.count。因ICGC-BLCA数据过大（1.67G），这里不再运行。感兴趣的小伙伴可以自己下载数据放入当前目录内运行。
Here, in order to reproduce the original text, sample names that match the original text are extracted, using the transcriptome data of ICGC-BLCA, but not the raw.count inside. Because the ICGC-BLCA data is too large (1.67G), it is no longer running here. Interested friends can download the data themselves and run it in the current directory.
