---
title: "FigureYa235scDEG"
author: "Yandong Zheng"
reviewers: "Ying Ge, Hui Huang"
date: "2025-9-12"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005

```{r setup, include = FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

单细胞测序分不同的cluster，然后在每个cluster上做DEGs的点图，最好是log2FC为负的标蓝。然后自定义显示想要的点。
输入数据用seurat用的PBMC那个测试数据吧

# Requirement Description

Single-cell sequencing is divided into different clusters, and then a dot plot of DEGs is made on each cluster, preferably with a negative log2FC in blue. Then customize the display of the desired points.
The input data will be the PBMC test data for the seurat

![](example.png)

出自<https://www.sciencedirect.com/science/article/pii/S0092867419312826?via%3Dihub>
from<https://www.sciencedirect.com/science/article/pii/S0092867419312826?via%3Dihub>

Figure 2. Global Spatiotemporal Analysis of Three Cardiac Developmental Stages
(H) Differential gene expression analysis showing up- and down-regulated genes across all ten clusters. An adjusted p value < 0.01 is indicated in red, while an adjusted p value R 0.01 is indicated in black.

# 应用场景

分类展示各cluster的marker gene表达趋势变化的分布情况。

可以像原文那样标注显著上调或下调表达的基因名，或者标注参与某一通路的基因名。

# Application Scenarios

The distribution of marker gene expression trends in each cluster is displayed by category.

Gene names that are significantly up- or down-regulated, or gene names involved in a pathway, can be labeled as in the original text.

# 环境设置


# Environment settings

```{r}
source("install_dependencies.R")
```

加载包
load packages

```{r}
library(dplyr)
library(Seurat)
library(patchwork)
library(reshape2)
library(RColorBrewer)
library(ggplot2)
library(ggrepel)  # Used to annotate text
library(magrittr)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入文件及预处理

## 输入文件

这部分与“FigureYa224scMarker”相同

以Seurat官网PBMC数据为例，跟FigureYa111markerGene和FigureYa160scGSVA用的是同一套数据，出自不同作者之手，可相互参考。

下载<https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz>，解压到当前文件夹。

预处理过程参考<https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html>

# Input files and preprocessing

## Input files

This part is the same as "FigureYa224scMarker".

Taking the PBMC data on the official website of Seurat as an example, it uses the same set of data as FigureYa111markerGene and FigureYa160scGSVA, which are from different authors and can be cross-referenced.

Download the <https://cf.10xgenomics.com/samples/cell/pbmc3k/pbmc3k_filtered_gene_bc_matrices.tar.gz> and extract it to the current folder.

Refer to <https://satijalab.org/seurat/v3.2/pbmc3k_tutorial.html> for the pretreatment process

```bash
# 解压缩
# Unzip
tar -zxvf pbmc3k_filtered_gene_bc_matrices.tar.gz
```
