---
title: "FigureYa315SingleRFraction"
params:
  author: "Xiaofan Lu; Yasi Zhang"
  reviewer: "Ying Ge"
  date: "2025-05-20"
output: html_document
---

**Author(s)**: `r params$author`
**Reviewer(s)**: `r params$reviewer`
**Date**: `r params$date`

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Demand description

在单细胞数据中，怎么用singleR预测不同细胞亚群之间的相关性。

In single-cell data, how to use SingleR to predict the correlations between different cell subpopulations.

![](method.png)

![](example.png)

出自：<https://www.cell.com/cell/fulltext/S0092-8674(21)00010-6?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867421000106%3Fshowall%3Dtrue>

图S5. LAMP3+ cDC细胞的起源与调控（与图4相关）。
(D) 箱线图展示不同癌症类型中，cDC2来源的LAMP3+ cDC细胞占所有LAMP3+ cDC细胞的比例。Kruskal-Wallis检验。

Source: <https://www.cell.com/cell/fulltext/S0092-8674(21)00010-6?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS0092867421000106%3Fshowall%3Dtrue>

Figure S5. Origins and regulation of LAMP3+ cDCs, related to Figure 4.
(D) Boxplot showing the fractions of cDC2-derived LAMP3+ cDCs in all LAMP3+ cDCs across different cancer types. Kruskal-Wallis test.

# 应用场景
# Application scenarios

例文针对`origins of the LAMP3+ cDCs`的问题做了多个角度的探讨，可以灵活运用到自己感兴趣的细胞类型中。

本文档使用SingleR计算cDC3_LAMP3来源于cDC2的占比，实现Figure S5D。

同一篇文章的Figure S5B（Stratification of cDC transcriptomes by scores generated from signature genes of three cDC subsets）可以用FigureYa314SingleRScore <https://k.youshop10.com/2DkhulHr>来计算和画图。

The example text explores the question of `origins of the LAMP3+ cDCs` from multiple perspectives, which can be flexibly applied to other cell types of interest.

This document uses SingleR to calculate the proportion of cDC3_LAMP3 derived from cDC2, reproducing Figure S5D.

For Figure S5B of the same paper (Stratification of cDC transcriptomes by scores generated from signature genes of three cDC subsets), the FigureYa314SingleRScore <https://k.youshop10.com/2DkhulHr>can be used for calculation and plotting.

# 环境设置
# Environment Setup


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

BiocManager::install("SingleR")
```

# 加载包
# Loading packages

```{r}
library(Matrix)
library(Seurat)

# 基于参考数据集对新细胞进行类型标注
#labels new cells from a test dataset based on similarity to the reference
library(SingleR)

library(ggplot2)

# 用于实现Kruskal-Wallis test
# Provides Kruskal-Wallis test
library(ggpubr)

# 显示英文报错信息
# Show English error messages
Sys.setenv(LANGUAGE = "en")

# 禁止chr转成factor
# Prevent character-to-factor conversion
options(stringsAsFactors = FALSE)
```

# 输入文件
# Input Files

原文图中有14个cohort，作者上传到GEO数据库的只有9个，<https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE154763>。我们从Supplementary file表格里下载csv.gz文件，放到当前目录下的InputData文件夹。

The original figure included data from 14 cohorts, but only 9 cohorts were uploaded by the authors to the GEO database, <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE154763>. We downloaded the csv.gz files from the Supplementary file table and saved them in the **InputData** folder under the current directory.

![](InputData.png)

实际应用中，直接跳到下一步“SingleR”，加载自己的seurat对象。

In practice, you can directly proceed to the next step "SingleR" and load your own Seurat object.

```{r}
data.path <- file.path("InputData")

# 读取表达谱
# Read expression matrices
emats <- lapply(list.files(data.path, pattern = "expression", full.names = T), function(x){
  emat = data.table::fread(x, data.table = F, sep = ",")
  rownames(emat) = emat$index; emat$index = NULL
  emat = as(as.matrix(t(emat)), "dgCMatrix")
  return(emat)
})

einfos <- lapply(list.files(data.path, pattern = "metadata", full.names = T), function(x){
  einfo = read.csv(x)
  rownames(einfo) = einfo$index
  return(einfo)
})

seu.objs <- mapply(function(x, y){
  seu = CreateSeuratObject(x, meta.data = y)
}, emats, einfos)
seu <- merge(seu.objs[[1]], seu.objs[-1])
# 目的是看cDC3中cDC2来源的占比，如果队列里没有cDC3细胞的话，画图时不会出现这个队列
# 运行这句就可以看到哪些队列里有cDC3了
# Purpose: To check the proportion of cDC2-derived cells in cDC3. Cohorts without cDC3 cells will be automatically excluded in plotting.
# Run this to identify cohorts containing cDC3 cells
table(seu$cancer[seu$MajorCluster=="M04_cDC2_CD1C"])

# 取子集来降低计算量
# Subset data to reduce computational load
seu <- subset(seu, MajorCluster %in% c("M03_cDC1_CLEC9A", "M04_cDC2_CD1C", "M05_cDC3_LAMP3"))
```

# SingleR评估cDC3（LAMP3+ cDCs）的cDC2来源占比
# Using SingleR to evaluate the proportion of cDC2-derived cDC3 (LAMP3+ cDCs)

原文：我们以**cDC1和cDC2的转录组作为参考**，预测**各数据集中LAMP3+ cDCs的细胞来源**。

Original Text：Here, we used the **transcriptomes of cDC1s and cDC2s as a reference** to predict the **origin of LAMP3+ cDCs in each dataset**.
