---
title: "FigureYa195PanPair"
params:
  author: "Xiaofan Lu; Yasi Zhang"
  reviewer: "Ying Ge"
  date: "2025-05-20"
output: html_document
---

**Author(s)**: `r params$author`
**Reviewer(s)**: `r params$reviewer`
**Date**: `r params$date`

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Requirements Description

直接比较了泛癌里面paired的基因的表达差异。之前的FigureYa55panCancer_violin比较癌和normal，没有paired的数据。

A direct comparison was made of the gene expression differences in paired samples across pan-cancer. The previous FigureYa55panCancer_violin compared cancer versus normal tissues but did not use paired data.

![](example.png)

出自：<https://www.nature.com/articles/s41388-019-1026-9>

图1 人类癌症中rRNA代谢相关基因的表达图谱。
a) TCGA数据库中按癌症类型分组的所有样本的rRNA代谢评分。Y轴表示基于TCGA基因表达数据通过ssGSEA方法计算的rRNA代谢评分。箱线图展示了中位数、四分位数、最小值和最大值，每个点代表一个独立样本。p值采用Mann-Whitney检验计算。
b) 与(a)相似，但分析的是TCGA中按癌症类型分组的配对样本。每个点代表一个样本。p值通过双尾Student t检验得出。

Source: <https://www.nature.com/articles/s41388-019-1026-9>

Fig. 1 Expression landscape of rRNA metabolism-related genes in human cancer.
a) rRNA metabolic score among all samples grouped by cancer from the TCGA. Y-axis representing rRNA metabolic score, which was calculated by ssGSEA based on the gene expression in the TCGA. Boxplots show median, quartiles, min, and max, each point representing one sample. p-values are based on the Mann–Whitneytest.
b) Similar to (a), but **in paired samples** grouped by cancer from the TCGA. Each point representing one sample. p-values are based on two-tailed Student t-test.

# 应用场景
# Application Scenario

提取TCGA泛癌里配对样本的表达量，再从中提取感兴趣的通路里的基因，计算通路富集得分，画出连线图。

根据barcode提取配对样本的方法同样适用于TCGA的其他类型数据，例如DNA甲基化等。

配对样本的表达量还可以用来做更能多分析，画更多图。例如用来筛选配对间的差异表达基因，用FigureYa149rankHeatmap来展示。

还可以把基因集换成你感兴趣的signature、biomarker。你在某个cohort里发现的差异基因，再来TCGA泛癌里看看异同。

更多泛癌的图看这里<https://k.koudai.com/Wi1xos9X>

Extract expression data from paired samples in TCGA pan-cancer, then select genes from pathways of interest, calculate pathway enrichment scores, and generate a connected-line plot.

The same barcode-based method for extracting paired samples can also be applied to other TCGA data types, such as DNA methylation.

The paired-sample expression data can be used for additional analyses and visualizations. For example, it can be used to identify differentially expressed genes (DEGs) between paired samples and display them using FigureYa149rankHeatmap.

You can also replace the gene sets with your own signatures or biomarkers of interest. For instance, DEGs identified in a specific cohort can be further examined in TCGA pan-cancer to compare similarities and differences.

For more pan-cancer visualizations, check here<https://k.koudai.com/Wi1xos9X>

# 环境设置
# Environment Setup


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

install.packages(c("data.table", "knitr"))
```

# 加载包
# Loading packages

```{r}
library(data.table)
library(GSVA)

# 显示英文报错信息
# Show English error messages
Sys.setenv(LANGUAGE = "en")

# 禁止chr转成factor
# Prevent character-to-factor conversion
options(stringsAsFactors = FALSE)
```

# 输入文件的获得
# Obtaining Input Files

如果你已经准备好easy_input_pair.txt、easy_input_expr.RDS和easy_input_gene.txt文件，就可以跳过这步，直接进入“输入文件”。

If you already have the files easy_input_pair.txt, easy_input_expr.RDS and easy_input_gene.txt prepared, you can skip this step and go directly to "Input Files".

## TCGA pan-cancer配对样本的表达矩阵提取
## Extracting Paired-Sample Expression Matrix from TCGA Pan-Cancer

这步运行一次就行，可直接跳到“获取感兴趣的基因集”。

这步提取出的配对样本及其表达矩阵，已保存到easy_input_pair.txt和easy_input_expr.RDS文件。你可以用这些配对样本的表达量做很多事情，不要局限于例图哦～

GDC-PANCAN.basic_phenotype.tsv，样本信息，从XENA下载<https://xenabrowser.net/datapages/>

EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv，表达矩阵，从TCGA PanCanAtlas下载：<https://gdc.cancer.gov/about-data/publications/pancanatlas>。太大，已上传微云<https://share.weiyun.com/K5dlEurP>

This step only needs to be run once. You can go directly to "Obtaining Gene Sets of Interest".

The extracted paired samples and their expression matrices have been saved in easy_input_pair.txt and easy_input_expr.RDS. You can use this paired-sample expression data for many analyses - don't limit yourself to the example plots.

GDC-PANCAN.basic_phenotype.tsv: Sample information, downloaded from XENA<https://xenabrowser.net/datapages/>

EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv: Expression matrix, downloaded from TCGA PanCanAtlas <https://gdc.cancer.gov/about-data/publications/pancanatlas>. (File is too large; already uploaded to Weiyun<https://share.weiyun.com/K5dlEurP>
