---
title: "FigureYa154immuneSubtypes"
author: "Yufang Wang"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-9-12"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述
## Requirement description

复现文章里的结果图
Reproduce the result graph in the article

![](example.png)

出自<https://doi.org/10.1016/j.immuni.2018.03.023>
From <https://doi.org/10.1016/j.immuni.2018.03.023>


Figure 1. Immune Subtypes in Cancer (A) Expression signature modules and identification of immune subtypes.

- Top: Consensus clustering of the pairwise correlation of cancer immune gene expression signature scores (rows and columns). Five modules of shared associations are indicated by boxes.
- Middle: Representative gene expression signatures from each module (columns), which robustly reproduced module clustering, were used to cluster TGCA tumor samples (rows), resulting in six immune subtypes C1–C6 (colored circles).
- Bottom: Distributions of signature scores within the six subtypes (rows), with dashed line indicating the median.

原文作者提供了代码to reproduce the signature scores and clusters，<https://github.com/Gibbsdavidl/Immune-Subtype-Clustering>。NOTE! This method is very sensitive to the software pipeline used in quantifying genes. It does not work well with FPKM, RPKM, TPM, etc.
The original author provided code to reproduce the signature scores and clusters, <https://github.com/Gibbsdavidl/Immune-Subtype-Clustering>. NOTE! This method is very sensitive to the software pipeline used to quantify genes. It does not work well with FPKM, RPKM, TPM, etc.

后来又写了个R包`ImmuneSubtypeClassifier`，more robust，<https://github.com/Gibbsdavidl/ImmuneSubtypeClassifier>，感兴趣的小伙伴可以去研究一下。
The author later wrote a more robust R package `ImmuneSubtypeClassifier`, <https://github.com/Gibbsdavidl/ImmuneSubtypeClassifier>. Interested readers are encouraged to explore it.

## 应用场景
## Application Scenarios

按免疫浸润定量结果给样本分类。
Classify samples based on immune infiltration quantification results.

## 环境设置
## Environment Setup

```{r}
source("install_dependencies.R")
library(devtools)
install_github("Gibbsdavidl/ImmuneSubtypeClassifier")
```

加载包
Loading the package

```{r}
library(ConsensusClusterPlus)
library(corrplot)
library(ggplot2)
library(dplyr)
library(reshape2)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 #Display English error messages
options(stringsAsFactors = FALSE) #禁止chr转成factor #Disable conversion of chr to factor
```

## 输入文件
## Input file

mmc2.xlsx，即例文的Table S1，包含TCGA里33种癌症1万多个样本的免疫浸润结果。
mmc2.xlsx, also known as Table S1 in the example article, contains immune infiltration results for over 10,000 samples from 33 cancer types collected by TCGA.

We integrated major immunogenomics methods to characterize
the immune tumor microenvironment (TME) across 33 cancers
analyzed by TCGA, applying methods for the assessment of
total lymphocytic infiltrate (from genomic and H&E image data),
immune cell fractions from deconvolution analysis of mRNAseq
data, immune gene expression signatures, neoantigen prediction,
TCR and BCR repertoire inference, viral RNA expression,
and somaticDNAalterations (Table S1).

实际应用时，可以提取其中某一种或相关的某几种癌症类型，做进一步分析。
In practical applications, one can extract specific or related cancer types for further analysis.
