---
title: "FigureYa180FGAplotV2"
author: "Enyu Lin"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

# Requirement

cBioPortal上的Fraction Genome Altered的算法
The Fraction Genome Altered algorithm on cBioPortal
<https://groups.google.com/forum/#!topic/cbioportal/4DvUVuRgrTw>

![](example.png)

出自<https://www.cbioportal.org/study/summary?id=kirc_tcga_pan_can_atlas_2018>
fromhttps://www.cbioportal.org/study/summary?id=kirc_tcga_pan_can_atlas_2018

另外一篇文章用了Fraction Genome Altered的概念，但算法跟cBioPortal但不同。
Another article used the concept of Fraction Genome Altered, but the algorithm was different from that of cBioPortal.

![](example.jpg)

出自<https://www.nature.com/articles/oncsis201214>
fromhttps://www.nature.com/articles/oncsis201214

Figure 1. Fraction (%) of the FGL, FGG and FGA among ccRCC case subgroups. P-values between subgroups were as follows: males vs females (P=0.002), age (<50, vs ⩾50 years, P=0.06), any family history of cancer (P=0.02), stage (P<0.00001), grade (P<0.0001), ever vs never smoking (P=0.05), VHL wild-type cases vs those with VHL promoter hypermethylation (P=0.03), VHL wild-type cases vs those with VHL sequence alterations (P=0.17). REF, referent group.

cBioPortal提供的Fraction Genome Altered算的是总体的，没有区分gain和lost，想用cBioPortal的算法分别计算fraction of genome gained (FGG) 和fraction of the genome lost (FGL)，并使用临床信息比较各个分组间的差异。输入TCGA的CNV数据，计算FGA、FGG、FGL，复现文章的图。
The Fraction Genome Altered provided by cBioPortal is overall and does not distinguish between gain and lost. the algorithm of cBioPortal was intended to calculate fraction of genome gained (FGG) and fraction of the genome lost (FGL) respectively, and clinical information was used to compare the differences among the groups. Input the CNV data of TCGA, calculate FGA, FGG, and FGL, and reproduce the figures of the article.

# 应用场景

# Application Scenarios

自己的测序数据，用cBioPortal的算法计算Fraction Genome Altered (FGA)。
Our own sequencing data were used to calculate Fraction Genome Altered (FGA) using the algorithm of cBioPortal.

或者使用TCGA的数据，计算不同临床分组的FGA、FGG和FGL，并进行差异检验。
Alternatively, the data from TCGA can be used to calculate the FGA, FGG and FGL of different clinical groups and conduct difference tests.

# 环境设置

# Environment Setup


```{r}
source("install_dependencies.R")
```

```{r, eval=FALSE}

install.packages(c("stringr", "magrittr", "ggplot2","patchwork","Rmisc","Cairo"))
```

加载包
Loading package

```{r}
library(ggplot2)
library(magrittr)
library(patchwork)
library(Rmisc)
library(Cairo)
library(stringr)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息  # Display an English error message
options(stringsAsFactors = FALSE) #禁止chr转成factor  # prohibit chr from being converted to factor
```

自定义函数计算mean、se和p值
Custom functions calculate mean, se and p values

```{r}
cal_ms <- function(x,y){
  table1=data.frame()
  for (i in y) {
    name=summarySE(x,measurevar=c(colnames(x)[ncol(x)]), groupvars=c(colnames(x)[i]))[[1]]
    mean=summarySE(x, measurevar=c(colnames(x)[ncol(x)]), groupvars=c(colnames(x)[i]))[[3]] #mean
    se=summarySE(x, measurevar=c(colnames(x)[ncol(x)]), groupvars=c(colnames(x)[i]))[[5]] #se

    table=data.frame(group=rep(colnames(x)[i],length(mean)),subgroup=name,mean=mean,se=se)
    table1=rbind(table1,table)

  }
  table1$subgroup=factor(table1$subgroup,levels=unique(table1$subgroup))
  table1$group=factor(table1$group,levels=unique(table1$group))
  return(table1)
}#计算mean和se,排序  # Calculate mean and se, sort
cal_p=function(x,y){
  table2=data.frame()
  for (i in y) {
    m<-aov(x[,ncol(x)]~as.factor(x[,i]))
    summary(m)
    table=cbind(rep(colnames(x)[i],nrow((TukeyHSD(m))[[1]])),(TukeyHSD(m))[[1]])
    table2=rbind(table2,table)
  }
  return(table2)
}#计算p值  # Calculate the p value
signif=function(table1,table2){
  signif1=c()
  for (i in 1:length(table(table2[,1]))) {
    signif1=c(signif1,"Ref",table2[table2[,1]==levels(table1[,1])[i],5][1:length(table1[table1[,1]==names(table(table2[,1])[i]),"subgroup"])-1])
  }#提取p值  # Extract the P-value
  for (i in 1:length(signif1)) {
    if (!is.na(as.numeric(signif1[i]))) {
      signif1[i]=ifelse(signif1[i]>0.05,"",ifelse(signif1[i]<=0.0001,"****",ifelse(signif1[i]<=0.001,"***",ifelse(signif1[i]<=0.01,"**",ifelse(signif1[i]<=0.05,'*')))))
    }
  }#转换标记  # Conversion Markup
  return(signif1)
}#准备显著性标记的数据  # Prepare the data for salient marking
mar=function(tableGL){
  mar=round(max(tableGL[,"mean"]+tableGL[,"se"]),1)+0.05
  return(mar)
}#ggplot坐标长度  #ggplot coordinate length
```

# 输入文件

# Input File

从Xena获取TCGA_KIRC的cnv数据、clinical数据。
Obtain the cnv data and clinical data of TCGA_KIRC from Xena.

数据来源：Data source:
<https://xenabrowser.net/datapages/?cohort=TCGA%20Kidney%20Clear%20Cell%20Carcinoma%20(KIRC)&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443>
