---
title: "FigureYa268panCNVexpr"
params:
  author: "Xiaofan Lu; Yasi Zhang"
  reviewer: "Ying Ge"
  date: "2025-05-20"
output: html_document
---

**Author(s)**: `r params$author`
**Reviewer(s)**: `r params$reviewer`
**Date**: `r params$date`

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Demand description

小丫姐好，我还有很多小伙伴想众筹这篇铁死亡泛癌的其他图片，主要是Figure 1。感觉这个套路只要换一个通路就可以测试，我看大鱼老师写了这篇文章的Figure 2A（铁死亡基因在泛癌肿瘤与正常的差异箱线图，FigureYa208FPI），Figure 3B（铁死亡与免疫通路的相关性，FigureYa253panGSEA），可否请大鱼老师详细写一下文章Figure 1的相关代码？

- 1A（铁死亡在泛癌中体细胞拷贝数变异情况，绘制条形图，可参考FigureYa265panCNV）
- 1B（铁死亡基因在肿瘤与正常的差异表达，绘制拼接的点图和柱状图，可参考FigureYa263panDiff）
- 1C（铁死亡基因体细胞拷贝数与基因表达的相关性点图）
- 1D（铁死亡基因肿瘤与正常组织甲基化状态点图）
- 1E（铁死亡转录组表达与启动子甲基化相关性的点图）

这些图都是泛癌的，不知道大鱼老师可不可以用例文数据分别众筹个教程 [皱眉] ？（感觉都可以出一期专栏——iSicence复现了哈哈）

这次众筹Figure 1C，铁死亡基因体细胞拷贝数与基因表达的相关性点图。

Hello Sister Xiaoya, I have many friends who would like to crowdfund the other figures from this pan-cancer ferroptosis study, mainly Figure 1. It seems that this approach can be tested just by switching to a different pathway. I noticed that Teacher Dayu has already written about Figure 2A (boxplot of ferroptosis gene differences between tumors and normal tissues in pan-cancer, FigureYa208FPI) and Figure 3B (correlation between ferroptosis and immune pathways, FigureYa253panGSEA) from this article. Could Teacher Dayu please provide a detailed explanation of the code for Figure 1 in the article?

- 1A (somatic copy number variation of ferroptosis in pan-cancer, bar plot, refer to FigureYa265panCNV)
- 1B (differential expression of ferroptosis genes in tumors vs. normal tissues, combined dot and bar plot, refer to FigureYa263panDiff)
- 1C (correlation scatter plot between somatic copy number and gene expression of ferroptosis genes)
- 1D (methylation status dot plot of ferroptosis genes in tumor vs. normal tissues)
- 1E (scatter plot of correlation between ferroptosis transcriptome expression and promoter methylation)

All these figures are pan-cancer analyses. I wonder if Teacher Dayu could create separate crowdfunding tutorials using the example data? (Feels like it could even be a column—iScience reproduction, haha!)

This time, we’re crowdfunding Figure 1C: the scatter plot showing the correlation between somatic copy number and gene expression of ferroptosis genes."

![](example.png)

出自<https://linkinghub.elsevier.com/retrieve/pii/S2589004220304892>

图1. 铁死亡调控基因（FRGs）的失调。
其中，正调控因子和负调控因子分别用红色和蓝色标注。
(C) FRGs体细胞拷贝数变异与基因表达的Spearman相关性分析。

接下来会陆续众筹Figure1DE，敬请关注。

Source: <https://linkinghub.elsevier.com/retrieve/pii/S2589004220304892>

Figure 1. The Dysregulation of Ferroptosis Regulator Genes (FRGs).
For which the positive and negative regulators are marked in red and blue, respectively.
(C) The Spearman’s correlation between somatic copy number alterations and the expression of FRGs.

The crowdfunding for Figure 1DE will be launched successively, so stay tuned.

# 应用场景
# Application scenarios

分析基因（以铁死亡为例）体细胞拷贝数与基因表达的相关性，并绘制点图。

Analyze the correlation between somatic copy number and gene expression (using ferroptosis as an example) and plot a scatter diagram.

# 环境设置
# Environment Setup


```{r}
source("install_dependencies.R")

```

# 加载包
# Loading packages

```{r}
library(ggplot2)
library(data.table)
library(randomcoloR)
library(ggpubr)
library(GSVA)
library(clusterProfiler)
library(impute)
library(ComplexHeatmap)
source("twoclasslimma.R")

# 显示英文报错信息
# Show English error messages
Sys.setenv(LANGUAGE = "en")

# 禁止chr转成factor
# Prevent character-to-factor conversion
options(stringsAsFactors = FALSE)
```

# 输入文件
# Input Files

跟FigureYa263panDiff、FigureYa265panCNV使用的是同一套数据，已经下载的小伙伴就不用重复下载了。

merged_sample_quality_annotations.tsv，下载自 <https://gdc.cancer.gov/about-data/publications/pancanatlas>，下载地址 <http://api.gdc.cancer.gov/data/1a7d7be8-675d-4e60-a105-19d4121bdebf>

EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv，表达矩阵，第一列是基因，之后是其在每个样本中的表达量。下载自 <http://api.gdc.cancer.gov/data/3586c0da-64d0-4b74-a449-5ff4d9136611>

TCGA.PANCAN.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz，CNV数据，下载地址 <https://xenabrowser.net/datapages/?dataset=TCGA.PANCAN.sampleMap%2FGistic2_CopyNumber_Gistic2_all_data_by_genes&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443>

The same dataset is used as in FigureYa263panDiff and FigureYa265panCNV, so those who have already downloaded it do not need to download it again.

merged_sample_quality_annotations.tsv: Downloaded from  <https://gdc.cancer.gov/about-data/publications/pancanatlas>, download link: <http://api.gdc.cancer.gov/data/1a7d7be8-675d-4e60-a105-19d4121bdebf>

EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv: Expression matrix, where the first column contains gene names, followed by their expression levels in each sample. Downloaded from <http://api.gdc.cancer.gov/data/3586c0da-64d0-4b74-a449-5ff4d9136611>

TCGA.PANCAN.sampleMap_Gistic2_CopyNumber_Gistic2_all_data_by_genes.gz: CNV data, download link: <https://xenabrowser.net/datapages/?dataset=TCGA.PANCAN.sampleMap%2FGistic2_CopyNumber_Gistic2_all_data_by_genes&host=https%3A%2F%2Ftcga.xenahubs.net&removeHub=https%3A%2F%2Fxena.treehouse.gi.ucsc.edu%3A443>
