---
title: "FigureYa117multilinearDE"
author: "Xiaofan Lu, Taojun Ye"
reviewer: "Ying Ge"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
# 设置knitr代码块的全局选项 / Set global options for knitr code chunks
```

## 需求描述

去除RNAseq表达量混杂因素，混杂因素包括年龄、性别、人种、批次等。
方法在文章method部分的Differential expression analyses里面描述。
结果图可以放两个热图进行比较

## Requirement description

Remove confounding factors such as age, gender, race, batch, etc. that affect RNAseq expression levels.
The method is described in the differential expression analyses section of the article.
The result chart can be compared with two heat maps

![](method.png)

出自<https://academic.oup.com/hmg/article/22/24/5001/568201>

from<https://academic.oup.com/hmg/article/22/24/5001/568201>

或参考：Evaluation of logistic regression models and effect of covariates for case–control study in RNA-Seq analysis，<https://link.springer.com/article/10.1186%2Fs12859-017-1498-y>

**题外话：**例文中采用这种多变量逻辑回归的方式考虑协变量的影响，个人觉得在一些成熟的差异表达算法（如DESeq2，edgeR，limma）的design matrix中加入这些变量可能更稳定。

Or reference: Evaluation of logistic regression models and effect of covariates for case–control study in RNA-Seq analysis，< https://link.springer.com/article/10.1186%2Fs12859 -017-1498-y>

**Off topic: In the example text, this multivariate logistic regression approach is used to consider the influence of covariates. Personally, I think adding these variables to the design matrix of some mature differential expression algorithms (such as DESeq2, edgeR, limma) may be more stable.

## 应用场景

用于做差异表达分析。大规模测序数据，往往混杂着各种协变量，影响差异基因筛选，常见的两种策略：

- 策略一，去除这些混杂因素，例文采取的是这种策略；
- 策略二，把这些因素加入到DESeq2的design里，推荐这种策略。

下面这篇文章详细探讨了第一种方式的缺点，并给出了实际操作时的指导意见：2016年的Biostatistics，Methods that remove batch effects while retaining group differences may lead to exaggerated confidence in downstream analyses，<https://t.zsxq.com/3j2beI6>

## Application scenarios

Used for differential expression analysis. Large scale sequencing data is often mixed with various covariates, which affect differential gene screening. There are two common strategies:

-Strategy one is to remove these confounding factors, which is adopted in the example text;
-Strategy 2: Incorporate these factors into the design of DESeq2 and recommend this strategy.

The following article discusses in detail the drawbacks of the first approach and provides guidance for practical operation: 2016 Biostatistics，Methods that remove batch effects while retaining group differences may lead to exaggerated confidence in downstream analyses，< https://t.zsxq.com/3j2beI6 >

## 环境设置


## Environment settings


```{r}
source("install_dependencies.R")

# 这将加速R包的下载过程（This will speed up the download process of R packages）

# 这将加速Bioconductor包的下载过程（This will speed up the download process of Bioconductor packages）

```

加载包

library packages
