---
title: "FigureYa252mclust"
author: "Zongcheng Li"
reviewers: "Ying Ge, Hui Huang"
date: "2025-9-12"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

求众筹这篇文章的Fig4A，好看的聚类图

# Requirement Description

Ask for crowdfunding for this article Fig4A, a good-looking cluster map

![](example.png)

出自<https://link.springer.com/article/10.1007%2Fs00497-018-00357-2>
from<https://link.springer.com/article/10.1007%2Fs00497-018-00357-2>

Fig. 4 Covariance clustering of expressed genes across the embryo time series. a Average maximum-normalized expression values for all genes in each of the 24 clusters generated by Mclust. Numbers in each panel indicate the number of genes grouped in that cluster, and polygons represents ± 1 standard deviation.

# 应用场景

这种曲线的方式很适合展示多时间点数据，例如多个发育时期、生物/非胁迫处理后多个时间点等。可以一目了然看出每个cluster的基因随时间的表达量变化规律。

更多时间序列数据结果展示，可参考FigureYa16fitting、FigureYa94STEMbox和FigureYa98STEMheatmap。

# Application Scenarios

This curve method is well suited to display data at multiple time points, such as multiple developmental periods, multiple time points after biological/non-stress treatment, etc. You can see at a glance the variation of gene expression in each cluster over time.

For more time series data results, see FigureYa16fitting, FigureYa94STEMbox, and FigureYa98STEMheatmap.

# 环境设置

# Environment settings

```{r}
source("install_dependencies.R")
library(tidyverse)
library(magrittr)
library(mclust)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入文件获得

如果你的数据已经整理成easy_input.csv格式，就可以跳过这步，直接进入“mclust”。

从NCBI的GEO数据库下载GSE121236_RAW.tar压缩包，<https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE121236>

# Input file obtained

If your data has been organized into easy_input.csv format, you can skip this step and go straight to "mcLust".

Download the GSE121236_RAW.tar archive from the GEO database of NCBI and <https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE121236>

![](download.png)

```{r}
files <- untar(tarfile = "GSE121236_RAW.tar", list = T)
temp_dir <- tempdir()
untar(tarfile =  "GSE121236_RAW.tar", exdir = temp_dir)

samples <- paste0(rep(c("pg", "gl", "eh", "lh", "et", "lt", "bc", "mg"), each = 3), "_", 1:3)
files_used <- paste0("GSM34282", 48:71, "_", samples, ".kallisto.tsv.gz")
files_used <- gsub("_bc_([123]).kallisto", "_bc_\\1.5ng.kallisto", files_used)

# 批量读取GSE121236_RAW.tar压缩包内每个样本的测序数据
# Read the sequencing data of each sample in the GSE121236_RAW.tar package in batches
tpm <- lapply(files_used, function(i) {
  read.table(file = gzfile(description = file.path(temp_dir, i)), header = T) %$%
    tapply(tpm, INDEX = gsub("\\..*", "", target_id), sum)
  #  Gene-level transcripts per million (TPM) were estimated by combining the TPM of all isoforms of protein-coding genes
}) %>% do.call(what = cbind)

colnames(tpm) <- samples

# 保存到文件，便于套用格式，每行一个基因，每列一个sample
# Save to a file for easy formatting, one gene per row, one sample per column
#write.csv(tpm, "easy_input.csv", quote = F)
```

# mclust

original text：To examine temporal changes in gene expression, proteincoding genes detected at **≥ 1 TPM in at least one embryonic stage** of the Smart-seq2 time series were subjected to model-based clustering by Mclust [(Scrucca et al. 2016), see “Materials and methods”]. Mclust classifies points according to a set of **Gaussian mixture models** and chooses the optimal model parameters to maximize the Bayesian information criterion (BIC). The mean TPM of 18,600 genes were **converted to z-scores**, and based on analysis with Mclust, the **optimal cluster number was 24** (Online Fig. S5).
