---
title: "FigureYa233genepair"
author: "Xiaofan Lu"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

想众筹按照文中的方法构建基因对预后模型


# Requirement Description

I want to crowdfund to construct a gene pair prognostic model according to the method in this paper

![](example.png)

出自<https://molecular-cancer.biomedcentral.com/articles/10.1186/s12943-017-0666-z>
from<https://molecular-cancer.biomedcentral.com/articles/10.1186/s12943-017-0666-z>

Fig. 1 Illustration of LncRIndiv method and work-flow of this study.
b The schematic diagram of **LncRIndiv method**.
Take lncRNA-A as an example to describe the LncRIndiv method. Circles with different color represent different lncRNAs.
The **blue and purple** human shapes represent the samples **with or without reversal lncRNA pairs in each line**, respectively.
The **red and green** human shapes represent the samples are determined as **with and without differential expression of lncRNA-A** by the LncRIndiv method.

# 应用场景

【提出问题】DESeq等方法 detect differentially expressed genes (DEGs) in **sub-groups of cancer samples**, considering the **high heterogeneity** of lncRNA expression among patients, **none have been used in detecting DE lncRNAs in individual patients**.

【前提假设】作者以前的工作发现 the relative expression rankings of genes (miRNAs) tend to be **highly stable in specific normal** human tissues but **widely disturbed in the corresponding cancer** tissues, and the reversal relationship of rank between genes (miRNAs) expression level can be used to identify DE genes (miRNAs) in individual patient.
这次用同样的方式来看DE lncRNAs in individual patients

【解决方案】例文建立了一个方法LncRIndiv（detect DE lncRNAs in individual patients），即使用gene-pair进行预后模型构建的方法，也就是通过成对基因来进行预后建模。

例如A和B基因，A表达量 > B表达量设为1，A表达量 < B表达量设为0。筛选出多对基因，都变成0和1的形式，进行预后模型构建。

本文档带你复现这一方法。

【优点】基于排序获得的模型具有平台独立性，只要单样本基因之间表达具有可比性就可以了。
...due to experimental batch effects and platform differences, the **score-based signatures** tend to produce spurious risk classification in independent samples measured by different laboratories and are infeasible in clinical application...
...we found prognostic signatures derived using the **relative genes (miRNAs) expression rankings** within samples, rather than the absolute expression values, are robust in **independent datasets from different laboratories and platforms**.

# Application Scenarios
【Ask a question】 DESeq and other methods detect differentially expressed genes (DEGs) in **sub-groups of cancer samples**, considering the **high heterogeneity** of lncRNA expression among patients, **none have been used in detecting DE lncRNAs in individual patients**.

【Hypothesis】The authors' previous work found that the relative expression rankings of genes (miRNAs) tend to be **highly stable in specific normal** human tissues but **widely disturbed in the corresponding cancer** tissues, and the reversal relationship of rank between genes (miRNAs) expression level can be used to identify DE genes (miRNAs) in individual patient. This time, we will look at DE lncRNAs in individual patients in the same way

【Solution】In this paper, a method LncRIndiv (detect DE lncRNAs in individual patients) is established, that is, a method of prognostic model construction using gene-pair, that is, prognostic modeling through pairs of genes.

For example, for genes A and B, the expression level of A > B is set to 1, and the expression level of A < B is set to 0. Multiple pairs of genes were screened out and turned into 0 and 1 forms for prognostic model construction.

I'm going to take you through the repetition of this approach.

【Advantages】The model obtained based on sequencing is platform independent, as long as the expression of single sample genes is comparable.
... due to experimental batch effects and platform differences, the **score-based signatures** tend to produce spurious risk classification in independent samples measured by different laboratories and are infeasible in clinical application...
... we found prognostic signatures derived using the **relative genes (miRNAs) expression rankings** within samples, rather than the absolute expression values, are robust in **independent datasets from different laboratories and platforms**.

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

install.packages("matlab")
install.packages("EnvStats")
```

加载包
load packages

```{r}
library(matlab)
library(EnvStats)
library(pheatmap)

Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

自定义函数
Custom functions

```{r}
display.progress = function (index, totalN, breakN=20) {
  if ( index %% ceiling(totalN/breakN)  ==0  ) {
    cat(paste(round(index*100/totalN), "% ", sep=""))
  }
}
```

# 输入文件

这里用的数据是原文的测试数据GSE27262，包括50例样本，25肿瘤和25正常。GEO数据下载和预处理方式可参考FigureYa203combat。

# Input files

The data used here is the original test data GSE27262, including 50 samples, 25 tumors and 25 normal. For details about how to download and preprocess GEO data, see FigureYa203combat.
