---
title: "FigureYa272scBulkCCCI"
params:
  author: "Xiaojian Liu; Yasi Zhang"
  reviewer: "Ying Ge"
  date: "2025-05-20"
output: html_document
---

**Author(s)**: `r params$author`
**Reviewer(s)**: `r params$reviewer`
**Date**: `r params$date`

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Demand description

基于细胞类型signature识别细胞-细胞互作

Identifying cell-cell interactions based on cell type signatures.

![](workflow.png)

出自：<https://linkinghub.elsevier.com/retrieve/pii/S009286742030341X>

图S8. 相关性细胞互作网络分析框架及人和小鼠肿瘤中髓系细胞亚群概述（与图3和STAR方法相关）。
(A) 相关性网络分析流程图。
从Smart-seq2单细胞RNA测序数据中鉴定出细胞**亚型特异性基因**，并用于**估算TCGA/GTEx批量RNA测序数据中的相对丰度**。在过滤掉自身表达的基因后，计算**高相关性基因**。最后，基于Smart-seq2表达谱进行**富集分析**，以**识别所有相关的细胞亚型并构建相关性网络**（STAR方法）。

Source: <https://linkinghub.elsevier.com/retrieve/pii/S009286742030341X>

Figure S8. Analysis Framework for Correlative Cell-Cell Interaction Networks and Summary of Myeloid Cell Subsets in Human and Mouse Tumors, Related to Figure 3 and STAR Methods.
(A) Workflow illustrating correlative network analysis.
Cell **subtype specific genes** were identified from the Smart-seq2 scRNA-seq dataset and were used to **estimate the relative abundance in TCGA/GTEx bulk RNA-seq datasets.** After filtering the self-expressed genes, **high correlative genes** were calculated. Finally, an **enrichment analysis** based on Smart-seq2 profile was performed to **identify all correlated cell subtypes and build the correlative network** (STAR Methods).

# 应用场景
# Application scenarios

借助scRNA-seq找到的signature genes，推断bulk RNA-seq样本的细胞类群互作网络，最后用igraph（静态）和networkD3（动态）两种网络可视化工具来展示结果。

本文档作者写了个R包scBulkCCCI，以便小伙伴运行例文的workflow。

scBulkCCCI：correlative cell-cell interaction networks based on combining scRNA-seq and TCGA datasets

R包下载地址：<https://www.yuque.com/figureya/figureyaplus/package>

下载后解压缩，可以查看R包内容：

- 包的自带描述文档，见ScBulkCCCI/README.md
- 示例部分数据，储存在ScBulkCCCI/data，scRNA-seq数据太大，未内置在包中，根据说明去下载；
- 数据经过预处理：数据预处理的code查看ScBulkCCCI包下ScBulkCCCI_DataPreprocess.R， 位于ScBulkCCCI/example；
- 示例运行代码ScBulkCCCI.R，于ScBulkCCCI/example；

Using signature genes identified from scRNA-seq, we inferred cell population interaction networks in bulk RNA-seq samples, and finally visualized the results using two network visualization tools: igraph (static) and networkD3 (dynamic).

The author of this document developed an R package, scBulkCCCI, to help users run the workflow described in the example.

scBulkCCCI: Correlative cell-cell interaction networks based on combining scRNA-seq and TCGA datasets

R package download link: <https://www.yuque.com/figureya/figureyaplus/package>

After downloading and extracting the package, you can view its contents:

- Package documentation: See ScBulkCCCI/README.md.
- Example datasets: Stored in ScBulkCCCI/data. Note that the scRNA-seq data is too large to be included in the package—please download it separately as instructed.
- Preprocessed data: The preprocessing code can be found in ScBulkCCCI/example/ScBulkCCCI_DataPreprocess.R.
- Example execution code: ScBulkCCCI.R, located in ScBulkCCCI/example

# 环境设置
# Environment Setup


```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}

install.packages("ggthemes")

# 下载后本地安装
# 下载地址：https://www.yuque.com/figureya/figureyaplus/package
# Download and install locally
# Download link: https://www.yuque.com/figureya/figureyaplus/package
install.packages("ScBulkCCCI_0.1.3.tar.gz", repos = NULL, type = "source")
```

# 加载包
# Loading packages

```{r}
library("ScBulkCCCI")
library("pheatmap")
library("RColorBrewer")
library("ggplot2")
library("ggthemes")
library("plyr")
library("dplyr")
library("stringr")
library("tibble")
library("tidyr")

# 显示英文报错信息
# Show English error messages
Sys.setenv(LANGUAGE = "en")

# 禁止chr转成factor
# Prevent character-to-factor conversion
options(stringsAsFactors = FALSE)
```

# 输入文件
# Input Files

输入文件包含两部分：scRNA-seq、bulk RNA-seq（以TCGA数据为例，也可以用其他RNA-seq数据）。

The input files consist of two parts: scRNA-seq and bulk RNA-seq (using TCGA data as an example, though other RNA-seq data can also be used).

## scRNA-seq 数据
## scRNA-seq data

下载地址：[GSE146771](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE146771)

Download link: [GSE146771](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE146771)

![](download.png)

CRC.Leukocyte.Smart-seq2.Metadata.txt: 细胞类群信息

- CellName: 细胞ID
- Tissue: 细胞所属组织类型
- Global_Cluster: 一级分类
- Sub_Cluster: 二级分类

CRC.Leukocyte.Smart-seq2.TPM.txt: 细胞基因定量矩阵，列是细胞，行是基因


CRC.Leukocyte.Smart-seq2.Metadata.txt: Cell population information

- CellName: Cell ID
- Tissue: Tissue type of the cell
- Global_Cluster: First-level classification
- Sub_Cluster: Second-level classification

CRC.Leukocyte.Smart-seq2.TPM.txt: Cell gene expression matrix (columns: cells, rows: genes)
