---
title: "FigureYa149rankHeatmap"
author: "Hao Li"
reviewer: "Ying Ge, Junyi Shen"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

## 需求描述
## Requirement description

分类变量的热图。
Heat map of categorical variables.

![](example.png)

出自<https://bmccancer.biomedcentral.com/articles/10.1186/s12885-019-5768-0>
From <https://bmccancer.biomedcentral.com/articles/10.1186/s12885-019-5768-0>

Fig. 1 Differential gene expression of 24 matrix metalloproteinases (MMPs) in 15 different cancer types. Fold change and p-values shown were obtained through comparison of **unmatched control tissue (N between 11 and 114) to tumor tissue** (N between 66 and 1097). Fold change was calculated as the median expression of a gene in tumor divided by the median gene expression in adjacent normal tissue.

**图的解读**
**Interpretation of the figure**

- 每行一个基因，每列一个癌症类型
- 颜色分为6档：5档变化倍数（红色上调，蓝色下调）和pvalue不显著（白色）
- 如果pvalue>0.05，无论变化倍数多大都是白色
- 白色和灰色时，字为黑色；蓝色和红色时，字为白色。
- One gene per row, one cancer type per column
- Colors are divided into 6 levels: 5 levels of change fold (red up, blue down) and pvalue not significant (white)
- If pvalue>0.05, it will be white regardless of the change fold
- When it is white and gray, the words are black; when it is blue and red, the words are white.

## 应用场景
## Application scenarios

展示某些基因差异表达结果的时候，经常是把连续性变量当作数值来直接进行热图绘制。也可以像例文那样，先设置成分类变量的，同时用数字来进一步注释。变成分类变量后，可能更清晰的看到趋势
When displaying the results of differential expression of certain genes, continuous variables are often used as numerical values to directly draw heat maps. You can also set it as a categorical variable as in the example text, and use numbers to further annotate it. After becoming a categorical variable, you may see the trend more clearly.

还可以用来转换和展示相关性或其他连续变量。
It can also be used to convert and display correlations or other continuous variables.

这里用complexheatmap画图，其实还可以用pheatmap画图，可参考FigureYa114ternaryCluster，按表达量分为3组（高/低/不变）来画图。
Here we use complexheatmap to draw the picture, but you can also use pheatmap to draw the picture. You can refer to FigureYa114ternaryCluster, which is divided into 3 groups (high/low/unchanged) according to the expression level to draw the picture.

## 环境设置
## Environment settings


```{r}
source("install_dependencies.R")

```

加载包
Load package

```{r}
library(ComplexHeatmap)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息
options(stringsAsFactors = FALSE) #禁止chr转成factor
```

## 输入文件
## Input file

- easy_input_logFc.csv，差异表达分析当中的**logFC**值，还可以替换为相关分析的相关系数。
- easy_input_PVal.csv，差异表达分析当中的**P**值。
- genetype.csv，基因的分组，用于画最左侧annotation，非必须。
- easy_input_logFc.csv, **logFC** value in differential expression analysis, can also be replaced by correlation coefficient of correlation analysis.
- easy_input_PVal.csv, **P** value in differential expression analysis.
- genetype.csv, gene grouping, used to draw the leftmost annotation, not required.

```{r}
logfc <- read.csv("easy_input_logFc.csv", row.names = 1)
logfc[1:2,]
p.val <- read.csv("easy_input_PVal.csv", row.names = 1)
p.val[1:2,]

genetype <- read.csv("genetype.csv")
head(genetype)
#为确保热图基因的顺序和分组文件的顺序是一致的
#因此按照基因分组文件里的顺序，对logfc和p.val两个文件排序
#To ensure that the order of heat map genes is consistent with the order of grouping files
#So sort the logfc and p.val files according to the order in the gene grouping file
logfcOrdered <- logfc[genetype$gene,]
p.valOrdered <- p.val[genetype$gene,]
```

## 连续变量转换为分类变量
## Convert continuous variables to categorical variables

分类规则：
Classification rules:

- 先按logFC值分成**5类**；
- 如果**pvalue>0.05**，无论倍数多大都作为第6类。
- First divide into **5 categories** according to logFC value;
- If **pvalue>0.05**, it will be regarded as the 6th category regardless of the multiple.
