---
title: "FigureYa42oncoprintV2_update"
author: "Ying Ge, Yi Xiong; Yijing Chen"
date: "2025-5-20"
output: html_document
---

# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述
# Requirement description

没有具体突变位点信息，只有临床信息和突变名称，对应着每个个体，01矩阵，需要画这样的oncoprint图，按照pathway分开。

There is no specific mutation site information, only clinical information and mutation names, corresponding to each individual, 01 matrix, which needs to be drawn such oncoprint plots, separated by pathway.

![](example1.png)

或者这种

Or this

![](example2.png)
出自<https://www.nejm.org/doi/full/10.1056/NEJMoa1301689>

from<https://www.nejm.org/doi/full/10.1056/NEJMoa1301689>

# 应用场景
# Application scenario

展示多个样本中多个基因的变异，以及临床信息。

- 场景一：01矩阵代表的一种变异类型。例如从基因检测报告中收集的变异基因。

- 场景二：多种变异类型，外显子测序数据。例如用FigureYa18oncoplot输出变异和临床信息，然后用本套代码修改成自己想要的效果。

如果在已发表的研究中使用，请注明出处：

Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional genomic data. Bioinformatics 2016.

Demonstrate variants in multiple genes in multiple samples, along with clinical information.

- Scenario 1: One variant type represented by the 01 matrix. For example, a variant gene collected from a genetic test report.

- Scenario 2: Multiple variant types with exome sequencing data. For example, use FigureYa18oncoplot to output variants and clinical information, and then use this code to modify it to your desired effect.

If you use it in published research, please cite:

Gu, Z. Complex heatmaps reveal patterns and correlations in multidimensional genomic data. Bioinformatics 2016.

# 环境设置
# Environment setting


```{r}
source("install_dependencies.R")

```

加载包

load package

```{r}
library(ComplexHeatmap)
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 display English error messages
options(stringsAsFactors = FALSE) #禁止chr转成factor prohibit the conversion of chr to factor
```

# 场景一：01矩阵代表的一种变异类型
# Scenario 1: A variant type represented by the 01 matrix

## 生成输入文件
## Generate input file

这一步仅用于产生示例数据。可跳过这步，直接进入“输入文件”。

此处借用FigureYa18oncoplot的输出文件`onco_matrix.txt`，整理成小伙伴常用的**01矩阵**。

This step is only for generating example data. You can skip this step and go directly to "Input file".

The output file `onco_matrix.txt` of FigureYa18oncoplot is used here, and is organized into the commonly used **01 matrix**.

```{r}
onco_matrix <- read.table("onco_matrix.txt", header = T, row.names = 1, sep = "\t", as.is = T)
#查看部分数据格式
#view some of the data formats
onco_matrix[1:3,1:4]

#把"Missense_Mutation"和"Sense_Mutation"替换为1，其他替换为0
#replace "Missense_Mutation" and "Sense_Mutation" with 1 and others with 0.
onco_matrix[onco_matrix == "Missense_Mutation"] <- "1"
onco_matrix[onco_matrix == "Sense_Mutation"] <- "1"
onco_matrix[onco_matrix != "1"] <- "0"

write.csv(onco_matrix, "easy_input.csv", quote = F)
```

## 输入文件
## Input file

`easy_input.csv`，01矩阵，每行一个基因，每列一个sample。01矩阵代表基因在这个sample里是否变异。

`easy_input_pathway.txt`，基因所在的pathway；

`easy_input_cli.csv`，临床信息，例如人种、病史、治疗方式等。

不同文件之间，基因名、sampleID必须一致。

`easy_input.csv`, 01 matrix, one gene per row, one sample per column. The 01 matrix represents whether the gene is mutated or not in this sample.

`easy_input_pathway.txt`, the pathway where the gene is located;

`easy_input_cli.csv`, clinical information such as ethnicity, medical history, treatment modality, etc.

The gene name and sampleID must be consistent between different files.
