---
title: "FigureYa253panGSEA"
author: "Xiaofan Lu"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

基于FPI最高30%和最低30%的两个肿瘤组的转录组，对每个肿瘤进行基因集富集分析(GSEA)，研究高FPI和低FPI肿瘤组织之间的通路差异。

# Requirement Description

Based on the transcriptome of the two tumor groups with the highest and lowest 30% FPI, gene set enrichment analysis (GSEA) was performed for each tumor to study the pathway differences between high and low FPI tumor tissues.

![](example.png)

出自<https://www.cell.com/iscience/fulltext/S2589-0042(20)30489-2?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2589004220304892%3Fshowall%3Dtrue>
from<https://www.cell.com/iscience/fulltext/S2589-0042(20)30489-2?_returnURL=https%3A%2F%2Flinkinghub.elsevier.com%2Fretrieve%2Fpii%2FS2589004220304892%3Fshowall%3Dtrue>

Figure 3. Relationships between Ferroptosis and Signaling Pathways and Immunophenotypes
(A and B) Enrichment analysis for metabolism pathway (A) and cancer signaling (B) between high- and low-FPI tumor tissues. NES is the normalized enrichment score in the GSEA algorithm.

# 应用场景

根据某一个基因的表达量高低给样本分组，然后做分组间的差异表达分析，进一步用差异分析结果做GSEA富集分析，可以推测该基因影响了哪些通路。

进一步拓展，可将上述“某一个基因的表达量”替换为“某个基因集的富集得分”，就像这篇例文，用一组基因计算出的FPI值给样本分组，然后做分组间的差异表达分析，进一步用差异分析结果做GSEA富集分析，可以推测该基因集（FPI）跟哪些通路相关。

# Application Scenarios

The samples were grouped according to the expression level of a certain gene, and then the differential expression analysis between the groups was performed, and the GSEA enrichment analysis was performed with the results of differential analysis, which pathways affected by the gene could be inferred.

For further expansion, you can replace the above "expression level of a certain gene" with "enrichment score of a certain gene set", just like this example, use the FPI value calculated by a group of genes to group the samples, and then do differential expression analysis between groups, and further use the results of differential analysis to do GSEA enrichment analysis, you can infer which pathways the gene set (FPI) is associated with.

> 本文档将提供这两种情况的解决方案。

更多富集分析展示方式看这里<https://k.youshop10.com/sm2p0xYn>

更多pancancer分析画图FigureYa看这里<https://k.youshop10.com/C8RkTgR3>

> This document will provide solutions for both scenarios.

For more enrichment analysis display methods, see here <https://k.youshop10.com/sm2p0xYn>

For more pancancer analysis plotting FigureYa see here <https://k.youshop10.com/C8RkTgR3>

# 环境设置


# Environment settings


```{r}
source("install_dependencies.R")
options(download.file.method = 'libcurl')
options(url.method='libcurl')
BiocManager::install("clusterProfiler")
```

加载包
load packages

```{r}
library(clusterProfiler)
library(limma)
library(ggplot2)
library(data.table)
library(ggpubr)
library(GSVA)
source("twoclasslimma.R")
Sys.setenv(LANGUAGE = "en") #显示英文报错信息 # error messages are displayed in English
options(stringsAsFactors = FALSE) #禁止chr转成factor # chr is not allowed to be converted to factor
```

# 输入文件

easy_input_sample_annotation.txt，样本分组信息。整理自注释文件merged_sample_quality_annotations.tsv，下载自<http://api.gdc.cancer.gov/data/1a7d7be8-675d-4e60-a105-19d4121bdebf>。

EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2.geneExp.tsv，表达矩阵，第一列是基因，之后是其在每个样本中的表达量。下载自<http://api.gdc.cancer.gov/data/3586c0da-64d0-4b74-a449-5ff4d9136611>。这里以TCGA数据为例，对于非肿瘤数据或其他物种数据，套用其格式即可，可忽略文中对筛选正常样本的代码，搜索“非TCGA数据”着重修改。

# Input files

easy_input_sample_annotation.txt. Sample grouping information. Organize the self-annotated file merged_sample_quality_annotations.tsv, download it from <http://api.gdc.cancer.gov/data/1a7d7be8-675d-4e60-a105-19d4121bdebf>.

EBPlusPlusAdjustPANCAN_IlluminaHiSeq_RNASeqV2. geneExp.tsv, the expression matrix, with the gene in the first column, followed by its expression in each sample. Download from <http://api.gdc.cancer.gov/data/3586c0da-64d0-4b74-a449-5ff4d9136611>. Taking TCGA data as an example, for non-tumor data or other species data, you can apply its format, you can ignore the code for screening normal samples in the text, and search for "non-TCGA data" to focus on modification.
