---
title: "FigureYa213customizeHeatmap"
author: "Xiaofan Lu"
reviewers: "Ying Ge,Hui Huang"
date: "2025-5-20"
output: html_document
---
# Academic Citation
If you use this code in your work or research, we kindly request that you cite our publication:

Xiaofan Lu, et al. (2025). FigureYa: A Standardized Visualization Framework for Enhancing Biomedical Data Interpretation and Research Efficiency. iMetaMed. https://doi.org/10.1002/imm3.70005


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# 需求描述

Figure6 的绘制方法以及Cmap score的计算，图里其他的内容如何获取可以无需讲解。

# Requirement Description

The drawing method of Figure 6 and the calculation of Cmap score, and how to obtain the rest of the content in the figure do not need to be explained.

![](example.png)

出自<https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbaa164/5891146>
from<https://academic.oup.com/bib/advance-article/doi/10.1093/bib/bbaa164/5891146>

Figure 6. Identification of most promising therapeutic agents for high PPS score patients according to the **evidence from multiple sources** (see also Table S15 available online at https://academic.oup.com/bib). **Six CTRP-derived agents** and **six PRISM-derived agents** were shown on the **left** and **right** of the diagram, respectively.

FigureYa131CMap也用到了CMap，并对感兴趣的小分子进行MoA药物作用机制分析，可相互参考。

CMap is also used in FigureYa131CMap, and MoA drug mechanism of action analysis is performed on the small molecule of interest, which can be cross-referenced.

# 应用场景

例文采用了多个来源的数据相互佐证，来说明筛选出的化合物/药物的潜在价值。作者设计了这个图，巧妙地把证据一、二、三和四同时展示出来，同时还能看到overlap。

- 证据一和二：上文用PRISM和CTRP发现了12个 candidate compounds identified showed a higher drug sensitivity in PPS score-high patients（可参考FigureYa212drugTarget）
- 证据三：We first used the **CMap analysis** to find compounds of which gene expression patterns were oppositional to the HCC-specific expression patterns (i.e. gene expression increased in tumor tissues but decreased by treatment of certain compounds).
- 证据四：Secondly, **fold-change differences of the expression levels (including mRNA- and protein-level) of candidates’ drug targets** between tumor and normal tissue were calculated, and a higher fold change value indicated a greater potential of candidate agent for HCC treatment.
- 证据五：Thirdly, a comprehensive **literature search** was performed in PubMed (https://www.ncbi.nlm.nih. gov/pubmed/) to find out the experimental and clinical evidence of candidate compounds in treating HCC.

# Application Scenarios

Data from multiple sources are used to corroborate each other to illustrate the potential value of the screened compounds/drugs. The author has designed this diagram to cleverly display evidence one, two, three, and four at the same time, while also seeing the overlap.

- Evidence 1 & 2: 12 candidate compounds identified showed a higher drug sensitivity in PPS score-high patients using PRISM and CTRP above (see FigureYa212drugTarget)
- Evidence 3: We first used the **CMap analysis** to find compounds of which gene expression patterns were oppositional to the HCC-specific expression patterns (i.e. gene expression increased in tumor tissues but decreased by treatment of certain compounds).
- Evidence 4: Secondly, **fold-change differences of the expression levels (including mRNA- and protein-level) of candidates' drug targets** between tumor and normal tissue were calculated, and a higher fold change value indicated a greater potential of candidate agent for HCC treatment.
- Evidence 5: Thirdly, a comprehensive **literature search** was performed in PubMed (https://www.ncbi.nlm.nih. gov/pubmed/) to find out the experimental and clinical evidence of candidate compounds in treating HCC.

下面将带你实现CMap analysis及画图。以左图CTRP vs. CMap为例，右图也是一样的。

> 使用本代码或分析思路请引用：

The following will take you through the implementation of CMap analysis and drawing. Take CTRP vs. CMap on the left as an example, and the same is true for the right.

> To use this code or analysis idea, please quote:

Chen Yang, Xiaowen Huang, Yan Li, Junfei Chen, Yuanyuan Lv, Shixue Dai, Prognosis and personalized treatment prediction in TP53-mutant hepatocellular carcinoma: an in silico strategy towards precision oncology, Briefings in Bioinformatics, bbaa164, https://doi.org/10.1093/bib/bbaa164

# 环境设置

画图用到了github最新版ComplexHeatmap里的pheatmap，两种方式都可以用上它，任选其一。

方式一：安装github版的ComplexHeatmap（大鱼海棠亲测有效）

# Environment settings

You can use the pheatmap in the latest version of ComplexHeatmap on github for drawing, and you can use it in either way.

Method 1: Install the github version of ComplexHeatmap (I personally tested it and found it to work)

```{r}
source("install_dependencies.R")
```

```{r eval=FALSE}
library(devtools)
install_github("jokergoo/ComplexHeatmap")
# 如果github网速困难，可本地安装
# If the network speed of GitHub is slow, you can install it locally
install.packages("ComplexHeatmap-master.tar.gz", repos = NULL, type = "source")
```

方式二：如果安装github版有困难，可以安装bioconductor版，然后加载最新的函数（小丫亲测有效）
Method 2: If you have difficulty installing the GitHub version, you can install the Bioconductor version and then load the latest function (Xiaoya's personal test is effective)
